{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "dabbe1d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "6e86e09e",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2023-11-10 11:25:49.957891: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcudart.so.11.0'; dlerror: libcudart.so.11.0: cannot open shared object file: No such file or directory\n",
      "2023-11-10 11:25:49.957927: I tensorflow/stream_executor/cuda/cudart_stub.cc:29] Ignore above cudart dlerror if you do not have a GPU set up on your machine.\n",
      "2023-11-10 11:25:52.090658: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcuda.so.1'; dlerror: libcuda.so.1: cannot open shared object file: No such file or directory\n",
      "2023-11-10 11:25:52.090691: W tensorflow/stream_executor/cuda/cuda_driver.cc:269] failed call to cuInit: UNKNOWN ERROR (303)\n",
      "2023-11-10 11:25:52.090722: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:156] kernel driver does not appear to be running on this host (ruby.gladstone.internal): /proc/driver/nvidia/version does not exist\n",
      "2023-11-10 11:25:52.090909: I tensorflow/core/platform/cpu_feature_guard.cc:151] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA\n",
      "To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Model: \"model_1\"\n",
      "__________________________________________________________________________________________________\n",
      " Layer (type)                   Output Shape         Param #     Connected to                     \n",
      "==================================================================================================\n",
      " sequence (InputLayer)          [(None, 1048576, 4)  0           []                               \n",
      "                                ]                                                                 \n",
      "                                                                                                  \n",
      " stochastic_reverse_complement   ((None, 1048576, 4)  0          ['sequence[0][0]']               \n",
      " (StochasticReverseComplement)  , ())                                                             \n",
      "                                                                                                  \n",
      " stochastic_shift (StochasticSh  (None, 1048576, 4)  0           ['stochastic_reverse_complement[0\n",
      " ift)                                                            ][0]']                           \n",
      "                                                                                                  \n",
      " re_lu (ReLU)                   (None, 1048576, 4)   0           ['stochastic_shift[0][0]']       \n",
      "                                                                                                  \n",
      " conv1d (Conv1D)                (None, 1048576, 96)  4224        ['re_lu[0][0]']                  \n",
      "                                                                                                  \n",
      " batch_normalization (BatchNorm  (None, 1048576, 96)  384        ['conv1d[0][0]']                 \n",
      " alization)                                                                                       \n",
      "                                                                                                  \n",
      " max_pooling1d (MaxPooling1D)   (None, 524288, 96)   0           ['batch_normalization[0][0]']    \n",
      "                                                                                                  \n",
      " re_lu_1 (ReLU)                 (None, 524288, 96)   0           ['max_pooling1d[0][0]']          \n",
      "                                                                                                  \n",
      " conv1d_1 (Conv1D)              (None, 524288, 96)   46080       ['re_lu_1[0][0]']                \n",
      "                                                                                                  \n",
      " batch_normalization_1 (BatchNo  (None, 524288, 96)  384         ['conv1d_1[0][0]']               \n",
      " rmalization)                                                                                     \n",
      "                                                                                                  \n",
      " max_pooling1d_1 (MaxPooling1D)  (None, 262144, 96)  0           ['batch_normalization_1[0][0]']  \n",
      "                                                                                                  \n",
      " re_lu_2 (ReLU)                 (None, 262144, 96)   0           ['max_pooling1d_1[0][0]']        \n",
      "                                                                                                  \n",
      " conv1d_2 (Conv1D)              (None, 262144, 96)   46080       ['re_lu_2[0][0]']                \n",
      "                                                                                                  \n",
      " batch_normalization_2 (BatchNo  (None, 262144, 96)  384         ['conv1d_2[0][0]']               \n",
      " rmalization)                                                                                     \n",
      "                                                                                                  \n",
      " max_pooling1d_2 (MaxPooling1D)  (None, 131072, 96)  0           ['batch_normalization_2[0][0]']  \n",
      "                                                                                                  \n",
      " re_lu_3 (ReLU)                 (None, 131072, 96)   0           ['max_pooling1d_2[0][0]']        \n",
      "                                                                                                  \n",
      " conv1d_3 (Conv1D)              (None, 131072, 96)   46080       ['re_lu_3[0][0]']                \n",
      "                                                                                                  \n",
      " batch_normalization_3 (BatchNo  (None, 131072, 96)  384         ['conv1d_3[0][0]']               \n",
      " rmalization)                                                                                     \n",
      "                                                                                                  \n",
      " max_pooling1d_3 (MaxPooling1D)  (None, 65536, 96)   0           ['batch_normalization_3[0][0]']  \n",
      "                                                                                                  \n",
      " re_lu_4 (ReLU)                 (None, 65536, 96)    0           ['max_pooling1d_3[0][0]']        \n",
      "                                                                                                  \n",
      " conv1d_4 (Conv1D)              (None, 65536, 96)    46080       ['re_lu_4[0][0]']                \n",
      "                                                                                                  \n",
      " batch_normalization_4 (BatchNo  (None, 65536, 96)   384         ['conv1d_4[0][0]']               \n",
      " rmalization)                                                                                     \n",
      "                                                                                                  \n",
      " max_pooling1d_4 (MaxPooling1D)  (None, 32768, 96)   0           ['batch_normalization_4[0][0]']  \n",
      "                                                                                                  \n",
      " re_lu_5 (ReLU)                 (None, 32768, 96)    0           ['max_pooling1d_4[0][0]']        \n",
      "                                                                                                  \n",
      " conv1d_5 (Conv1D)              (None, 32768, 96)    46080       ['re_lu_5[0][0]']                \n",
      "                                                                                                  \n",
      " batch_normalization_5 (BatchNo  (None, 32768, 96)   384         ['conv1d_5[0][0]']               \n",
      " rmalization)                                                                                     \n",
      "                                                                                                  \n",
      " max_pooling1d_5 (MaxPooling1D)  (None, 16384, 96)   0           ['batch_normalization_5[0][0]']  \n",
      "                                                                                                  \n",
      " re_lu_6 (ReLU)                 (None, 16384, 96)    0           ['max_pooling1d_5[0][0]']        \n",
      "                                                                                                  \n",
      " conv1d_6 (Conv1D)              (None, 16384, 96)    46080       ['re_lu_6[0][0]']                \n",
      "                                                                                                  \n",
      " batch_normalization_6 (BatchNo  (None, 16384, 96)   384         ['conv1d_6[0][0]']               \n",
      " rmalization)                                                                                     \n",
      "                                                                                                  \n",
      " max_pooling1d_6 (MaxPooling1D)  (None, 8192, 96)    0           ['batch_normalization_6[0][0]']  \n",
      "                                                                                                  \n",
      " re_lu_7 (ReLU)                 (None, 8192, 96)     0           ['max_pooling1d_6[0][0]']        \n",
      "                                                                                                  \n",
      " conv1d_7 (Conv1D)              (None, 8192, 96)     46080       ['re_lu_7[0][0]']                \n",
      "                                                                                                  \n",
      " batch_normalization_7 (BatchNo  (None, 8192, 96)    384         ['conv1d_7[0][0]']               \n",
      " rmalization)                                                                                     \n",
      "                                                                                                  \n",
      " max_pooling1d_7 (MaxPooling1D)  (None, 4096, 96)    0           ['batch_normalization_7[0][0]']  \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                                                                                  \n",
      " re_lu_8 (ReLU)                 (None, 4096, 96)     0           ['max_pooling1d_7[0][0]']        \n",
      "                                                                                                  \n",
      " conv1d_8 (Conv1D)              (None, 4096, 96)     46080       ['re_lu_8[0][0]']                \n",
      "                                                                                                  \n",
      " batch_normalization_8 (BatchNo  (None, 4096, 96)    384         ['conv1d_8[0][0]']               \n",
      " rmalization)                                                                                     \n",
      "                                                                                                  \n",
      " max_pooling1d_8 (MaxPooling1D)  (None, 2048, 96)    0           ['batch_normalization_8[0][0]']  \n",
      "                                                                                                  \n",
      " re_lu_9 (ReLU)                 (None, 2048, 96)     0           ['max_pooling1d_8[0][0]']        \n",
      "                                                                                                  \n",
      " conv1d_9 (Conv1D)              (None, 2048, 96)     46080       ['re_lu_9[0][0]']                \n",
      "                                                                                                  \n",
      " batch_normalization_9 (BatchNo  (None, 2048, 96)    384         ['conv1d_9[0][0]']               \n",
      " rmalization)                                                                                     \n",
      "                                                                                                  \n",
      " max_pooling1d_9 (MaxPooling1D)  (None, 1024, 96)    0           ['batch_normalization_9[0][0]']  \n",
      "                                                                                                  \n",
      " re_lu_10 (ReLU)                (None, 1024, 96)     0           ['max_pooling1d_9[0][0]']        \n",
      "                                                                                                  \n",
      " conv1d_10 (Conv1D)             (None, 1024, 96)     46080       ['re_lu_10[0][0]']               \n",
      "                                                                                                  \n",
      " batch_normalization_10 (BatchN  (None, 1024, 96)    384         ['conv1d_10[0][0]']              \n",
      " ormalization)                                                                                    \n",
      "                                                                                                  \n",
      " max_pooling1d_10 (MaxPooling1D  (None, 512, 96)     0           ['batch_normalization_10[0][0]'] \n",
      " )                                                                                                \n",
      "                                                                                                  \n",
      " re_lu_11 (ReLU)                (None, 512, 96)      0           ['max_pooling1d_10[0][0]']       \n",
      "                                                                                                  \n",
      " conv1d_11 (Conv1D)             (None, 512, 48)      13824       ['re_lu_11[0][0]']               \n",
      "                                                                                                  \n",
      " batch_normalization_11 (BatchN  (None, 512, 48)     192         ['conv1d_11[0][0]']              \n",
      " ormalization)                                                                                    \n",
      "                                                                                                  \n",
      " re_lu_12 (ReLU)                (None, 512, 48)      0           ['batch_normalization_11[0][0]'] \n",
      "                                                                                                  \n",
      " conv1d_12 (Conv1D)             (None, 512, 96)      4608        ['re_lu_12[0][0]']               \n",
      "                                                                                                  \n",
      " batch_normalization_12 (BatchN  (None, 512, 96)     384         ['conv1d_12[0][0]']              \n",
      " ormalization)                                                                                    \n",
      "                                                                                                  \n",
      " dropout (Dropout)              (None, 512, 96)      0           ['batch_normalization_12[0][0]'] \n",
      "                                                                                                  \n",
      " add (Add)                      (None, 512, 96)      0           ['max_pooling1d_10[0][0]',       \n",
      "                                                                  'dropout[0][0]']                \n",
      "                                                                                                  \n",
      " re_lu_13 (ReLU)                (None, 512, 96)      0           ['add[0][0]']                    \n",
      "                                                                                                  \n",
      " conv1d_13 (Conv1D)             (None, 512, 48)      13824       ['re_lu_13[0][0]']               \n",
      "                                                                                                  \n",
      " batch_normalization_13 (BatchN  (None, 512, 48)     192         ['conv1d_13[0][0]']              \n",
      " ormalization)                                                                                    \n",
      "                                                                                                  \n",
      " re_lu_14 (ReLU)                (None, 512, 48)      0           ['batch_normalization_13[0][0]'] \n",
      "                                                                                                  \n",
      " conv1d_14 (Conv1D)             (None, 512, 96)      4608        ['re_lu_14[0][0]']               \n",
      "                                                                                                  \n",
      " batch_normalization_14 (BatchN  (None, 512, 96)     384         ['conv1d_14[0][0]']              \n",
      " ormalization)                                                                                    \n",
      "                                                                                                  \n",
      " dropout_1 (Dropout)            (None, 512, 96)      0           ['batch_normalization_14[0][0]'] \n",
      "                                                                                                  \n",
      " add_1 (Add)                    (None, 512, 96)      0           ['add[0][0]',                    \n",
      "                                                                  'dropout_1[0][0]']              \n",
      "                                                                                                  \n",
      " re_lu_15 (ReLU)                (None, 512, 96)      0           ['add_1[0][0]']                  \n",
      "                                                                                                  \n",
      " conv1d_15 (Conv1D)             (None, 512, 48)      13824       ['re_lu_15[0][0]']               \n",
      "                                                                                                  \n",
      " batch_normalization_15 (BatchN  (None, 512, 48)     192         ['conv1d_15[0][0]']              \n",
      " ormalization)                                                                                    \n",
      "                                                                                                  \n",
      " re_lu_16 (ReLU)                (None, 512, 48)      0           ['batch_normalization_15[0][0]'] \n",
      "                                                                                                  \n",
      " conv1d_16 (Conv1D)             (None, 512, 96)      4608        ['re_lu_16[0][0]']               \n",
      "                                                                                                  \n",
      " batch_normalization_16 (BatchN  (None, 512, 96)     384         ['conv1d_16[0][0]']              \n",
      " ormalization)                                                                                    \n",
      "                                                                                                  \n",
      " dropout_2 (Dropout)            (None, 512, 96)      0           ['batch_normalization_16[0][0]'] \n",
      "                                                                                                  \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " add_2 (Add)                    (None, 512, 96)      0           ['add_1[0][0]',                  \n",
      "                                                                  'dropout_2[0][0]']              \n",
      "                                                                                                  \n",
      " re_lu_17 (ReLU)                (None, 512, 96)      0           ['add_2[0][0]']                  \n",
      "                                                                                                  \n",
      " conv1d_17 (Conv1D)             (None, 512, 48)      13824       ['re_lu_17[0][0]']               \n",
      "                                                                                                  \n",
      " batch_normalization_17 (BatchN  (None, 512, 48)     192         ['conv1d_17[0][0]']              \n",
      " ormalization)                                                                                    \n",
      "                                                                                                  \n",
      " re_lu_18 (ReLU)                (None, 512, 48)      0           ['batch_normalization_17[0][0]'] \n",
      "                                                                                                  \n",
      " conv1d_18 (Conv1D)             (None, 512, 96)      4608        ['re_lu_18[0][0]']               \n",
      "                                                                                                  \n",
      " batch_normalization_18 (BatchN  (None, 512, 96)     384         ['conv1d_18[0][0]']              \n",
      " ormalization)                                                                                    \n",
      "                                                                                                  \n",
      " dropout_3 (Dropout)            (None, 512, 96)      0           ['batch_normalization_18[0][0]'] \n",
      "                                                                                                  \n",
      " add_3 (Add)                    (None, 512, 96)      0           ['add_2[0][0]',                  \n",
      "                                                                  'dropout_3[0][0]']              \n",
      "                                                                                                  \n",
      " re_lu_19 (ReLU)                (None, 512, 96)      0           ['add_3[0][0]']                  \n",
      "                                                                                                  \n",
      " conv1d_19 (Conv1D)             (None, 512, 48)      13824       ['re_lu_19[0][0]']               \n",
      "                                                                                                  \n",
      " batch_normalization_19 (BatchN  (None, 512, 48)     192         ['conv1d_19[0][0]']              \n",
      " ormalization)                                                                                    \n",
      "                                                                                                  \n",
      " re_lu_20 (ReLU)                (None, 512, 48)      0           ['batch_normalization_19[0][0]'] \n",
      "                                                                                                  \n",
      " conv1d_20 (Conv1D)             (None, 512, 96)      4608        ['re_lu_20[0][0]']               \n",
      "                                                                                                  \n",
      " batch_normalization_20 (BatchN  (None, 512, 96)     384         ['conv1d_20[0][0]']              \n",
      " ormalization)                                                                                    \n",
      "                                                                                                  \n",
      " dropout_4 (Dropout)            (None, 512, 96)      0           ['batch_normalization_20[0][0]'] \n",
      "                                                                                                  \n",
      " add_4 (Add)                    (None, 512, 96)      0           ['add_3[0][0]',                  \n",
      "                                                                  'dropout_4[0][0]']              \n",
      "                                                                                                  \n",
      " re_lu_21 (ReLU)                (None, 512, 96)      0           ['add_4[0][0]']                  \n",
      "                                                                                                  \n",
      " conv1d_21 (Conv1D)             (None, 512, 48)      13824       ['re_lu_21[0][0]']               \n",
      "                                                                                                  \n",
      " batch_normalization_21 (BatchN  (None, 512, 48)     192         ['conv1d_21[0][0]']              \n",
      " ormalization)                                                                                    \n",
      "                                                                                                  \n",
      " re_lu_22 (ReLU)                (None, 512, 48)      0           ['batch_normalization_21[0][0]'] \n",
      "                                                                                                  \n",
      " conv1d_22 (Conv1D)             (None, 512, 96)      4608        ['re_lu_22[0][0]']               \n",
      "                                                                                                  \n",
      " batch_normalization_22 (BatchN  (None, 512, 96)     384         ['conv1d_22[0][0]']              \n",
      " ormalization)                                                                                    \n",
      "                                                                                                  \n",
      " dropout_5 (Dropout)            (None, 512, 96)      0           ['batch_normalization_22[0][0]'] \n",
      "                                                                                                  \n",
      " add_5 (Add)                    (None, 512, 96)      0           ['add_4[0][0]',                  \n",
      "                                                                  'dropout_5[0][0]']              \n",
      "                                                                                                  \n",
      " re_lu_23 (ReLU)                (None, 512, 96)      0           ['add_5[0][0]']                  \n",
      "                                                                                                  \n",
      " conv1d_23 (Conv1D)             (None, 512, 48)      13824       ['re_lu_23[0][0]']               \n",
      "                                                                                                  \n",
      " batch_normalization_23 (BatchN  (None, 512, 48)     192         ['conv1d_23[0][0]']              \n",
      " ormalization)                                                                                    \n",
      "                                                                                                  \n",
      " re_lu_24 (ReLU)                (None, 512, 48)      0           ['batch_normalization_23[0][0]'] \n",
      "                                                                                                  \n",
      " conv1d_24 (Conv1D)             (None, 512, 96)      4608        ['re_lu_24[0][0]']               \n",
      "                                                                                                  \n",
      " batch_normalization_24 (BatchN  (None, 512, 96)     384         ['conv1d_24[0][0]']              \n",
      " ormalization)                                                                                    \n",
      "                                                                                                  \n",
      " dropout_6 (Dropout)            (None, 512, 96)      0           ['batch_normalization_24[0][0]'] \n",
      "                                                                                                  \n",
      " add_6 (Add)                    (None, 512, 96)      0           ['add_5[0][0]',                  \n",
      "                                                                  'dropout_6[0][0]']              \n",
      "                                                                                                  \n",
      " re_lu_25 (ReLU)                (None, 512, 96)      0           ['add_6[0][0]']                  \n",
      "                                                                                                  \n",
      " conv1d_25 (Conv1D)             (None, 512, 48)      13824       ['re_lu_25[0][0]']               \n",
      "                                                                                                  \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " batch_normalization_25 (BatchN  (None, 512, 48)     192         ['conv1d_25[0][0]']              \n",
      " ormalization)                                                                                    \n",
      "                                                                                                  \n",
      " re_lu_26 (ReLU)                (None, 512, 48)      0           ['batch_normalization_25[0][0]'] \n",
      "                                                                                                  \n",
      " conv1d_26 (Conv1D)             (None, 512, 96)      4608        ['re_lu_26[0][0]']               \n",
      "                                                                                                  \n",
      " batch_normalization_26 (BatchN  (None, 512, 96)     384         ['conv1d_26[0][0]']              \n",
      " ormalization)                                                                                    \n",
      "                                                                                                  \n",
      " dropout_7 (Dropout)            (None, 512, 96)      0           ['batch_normalization_26[0][0]'] \n",
      "                                                                                                  \n",
      " add_7 (Add)                    (None, 512, 96)      0           ['add_6[0][0]',                  \n",
      "                                                                  'dropout_7[0][0]']              \n",
      "                                                                                                  \n",
      " re_lu_27 (ReLU)                (None, 512, 96)      0           ['add_7[0][0]']                  \n",
      "                                                                                                  \n",
      " conv1d_27 (Conv1D)             (None, 512, 64)      30720       ['re_lu_27[0][0]']               \n",
      "                                                                                                  \n",
      " batch_normalization_27 (BatchN  (None, 512, 64)     256         ['conv1d_27[0][0]']              \n",
      " ormalization)                                                                                    \n",
      "                                                                                                  \n",
      " re_lu_28 (ReLU)                (None, 512, 64)      0           ['batch_normalization_27[0][0]'] \n",
      "                                                                                                  \n",
      " one_to_two (OneToTwo)          (None, 512, 512, 64  0           ['re_lu_28[0][0]']               \n",
      "                                )                                                                 \n",
      "                                                                                                  \n",
      " concat_dist2d (ConcatDist2D)   (None, 512, 512, 65  0           ['one_to_two[0][0]']             \n",
      "                                )                                                                 \n",
      "                                                                                                  \n",
      " re_lu_29 (ReLU)                (None, 512, 512, 65  0           ['concat_dist2d[0][0]']          \n",
      "                                )                                                                 \n",
      "                                                                                                  \n",
      " conv2d (Conv2D)                (None, 512, 512, 48  28080       ['re_lu_29[0][0]']               \n",
      "                                )                                                                 \n",
      "                                                                                                  \n",
      " batch_normalization_28 (BatchN  (None, 512, 512, 48  192        ['conv2d[0][0]']                 \n",
      " ormalization)                  )                                                                 \n",
      "                                                                                                  \n",
      " symmetrize2d (Symmetrize2D)    (None, 512, 512, 48  0           ['batch_normalization_28[0][0]'] \n",
      "                                )                                                                 \n",
      "                                                                                                  \n",
      " re_lu_30 (ReLU)                (None, 512, 512, 48  0           ['symmetrize2d[0][0]']           \n",
      "                                )                                                                 \n",
      "                                                                                                  \n",
      " conv2d_1 (Conv2D)              (None, 512, 512, 24  10368       ['re_lu_30[0][0]']               \n",
      "                                )                                                                 \n",
      "                                                                                                  \n",
      " batch_normalization_29 (BatchN  (None, 512, 512, 24  96         ['conv2d_1[0][0]']               \n",
      " ormalization)                  )                                                                 \n",
      "                                                                                                  \n",
      " re_lu_31 (ReLU)                (None, 512, 512, 24  0           ['batch_normalization_29[0][0]'] \n",
      "                                )                                                                 \n",
      "                                                                                                  \n",
      " conv2d_2 (Conv2D)              (None, 512, 512, 48  1152        ['re_lu_31[0][0]']               \n",
      "                                )                                                                 \n",
      "                                                                                                  \n",
      " batch_normalization_30 (BatchN  (None, 512, 512, 48  192        ['conv2d_2[0][0]']               \n",
      " ormalization)                  )                                                                 \n",
      "                                                                                                  \n",
      " dropout_8 (Dropout)            (None, 512, 512, 48  0           ['batch_normalization_30[0][0]'] \n",
      "                                )                                                                 \n",
      "                                                                                                  \n",
      " add_8 (Add)                    (None, 512, 512, 48  0           ['symmetrize2d[0][0]',           \n",
      "                                )                                 'dropout_8[0][0]']              \n",
      "                                                                                                  \n",
      " symmetrize2d_1 (Symmetrize2D)  (None, 512, 512, 48  0           ['add_8[0][0]']                  \n",
      "                                )                                                                 \n",
      "                                                                                                  \n",
      " re_lu_32 (ReLU)                (None, 512, 512, 48  0           ['symmetrize2d_1[0][0]']         \n",
      "                                )                                                                 \n",
      "                                                                                                  \n",
      " conv2d_3 (Conv2D)              (None, 512, 512, 24  10368       ['re_lu_32[0][0]']               \n",
      "                                )                                                                 \n",
      "                                                                                                  \n",
      " batch_normalization_31 (BatchN  (None, 512, 512, 24  96         ['conv2d_3[0][0]']               \n",
      " ormalization)                  )                                                                 \n",
      "                                                                                                  \n",
      " re_lu_33 (ReLU)                (None, 512, 512, 24  0           ['batch_normalization_31[0][0]'] \n",
      "                                )                                                                 \n",
      "                                                                                                  \n",
      " conv2d_4 (Conv2D)              (None, 512, 512, 48  1152        ['re_lu_33[0][0]']               \n",
      "                                )                                                                 \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                                                                                  \n",
      " batch_normalization_32 (BatchN  (None, 512, 512, 48  192        ['conv2d_4[0][0]']               \n",
      " ormalization)                  )                                                                 \n",
      "                                                                                                  \n",
      " dropout_9 (Dropout)            (None, 512, 512, 48  0           ['batch_normalization_32[0][0]'] \n",
      "                                )                                                                 \n",
      "                                                                                                  \n",
      " add_9 (Add)                    (None, 512, 512, 48  0           ['symmetrize2d_1[0][0]',         \n",
      "                                )                                 'dropout_9[0][0]']              \n",
      "                                                                                                  \n",
      " symmetrize2d_2 (Symmetrize2D)  (None, 512, 512, 48  0           ['add_9[0][0]']                  \n",
      "                                )                                                                 \n",
      "                                                                                                  \n",
      " re_lu_34 (ReLU)                (None, 512, 512, 48  0           ['symmetrize2d_2[0][0]']         \n",
      "                                )                                                                 \n",
      "                                                                                                  \n",
      " conv2d_5 (Conv2D)              (None, 512, 512, 24  10368       ['re_lu_34[0][0]']               \n",
      "                                )                                                                 \n",
      "                                                                                                  \n",
      " batch_normalization_33 (BatchN  (None, 512, 512, 24  96         ['conv2d_5[0][0]']               \n",
      " ormalization)                  )                                                                 \n",
      "                                                                                                  \n",
      " re_lu_35 (ReLU)                (None, 512, 512, 24  0           ['batch_normalization_33[0][0]'] \n",
      "                                )                                                                 \n",
      "                                                                                                  \n",
      " conv2d_6 (Conv2D)              (None, 512, 512, 48  1152        ['re_lu_35[0][0]']               \n",
      "                                )                                                                 \n",
      "                                                                                                  \n",
      " batch_normalization_34 (BatchN  (None, 512, 512, 48  192        ['conv2d_6[0][0]']               \n",
      " ormalization)                  )                                                                 \n",
      "                                                                                                  \n",
      " dropout_10 (Dropout)           (None, 512, 512, 48  0           ['batch_normalization_34[0][0]'] \n",
      "                                )                                                                 \n",
      "                                                                                                  \n",
      " add_10 (Add)                   (None, 512, 512, 48  0           ['symmetrize2d_2[0][0]',         \n",
      "                                )                                 'dropout_10[0][0]']             \n",
      "                                                                                                  \n",
      " symmetrize2d_3 (Symmetrize2D)  (None, 512, 512, 48  0           ['add_10[0][0]']                 \n",
      "                                )                                                                 \n",
      "                                                                                                  \n",
      " re_lu_36 (ReLU)                (None, 512, 512, 48  0           ['symmetrize2d_3[0][0]']         \n",
      "                                )                                                                 \n",
      "                                                                                                  \n",
      " conv2d_7 (Conv2D)              (None, 512, 512, 24  10368       ['re_lu_36[0][0]']               \n",
      "                                )                                                                 \n",
      "                                                                                                  \n",
      " batch_normalization_35 (BatchN  (None, 512, 512, 24  96         ['conv2d_7[0][0]']               \n",
      " ormalization)                  )                                                                 \n",
      "                                                                                                  \n",
      " re_lu_37 (ReLU)                (None, 512, 512, 24  0           ['batch_normalization_35[0][0]'] \n",
      "                                )                                                                 \n",
      "                                                                                                  \n",
      " conv2d_8 (Conv2D)              (None, 512, 512, 48  1152        ['re_lu_37[0][0]']               \n",
      "                                )                                                                 \n",
      "                                                                                                  \n",
      " batch_normalization_36 (BatchN  (None, 512, 512, 48  192        ['conv2d_8[0][0]']               \n",
      " ormalization)                  )                                                                 \n",
      "                                                                                                  \n",
      " dropout_11 (Dropout)           (None, 512, 512, 48  0           ['batch_normalization_36[0][0]'] \n",
      "                                )                                                                 \n",
      "                                                                                                  \n",
      " add_11 (Add)                   (None, 512, 512, 48  0           ['symmetrize2d_3[0][0]',         \n",
      "                                )                                 'dropout_11[0][0]']             \n",
      "                                                                                                  \n",
      " symmetrize2d_4 (Symmetrize2D)  (None, 512, 512, 48  0           ['add_11[0][0]']                 \n",
      "                                )                                                                 \n",
      "                                                                                                  \n",
      " re_lu_38 (ReLU)                (None, 512, 512, 48  0           ['symmetrize2d_4[0][0]']         \n",
      "                                )                                                                 \n",
      "                                                                                                  \n",
      " conv2d_9 (Conv2D)              (None, 512, 512, 24  10368       ['re_lu_38[0][0]']               \n",
      "                                )                                                                 \n",
      "                                                                                                  \n",
      " batch_normalization_37 (BatchN  (None, 512, 512, 24  96         ['conv2d_9[0][0]']               \n",
      " ormalization)                  )                                                                 \n",
      "                                                                                                  \n",
      " re_lu_39 (ReLU)                (None, 512, 512, 24  0           ['batch_normalization_37[0][0]'] \n",
      "                                )                                                                 \n",
      "                                                                                                  \n",
      " conv2d_10 (Conv2D)             (None, 512, 512, 48  1152        ['re_lu_39[0][0]']               \n",
      "                                )                                                                 \n",
      "                                                                                                  \n",
      " batch_normalization_38 (BatchN  (None, 512, 512, 48  192        ['conv2d_10[0][0]']              \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " ormalization)                  )                                                                 \n",
      "                                                                                                  \n",
      " dropout_12 (Dropout)           (None, 512, 512, 48  0           ['batch_normalization_38[0][0]'] \n",
      "                                )                                                                 \n",
      "                                                                                                  \n",
      " add_12 (Add)                   (None, 512, 512, 48  0           ['symmetrize2d_4[0][0]',         \n",
      "                                )                                 'dropout_12[0][0]']             \n",
      "                                                                                                  \n",
      " symmetrize2d_5 (Symmetrize2D)  (None, 512, 512, 48  0           ['add_12[0][0]']                 \n",
      "                                )                                                                 \n",
      "                                                                                                  \n",
      " re_lu_40 (ReLU)                (None, 512, 512, 48  0           ['symmetrize2d_5[0][0]']         \n",
      "                                )                                                                 \n",
      "                                                                                                  \n",
      " conv2d_11 (Conv2D)             (None, 512, 512, 24  10368       ['re_lu_40[0][0]']               \n",
      "                                )                                                                 \n",
      "                                                                                                  \n",
      " batch_normalization_39 (BatchN  (None, 512, 512, 24  96         ['conv2d_11[0][0]']              \n",
      " ormalization)                  )                                                                 \n",
      "                                                                                                  \n",
      " re_lu_41 (ReLU)                (None, 512, 512, 24  0           ['batch_normalization_39[0][0]'] \n",
      "                                )                                                                 \n",
      "                                                                                                  \n",
      " conv2d_12 (Conv2D)             (None, 512, 512, 48  1152        ['re_lu_41[0][0]']               \n",
      "                                )                                                                 \n",
      "                                                                                                  \n",
      " batch_normalization_40 (BatchN  (None, 512, 512, 48  192        ['conv2d_12[0][0]']              \n",
      " ormalization)                  )                                                                 \n",
      "                                                                                                  \n",
      " dropout_13 (Dropout)           (None, 512, 512, 48  0           ['batch_normalization_40[0][0]'] \n",
      "                                )                                                                 \n",
      "                                                                                                  \n",
      " add_13 (Add)                   (None, 512, 512, 48  0           ['symmetrize2d_5[0][0]',         \n",
      "                                )                                 'dropout_13[0][0]']             \n",
      "                                                                                                  \n",
      " symmetrize2d_6 (Symmetrize2D)  (None, 512, 512, 48  0           ['add_13[0][0]']                 \n",
      "                                )                                                                 \n",
      "                                                                                                  \n",
      " cropping2d (Cropping2D)        (None, 448, 448, 48  0           ['symmetrize2d_6[0][0]']         \n",
      "                                )                                                                 \n",
      "                                                                                                  \n",
      " upper_tri (UpperTri)           (None, 99681, 48)    0           ['cropping2d[0][0]']             \n",
      "                                                                                                  \n",
      " dense (Dense)                  (None, 99681, 5)     245         ['upper_tri[0][0]']              \n",
      "                                                                                                  \n",
      " switch_reverse_triu (SwitchRev  (None, 99681, 5)    0           ['dense[0][0]',                  \n",
      " erseTriu)                                                        'stochastic_reverse_complement[0\n",
      "                                                                 ][1]']                           \n",
      "                                                                                                  \n",
      "==================================================================================================\n",
      "Total params: 751,653\n",
      "Trainable params: 746,149\n",
      "Non-trainable params: 5,504\n",
      "__________________________________________________________________________________________________\n",
      "None\n",
      "model_strides [2048]\n",
      "target_lengths [99681]\n",
      "target_crops [-49585]\n",
      "model successfully loaded\n"
     ]
    }
   ],
   "source": [
    "import akita_utils_forplotting as utils\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import sys, os, psutil, time\n",
    "import matplotlib.pyplot as plt\n",
    "import pysam\n",
    "    \n",
    "from skimage.transform import resize\n",
    "\n",
    "\n",
    "#from optparse import OptionParser\n",
    "\n",
    "\n",
    "nt = ['A', 'T', 'C', 'G']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "3b85a63c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import h5py\n",
    "import pyBigWig\n",
    "import intervaltree\n",
    "import basenji\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# hic imports\n",
    "import cooler\n",
    "from cooltools.lib.numutils import observed_over_expected, adaptive_coarsegrain\n",
    "from cooltools.lib.numutils import interpolate_bad_singletons, set_diag, interp_nan\n",
    "from astropy.convolution import Gaussian2DKernel\n",
    "from astropy.convolution import convolve\n",
    "\n",
    "import collections "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "29609b85",
   "metadata": {},
   "outputs": [],
   "source": [
    "import cooltools.lib.plotting\n",
    "from matplotlib.colors import LogNorm\n",
    "from mpl_toolkits.axes_grid1 import make_axes_locatable\n",
    "from matplotlib.ticker import EngFormatter\n",
    "import seaborn as sns\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "8c25870a",
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy.ndimage import gaussian_filter, gaussian_filter1d, uniform_filter\n",
    "from astropy.convolution import Gaussian2DKernel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "b573b5bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "#REMEMBER!!!!!!! +1MB SEQUENCE GOES IN, BUT MAKES PREDICTION FOR 917504BP\n",
    "#1MB CORRESPONDS TO 512, 2048 BP BINS THAT GETS CUT DOWN TO 448 BINS "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "b0626ed2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>start_gene</th>\n",
       "      <th>end_gene</th>\n",
       "      <th>Gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>DDX11L1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14404</td>\n",
       "      <td>29570</td>\n",
       "      <td>WASH7P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>17369</td>\n",
       "      <td>17436</td>\n",
       "      <td>MIR6859-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>29554</td>\n",
       "      <td>31109</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>30366</td>\n",
       "      <td>30503</td>\n",
       "      <td>MIR1302-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60567</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57184101</td>\n",
       "      <td>57197337</td>\n",
       "      <td>IL9R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60568</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57190738</td>\n",
       "      <td>57208756</td>\n",
       "      <td>AJ271736.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60569</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57201143</td>\n",
       "      <td>57203357</td>\n",
       "      <td>WASIR1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60570</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57207346</td>\n",
       "      <td>57212230</td>\n",
       "      <td>WASH6P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60571</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57212184</td>\n",
       "      <td>57214397</td>\n",
       "      <td>DDX11L16</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>60572 rows  4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      CHROM  start_gene  end_gene         Gene\n",
       "0      chr1       11869     14409      DDX11L1\n",
       "1      chr1       14404     29570       WASH7P\n",
       "2      chr1       17369     17436    MIR6859-1\n",
       "3      chr1       29554     31109  MIR1302-2HG\n",
       "4      chr1       30366     30503    MIR1302-2\n",
       "...     ...         ...       ...          ...\n",
       "60567  chrY    57184101  57197337         IL9R\n",
       "60568  chrY    57190738  57208756   AJ271736.1\n",
       "60569  chrY    57201143  57203357       WASIR1\n",
       "60570  chrY    57207346  57212230       WASH6P\n",
       "60571  chrY    57212184  57214397     DDX11L16\n",
       "\n",
       "[60572 rows x 4 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#TOTAL AKITA INPUT LENGTH \n",
    "\n",
    "akita_seq_len=1048576\n",
    "\n",
    "\n",
    "#HG38 REF FA\n",
    "hg38_file = '/pollard/home/shzhang/akita/run_akita/refs/hg38.fa'\n",
    "fasta_hg38 = pysam.Fastafile(hg38_file)\n",
    "\n",
    "#GENCODE\n",
    "gene_annot = pd.read_csv('/pollard/data/genetics/GENCODE/release32/gencode.v32.genes_only.bed', sep = '\\t', names = ['CHROM', 'start_gene', 'end_gene', 'Gene'])\n",
    "gene_annot[gene_annot.CHROM != 'chrM']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "b115d2d4",
   "metadata": {},
   "outputs": [],
   "source": [
    "#to do normalization\n",
    "#functions from ohther files \n",
    "ModelSeq = collections.namedtuple('ModelSeq', ['chr', 'start', 'end', 'label'])\n",
    "\n",
    "def read_blacklist(blacklist_bed, black_buffer=20):\n",
    "    \"\"\"Construct interval trees of blacklist\n",
    "     regions for each chromosome.\"\"\"\n",
    "    black_chr_trees = {}\n",
    "\n",
    "    if blacklist_bed is not None and os.path.isfile(blacklist_bed):\n",
    "        for line in open(blacklist_bed):\n",
    "            a = line.split()\n",
    "            chrm = a[0]\n",
    "            start = max(0, int(a[1]) - black_buffer)\n",
    "            end = int(a[2]) + black_buffer\n",
    "\n",
    "        if chrm not in black_chr_trees:\n",
    "            black_chr_trees[chrm] = intervaltree.IntervalTree()\n",
    "\n",
    "        black_chr_trees[chrm][start:end] = True\n",
    "\n",
    "    return black_chr_trees\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "2ca1cfcb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "524288"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "utils.half_patch_size"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "645677f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def format_ticks(ax, x=True, y=True, rotate=True):\n",
    "    if y:\n",
    "        ax.yaxis.set_major_formatter(bp_formatter)\n",
    "    if x:\n",
    "        ax.xaxis.set_major_formatter(bp_formatter)\n",
    "        ax.xaxis.tick_bottom()\n",
    "    if rotate:\n",
    "        ax.tick_params(axis='x',rotation=45)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "2ebfcef6",
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_fasta(fa):\n",
    "    #read in fasta and return sequence only \n",
    "    seq=''\n",
    "    with open(fa, \"r\") as fa_file:\n",
    "        for line in fa_file:\n",
    "            if not (line.startswith('>')):\n",
    "                seq += line.strip()\n",
    "    fa_file.close()                \n",
    "    return(seq)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "6ba4aa94",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['SMF-12-23_walk_9_MDM2_containing.fa.mcool',\n",
       " 'SMF-12-23_walk_8_MDM2_containing.fa.mcool',\n",
       " 'SMF-12-12_walk_3_KRAS_containing.fa.mcool',\n",
       " 'SMF-12-23_walk_9_GLI1_containing.fa.mcool',\n",
       " 'SMF-12-12_walk_4_KRAS_containing.fa.mcool',\n",
       " 'SMF-12-12_walk_5_KRAS_containing.fa.mcool',\n",
       " 'SMF-12-17_walk_7_MDM2_containing.fa.mcool',\n",
       " 'SMF-12-23_walk_11_GLI1_containing.fa.mcool',\n",
       " 'SMF-12-12_walk_4_GLI1_containing.fa.mcool',\n",
       " 'SMF-12-17_walk_4_GLI1_containing.fa.mcool',\n",
       " 'SMF-12-23_walk_11_MDM2_containing.fa.mcool',\n",
       " 'SMF-12-17_walk_9_MDM2_containing.fa.mcool',\n",
       " 'SMF-12-23_walk_10_MDM2_containing.fa.mcool',\n",
       " 'SMF-12-17_walk_8_MDM2_containing.fa.mcool',\n",
       " 'SMF-12-17_walk_5_GLI1_containing.fa.mcool',\n",
       " 'SMF-12-17_walk_3_GLI1_containing.fa.mcool']"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#data directories\n",
    "date = '20231110_v1'\n",
    "d='/pollard/home/shzhang/akita_variant_scoring/akdemir_collab/'\n",
    "data_dir = d + 'data/'\n",
    "fig_dir = f'{d}figs/{date}/'\n",
    "\n",
    "\n",
    "#two folders are colab_fasta and new_fasta\n",
    "list_mcool=os.listdir(data_dir + 'mcool/')\n",
    "list_mcool"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "50dc077a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#make akita predictions for each walk and plot \n",
    "def akita_pred(mcool):\n",
    "    i=mcool.replace('.fa.mcool', '.fa')\n",
    "    print(i)\n",
    "    bed_name=i[:-3] +'.bed'\n",
    "    name_parts = i.rsplit('_', 2)\n",
    "\n",
    "    gencode_coord=gene_annot[gene_annot.Gene == name_parts[1]]    \n",
    "    gene_len=abs(gencode_coord.end_gene-gencode_coord.start_gene)\n",
    "    gene_mid_coord=np.floor(gencode_coord.start_gene + (gene_len/2))\n",
    "    #print('gene mid coord', gene_mid_coord)\n",
    "    #print('gencode coord', gencode_coord)\n",
    "    \n",
    "    #read in bed file and figure out where to center the file \n",
    "    bed=pd.read_csv(data_dir+ 'new_fasta/walks.bed/' + bed_name, sep='\\t', header=None)\n",
    "    #rename columns\n",
    "    bed.columns = ['chr', 'start', 'end', 'strand', '4', '5']\n",
    "    bed['chr'] = bed['chr'].astype('str') \n",
    "    #get spans of each walk \n",
    "    bed['span']=abs(bed['end']-bed['start']) \n",
    "    \n",
    "    #print('bed span', np.sum(bed['span']))\n",
    "\n",
    "\n",
    "    #get the row where gene_mid_coord \n",
    "    #print(gencode_coord['CHROM'])\n",
    "    condition = ((bed['chr']==gencode_coord.CHROM.values[0])\n",
    "                & (bed['start'] <= int(gene_mid_coord)) & (int(gene_mid_coord) <= bed['end']) ) \n",
    "    #in cases where the middle of the gene shows up more than once, just takes first instance \n",
    "    row_midgene = (condition.index[condition].tolist())[0]\n",
    "    #print('row_midgene index', row_midgene)\n",
    "                \n",
    "\n",
    "    #get span for the entire walk up to the middle of the gene \n",
    "    bed_row=bed.iloc[row_midgene, :]\n",
    "    #print('bed_Row', bed_row)\n",
    "    #print('gene row span', np.sum(bed_row['span']))\n",
    "    #print('gene_mid_Coord', gene_mid_coord)\n",
    "    if row_midgene>0:\n",
    "        bed_begin=bed.iloc[0:row_midgene, :]\n",
    "        #print('bed_begin', bed_begin)\n",
    "        first_half_span=(np.sum(bed_begin['span'])) + int(gene_mid_coord-bed_row['start'])\n",
    "        bed_before=pd.concat([bed_begin, bed_row])\n",
    "\n",
    "    else:\n",
    "        first_half_span=int(gene_mid_coord-bed_row['start'])\n",
    "        bed_before=bed_row\n",
    "    #print('bed_before', bed_row)\n",
    "            \n",
    "    #then, get span for entire walk from middle of gene to end \n",
    "    if row_midgene<(len(bed)-1):\n",
    "        bed_end=bed.iloc[(row_midgene+1):len(bed), :]\n",
    "        #print('bed_end', bed_end)\n",
    "        second_half_span=(np.sum(bed_end['span'])) + int(bed_row['end']-gene_mid_coord)\n",
    "    else:\n",
    "        second_half_span=int(bed_row['end']-gene_mid_coord)\n",
    "\n",
    "        \n",
    "    if 1==1:   \n",
    "        #read in fasta file \n",
    "        seq=read_fasta(data_dir + 'new_fasta/walks.fasta/' + i)\n",
    "\n",
    "     \n",
    "    \n",
    "        #define ALT from fasta sequences given to us       \n",
    "        #check length of walk start to middle of gene, must be at lesat half of akita_seq_len\n",
    "        mid_seq_idx=0\n",
    "        gene_bin_start=0\n",
    "        gene_bin_end=0\n",
    "        seq_start=0\n",
    "        seq_end=0\n",
    "        gene_half=round(gene_len/2)\n",
    "        \n",
    "        #at least right now, no sequences need to be padded on both left and right so this works.\n",
    "        #in future, if that's the case, we will need to change this\n",
    "        if first_half_span<=(utils.half_patch_size):\n",
    "            print('starting from left most sequence')\n",
    "            \n",
    "            seq_start=0\n",
    "            seq_end=utils.MB\n",
    "            ALT_seq=seq[0:utils.MB]\n",
    "            \n",
    "            gene_bin_start=utils.get_bin(first_half_span-gene_half)\n",
    "            gene_bin_end=utils.get_bin(first_half_span+gene_half)\n",
    "          \n",
    "        \n",
    "            #get REF POS \n",
    "            #POS = int(gene_mid_coord-utils.half_patch_size) + (utils.half_patch_size-first_half_span)\n",
    "            POS = int(gene_mid_coord-first_half_span)\n",
    "\n",
    "        \n",
    "           \n",
    "        elif second_half_span<=(utils.half_patch_size):          \n",
    "            print('ending at rightmost sequence ')\n",
    "            \n",
    "            seq_start=(len(seq)-utils.MB)\n",
    "            seq_end=len(seq)       \n",
    "            ALT_seq=seq[seq_start:seq_end]\n",
    "            \n",
    "            gene_bin_start=utils.get_bin((utils.MB-second_half_span-gene_half))\n",
    "            gene_bin_end=utils.get_bin((utils.MB-second_half_span+gene_half))\n",
    "\n",
    "                    \n",
    "            #get REF POS \n",
    "            POS = int(gene_mid_coord-utils.half_patch_size) - (utils.half_patch_size-second_half_span)\n",
    "\n",
    "            \n",
    "        else:  \n",
    "            seq_start=int(first_half_span-(utils.half_patch_size))\n",
    "            seq_end=int(seq_start+utils.MB)  \n",
    "            print(seq_start)\n",
    "            print(seq_end)\n",
    "            print(len(seq))\n",
    "            ALT_seq=seq[seq_start:seq_end]\n",
    "\n",
    "            gene_bin_start=utils.get_bin(utils.half_patch_size - round(gene_len/2))\n",
    "            gene_bin_end=utils.get_bin(utils.half_patch_size + round(gene_len/2))\n",
    "            \n",
    "                    \n",
    "            #get REF POS \n",
    "            POS = int(gene_mid_coord-utils.half_patch_size)\n",
    "            \n",
    "            \n",
    "            \n",
    "        #get REF                   \n",
    "        #define start and end coordinates of reference matrices\n",
    "        #currently, just centered around the middle of the gene of interest        \n",
    "        CHR = gencode_coord.CHROM.values[0]\n",
    "        END= int(POS+utils.seq_length)\n",
    "        REF_seq = fasta_hg38.fetch(CHR, POS, END).upper()\n",
    "            \n",
    "        ####################\n",
    "        \n",
    "        assert len(REF_seq) == utils.seq_length, 'REF_seq is not right length'\n",
    "        assert len(ALT_seq) == utils.seq_length, 'ALT_seq is not right length'\n",
    "        \n",
    "#         print('gene bin start', gene_bin_start)\n",
    "#         print('gene bin end', gene_bin_end)\n",
    "        \n",
    "#         print('seq start', seq_start)\n",
    "#         print('seq end', seq_end)\n",
    "#         print('seq len', len(seq))\n",
    "\n",
    "    \n",
    "        #this is where akita is actually run, output vector size is (99681,)\n",
    "        REF_vector, ALT_vector = utils.vector_from_seq(REF_seq), utils.vector_from_seq(ALT_seq)\n",
    "        REF_mat, ALT_mat = utils.mat_from_vector(REF_vector), utils.mat_from_vector(ALT_vector)\n",
    "\n",
    "\n",
    "        ALT_label=i[:-3]\n",
    "        REF_label='{}:{}-{}'.format(CHR, POS, END)\n",
    "        return(REF_mat, ALT_mat, REF_label, ALT_label, \n",
    "               gene_bin_start, gene_bin_end, gene_len, \n",
    "               CHR, seq_start, seq_end)\n",
    "\n",
    "\n",
    "#         plot_complex_pred(ALT_mat, REF_mat, ALT_label, REF_label, gene_len)\n",
    "            \n",
    "        \n",
    "    else:\n",
    "        print('error!!!! in file ', i )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fab47d43",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8cec32d8",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "b5d5ab3e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#below is to normalize original hi-c maps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "bafa8e76",
   "metadata": {},
   "outputs": [],
   "source": [
    "def normalize_hic(file_path, seqs_info, resolution, clip, crop_bp, kernel_stddev, pool_width, \n",
    "                  no_log, crop, as_obsexp, global_obsexp, diagonal_offset, norm_type, output_file):\n",
    "    genome_hic_file = f'{file_path}::resolutions/{resolution}'\n",
    "    print(genome_hic_file)\n",
    "    #bed file of sequences that we want from the initial hic file \n",
    "\n",
    "\n",
    "    #path to where output file should be stored\n",
    "    seqs_hic_file = output_file\n",
    "    \n",
    "    ####################\n",
    "    # read model sequences\n",
    "    model_seqs = []\n",
    "    model_seqs.append(ModelSeq(seqs_info[0],int(seqs_info[1]),int(seqs_info[2]),None))\n",
    "\n",
    "    # compute dimensions\n",
    "    num_seqs = len(model_seqs)\n",
    "    seq_len_nt = model_seqs[0].end - model_seqs[0].start\n",
    "    seq_len_pool = seq_len_nt // pool_width\n",
    "\n",
    "    if crop_bp == 0:\n",
    "        seq_len_crop = seq_len_pool\n",
    "    else:\n",
    "        crop_start = crop_bp // pool_width\n",
    "        crop_end = seq_len_pool - crop_start\n",
    "        seq_len_crop = seq_len_pool - 2*crop_start\n",
    "\n",
    "    # compute upper triangular indexes\n",
    "    triu_tup = np.triu_indices(seq_len_crop, diagonal_offset)\n",
    "    seq_len_nodiag = seq_len_crop - diagonal_offset\n",
    "    seq_len_hic = seq_len_nodiag*(seq_len_nodiag + 1) // 2\n",
    "\n",
    "\n",
    "    if kernel_stddev > 0:\n",
    "        # initialize Gaussian kernel\n",
    "        kernel = Gaussian2DKernel(x_stddev=kernel_stddev)\n",
    "        #kernel = Gaussian2DKernel(x_stddev=kernel_stddev, x_size=kernel_size)\n",
    "    else:\n",
    "        kernel = None\n",
    "        \n",
    "        \n",
    "    # open genome coverage file\n",
    "    genome_hic_cool = cooler.Cooler(genome_hic_file)\n",
    "    #check for \"chr\" prefix\n",
    "    chr_pre = 'chr1' in genome_hic_cool.chromnames\n",
    "    # assert that resolution matches\n",
    "    assert(pool_width == genome_hic_cool.info['bin-size'])\n",
    "    \n",
    "    \n",
    "        # initialize sequences coverage file\n",
    "    #seqs_hic_open = h5py.File(seqs_hic_file, 'w')\n",
    "    #seqs_hic_open.create_dataset('targets', shape=(num_seqs, seq_len_hic), dtype='float16')\n",
    "    targets=np.zeros((num_seqs, seq_len_hic))\n",
    "\n",
    "    # for each model sequence\n",
    "    for si in range(num_seqs):\n",
    "        mseq = model_seqs[si]\n",
    "        print(mseq)\n",
    "        if 1==1:\n",
    "          # pull hic values\n",
    "            if chr_pre:\n",
    "                mseq_str = '%s:%d-%d' % (mseq.chr[3:], mseq.start, mseq.end)\n",
    "            else:\n",
    "                mseq_str = '%s:%d-%d' % (mseq.chr, mseq.start, mseq.end)\n",
    "                #print('mseq_str:', mseq_str)\n",
    "\n",
    "            seq_hic_raw = genome_hic_cool.matrix(balance=norm_type).fetch(mseq_str)\n",
    "            seq_hic_nan = np.isnan(seq_hic_raw)\n",
    "            num_filtered_bins = np.sum(np.sum(seq_hic_nan,axis=0) == len(seq_hic_nan))\n",
    "\n",
    "\n",
    "            if num_filtered_bins > (.5*len(seq_hic_nan)):\n",
    "                print(\"WARNING: %s >50%% bins filtered, check:  %s. \" % (genome_hic_file, mseq_str))\n",
    "\n",
    "\n",
    "          # clip first diagonals and high values\n",
    "            clipval = np.nanmedian(np.diag(seq_hic_raw,diagonal_offset))\n",
    "            for i in range(-diagonal_offset+1,diagonal_offset):\n",
    "                set_diag(seq_hic_raw, clipval, i)\n",
    "            seq_hic_raw = np.clip(seq_hic_raw, 0, clipval)\n",
    "            seq_hic_raw[seq_hic_nan] = np.nan\n",
    "\n",
    "          # adaptively coarsegrain based on raw counts\n",
    "            seq_hic_smoothed = adaptive_coarsegrain(\n",
    "                seq_hic_raw,\n",
    "                #use KR in this case \n",
    "                genome_hic_cool.matrix(balance=norm_type).fetch(mseq_str),\n",
    "                cutoff=2, max_levels=8)\n",
    "            seq_hic_nan = np.isnan(seq_hic_smoothed)\n",
    "          #todo: pass an option to add a certain pseudocount value, or the minimum nonzero value\n",
    "\n",
    "            #THIS REQUIRES SOME PREKNOWN EXPECTED HIC COUNTS FILE WHICH WE DON'T HAVE \n",
    "            if as_obsexp:\n",
    "                # compute obs/exp        \n",
    "                if global_obsexp: # compute global obs/exp\n",
    "                    exp_chr = genome_hic_expected.iloc[ genome_hic_expected['chrom'].values ==mseq.chr][0:seq_len_pool]\n",
    "                    if len(exp_chr)==0: \n",
    "                          raise ValueError('no expected values found for chr:'+mseq.chr)\n",
    "                    exp_map= np.zeros((seq_len_pool,seq_len_pool))\n",
    "                    for i in range(seq_len_pool):\n",
    "                        set_diag(exp_map,exp_chr['balanced.avg'].values[i],i)\n",
    "                        set_diag(exp_map,exp_chr['balanced.avg'].values[i],-i)\n",
    "                    seq_hic_obsexp = seq_hic_smoothed / exp_map\n",
    "                    for i in range(-diagonal_offset+1, diagonal_offset): set_diag(seq_hic_obsexp,1.0,i)\n",
    "                    seq_hic_obsexp[seq_hic_nan] = np.nan          \n",
    "\n",
    "                else: # compute local obs/exp\n",
    "                    seq_hic_obsexp = observed_over_expected(seq_hic_smoothed, ~seq_hic_nan)[0]\n",
    "\n",
    "                # TAKE NATURAL LOG \n",
    "                #AND CLIP TO (-2,2)\n",
    "                #AND ALSO INTERPOLATE \n",
    "                if no_log==False:\n",
    "                    seq_hic_obsexp = np.log(seq_hic_obsexp)\n",
    "                    if clip is not None:\n",
    "                        seq_hic_obsexp = np.clip(seq_hic_obsexp, -clip, clip)\n",
    "                    seq_hic_obsexp = interp_nan(seq_hic_obsexp)\n",
    "                    for i in range(-diagonal_offset+1, diagonal_offset): set_diag(seq_hic_obsexp, 0,i)\n",
    "\n",
    "                else:\n",
    "                    if clip is not None:\n",
    "                        seq_hic_obsexp = np.clip(seq_hic_obsexp, 0, clip)\n",
    "                    seq_hic_obsexp = interp_nan(seq_hic_obsexp)\n",
    "                    for i in range(-diagonal_offset+1, diagonal_offset): set_diag(seq_hic_obsexp, 1,i)\n",
    "\n",
    "                # apply kernel\n",
    "                if kernel is not None:\n",
    "                    seq_hic = convolve(seq_hic_obsexp, kernel)\n",
    "                else:\n",
    "                    seq_hic = seq_hic_obsexp\n",
    "\n",
    "            else:\n",
    "                # interpolate all missing bins\n",
    "                seq_hic_interpolated = interp_nan(seq_hic_smoothed)\n",
    "\n",
    "                # rescale, reclip\n",
    "                seq_hic = 100000*seq_hic_interpolated\n",
    "                clipval = np.nanmedian(np.diag(seq_hic,diagonal_offset))\n",
    "                for i in range(-diagonal_offset+1, diagonal_offset):\n",
    "                    set_diag(seq_hic,clipval,i)\n",
    "                seq_hic = np.clip(seq_hic, 0, clipval)\n",
    "\n",
    "                #extra smoothing. todo pass kernel specs\n",
    "                if kernel is not None:\n",
    "                    seq_hic = convolve(seq_hic, kernel)\n",
    "        else:\n",
    "        #except ValueError:\n",
    "            print(\"WARNING: %s doesn't see %s. Setting to all zeros.\" % (genome_hic_file, mseq_str))\n",
    "            seq_hic = np.zeros((seq_len_pool,seq_len_pool), dtype='float16')\n",
    "\n",
    "        # crop\n",
    "        if crop_bp > 0:\n",
    "            seq_hic = seq_hic[crop_start:crop_end,:]\n",
    "            seq_hic = seq_hic[:,crop_start:crop_end]\n",
    "\n",
    "        # unroll upper triangular\n",
    "        seq_hic = seq_hic[triu_tup]\n",
    "        print(seq_hic.shape)\n",
    "\n",
    "        # write\n",
    "        #seqs_hic_open['targets'][si,:] = seq_hic.astype('float16')\n",
    "        targets[si,:] = seq_hic.astype('float16')\n",
    "\n",
    "    # close sequences coverage file\n",
    "    #seqs_hic_open.close()\n",
    "    \n",
    "\n",
    "    targets=targets.reshape(targets.shape[1])       \n",
    "    targets_mat = utils.from_upper_triu(targets,seq_len_pool,2)\n",
    "\n",
    "    \n",
    "    return(targets_mat)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0d155bbc",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "1478acda",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Define the kernel function (e.g., Gaussian)\n",
    "# def kernel(x, bandwidth):\n",
    "#     return np.exp(-0.5 * (x / bandwidth)**2) / (bandwidth * np.sqrt(2 * np.pi))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "98424fa2",
   "metadata": {},
   "outputs": [],
   "source": [
    "def condense_matrices(target_size, original_size, akita_pred_matrix):\n",
    "    \n",
    "    original_matrix=akita_pred_matrix.copy()\n",
    "    \n",
    "    target_matrix = np.zeros((target_size, target_size))\n",
    "    for target_row in range(target_size):\n",
    "        for target_col in range(target_size):\n",
    "            # Calculate the range of original bins to consider\n",
    "            min_original_row = target_row * original_size // target_size\n",
    "            max_original_row = (target_row + 1) * original_size // target_size\n",
    "            min_original_col = target_col * original_size // target_size\n",
    "            max_original_col = (target_col + 1) * original_size // target_size\n",
    "\n",
    "            # Crop the last row and column by applying crop_weight\n",
    "            cropped_matrix = original_matrix[:max_original_row, :max_original_col]\n",
    "            cropped_matrix[-1, :] *= crop_weight\n",
    "            cropped_matrix[:, -1] *= crop_weight\n",
    "            \n",
    "            kernel_std=2\n",
    "            #kernel = Gaussian2DKernel(x_stddev=kernel_std)\n",
    "\n",
    "            #print('cropped matrix shape', cropped_matrix.shape)\n",
    "            # Apply 2D Gaussian smoothing\n",
    "            smoothed_matrix = gaussian_filter(cropped_matrix, sigma=kernel_std)\n",
    "            #smoothed_matrix=uniform_filter(original_matrix, size=3, mode='nearest', cval=np.nan)\n",
    "\n",
    "            # Store the smoothed value in the target matrix\n",
    "            target_matrix[target_row, target_col] = smoothed_matrix[-1, -1]\n",
    "            #print(smoothed_matrix.shape)\n",
    "            \n",
    "            #renormalize so mean values still same across both matrices \n",
    "            target_matrix=(target_matrix/np.nanmax(target_matrix))*np.nanmax(akita_pred_matrix)\n",
    "            \n",
    "    return(target_matrix)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2535abd3",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "db8aee59",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9f7e7bed",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "45f0d308",
   "metadata": {},
   "outputs": [],
   "source": [
    " "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "8339d000",
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_complex_pred(REF_mat, ALT_mat, REF_label, ALT_label, gene_bin_start, gene_bin_end, \n",
    "                      gene_len, hic_region, file_path, norm_hic_map, ALT_mat_condensed):\n",
    "    \n",
    "    print('plotting ' + ALT_label)\n",
    "    \n",
    "    #ref_gene_bin_start=utils.get_bin(utils.half_patch_size - round(gene_len/2))\n",
    "    #ref_gene_bin_end=utils.get_bin(utils.half_patch_size + round(gene_len/2))\n",
    "      \n",
    "    target_index = 0\n",
    "    vmin=-2; vmax=2\n",
    "    fig, axes = plt.subplots(2, 3, figsize=(16, 10))\n",
    "\n",
    "    # Plot 1, raw HiC\n",
    "    clr = cooler.Cooler(f'{file_path}::resolutions/5000')\n",
    "    fruitpunch = sns.blend_palette(['white', 'red'], as_cmap=True)\n",
    "\n",
    "    CHR=hic_region[0]\n",
    "    seq_start=hic_region[1]\n",
    "    seq_end=hic_region[2]\n",
    "    region = (CHR, seq_start, seq_end)\n",
    "    print('region', region)\n",
    "    \n",
    "    im = axes[0,0].matshow(clr.matrix(balance='KR').fetch(region), cmap=fruitpunch, \n",
    "                     norm=LogNorm(vmin=1, vmax=3), extent=(seq_start, seq_end, seq_end, seq_start))\n",
    "    region_str=(f'{CHR}:{seq_start:,}-{seq_end:,}')\n",
    "    axes[0,0].set_title(f'HiC (5kbp bins) \\n chr{region_str}')\n",
    "\n",
    "    fig.colorbar(im, ax=axes[0,0], fraction=0.046, pad=0.04, label='KR balance')\n",
    "\n",
    "    # Plot 2, Akita ALT pred\n",
    "    im = axes[0,1].matshow(ALT_mat, cmap='RdBu_r', vmax=vmax, vmin=vmin)\n",
    "    axes[0,1].axvline(x=gene_bin_start, color='b', linestyle='--', linewidth=1, label='gene start')\n",
    "    axes[0,1].axvline(x=gene_bin_end, color='b', linestyle='--', linewidth=1, label='gene end')\n",
    "    fig.colorbar(im, ax=axes[0,1], fraction=0.04, pad=0.05, ticks=[-2, -1, 0, 1, 2])\n",
    "    axes[0,1].set_title('Akita ALT Prediction: \\n ' + ALT_label)\n",
    "\n",
    "    # Plot 3, Akita REF pred\n",
    "    im = axes[0,2].matshow(REF_mat, cmap='RdBu_r', vmax=vmax, vmin=vmin)\n",
    "    axes[0,2].axvline(x=gene_bin_start, color='b', linestyle='--', linewidth=1, label='gene start')\n",
    "    axes[0,2].axvline(x=gene_bin_end, color='b', linestyle='--', linewidth=1, label='gene end')\n",
    "    cbar = fig.colorbar(im, ax=axes[0,2], fraction=0.04, pad=0.05, ticks=[-2, -1, 0, 1, 2])\n",
    "    axes[0,2].set_title('Akita REF Prediction' + '\\n' + REF_label)\n",
    "    \n",
    "    \n",
    "    #Plot 4, HiC normalized to Akita\n",
    "    im = axes[1,0].matshow(norm_hic_map, cmap= 'RdBu_r', vmax=vmax, vmin=vmin)\n",
    "    fig.colorbar(im, ax=axes[1,0], fraction=.04, pad = 0.05, ticks=[-2,-1, 0, 1,2]);\n",
    "    region_str=(f'{CHR}:{seq_start:,}-{seq_end:,}')\n",
    "    axes[1,0].set_title(f'Normalized HiC (5kbp bins) \\n chr{region_str}')\n",
    "\n",
    "    #Plot 5, Akita prediction with 200 bins total instead of 448 \n",
    "    im = axes[1,1].matshow(ALT_mat_condensed, cmap= 'RdBu_r', vmax=vmax, vmin=vmin)\n",
    "    fig.colorbar(im, ax=axes[1,1], fraction=.04, pad = 0.05, ticks=[-2,-1, 0, 1,2]);\n",
    "    axes[1,1].set_title(f'Akita ALT Prediction (100 bins) \\n ' + ALT_label)\n",
    "\n",
    "\n",
    "\n",
    "    #Plot 6, currently empty    \n",
    "    im = axes[1,2].matshow(np.zeros((10,10)), cmap= 'RdBu_r', vmax=vmax, vmin=vmin)\n",
    "    fig.colorbar(im, ax=axes[1,2], fraction=.04, pad = 0.05, ticks=[-2,-1, 0, 1,2]);\n",
    "\n",
    "\n",
    "\n",
    "    plt.tight_layout()\n",
    "    plt.savefig('{}{}_Predictions.pdf'.format(fig_dir, ALT_label), dpi=300)   # save the figure to file\n",
    "    plt.close('all')\n",
    "    #plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "573f8a7a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "0fecd6a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "#PARAMETERS FOR NORMALIZING \n",
    "\n",
    "clip=2\n",
    "crop_bp=0\n",
    "resolution=5000\n",
    "#sigma 1, width 5\n",
    "#stddev is sigma\n",
    "#kernel_size (which i added) is width, i think \n",
    "kernel_stddev=1\n",
    "#kernel_size=5\n",
    "pool_width=resolution\n",
    "no_log=False\n",
    "crop=False\n",
    "as_obsexp=True\n",
    "global_obsexp=False\n",
    "diagonal_offset=2\n",
    "#options are KR, VC, VC_SQRT\n",
    "norm_type='VC_SQRT'\n",
    "output_file='/pollard/home/shzhang/test1.cool'\n",
    "\n",
    "\n",
    "def myceil(x, base=resolution):\n",
    "    return base * np.ceil(x/base)\n",
    "\n",
    "def myfloor(x, base=resolution):\n",
    "    return base * np.floor(x/base)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "c83fb544",
   "metadata": {},
   "outputs": [],
   "source": [
    "#PARAMETERS FOR TURNING INTO SAME SIZE MATRICES \n",
    "\n",
    "# Define the sizes of the original and target matrices\n",
    "original_size = 448\n",
    "target_size = 100\n",
    "\n",
    "# Define the weight for cropping the last row and column by 60%\n",
    "crop_weight = 0.4 \n",
    "\n",
    "# Define the kernel bandwidth\n",
    "bandwidth = 5  # Adjust as needed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e47d8e56",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "a95f0d86",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SMF-12-23_walk_9_MDM2_containing.fa\n",
      "starting from left most sequence\n",
      "/pollard/home/shzhang/akita_variant_scoring/akdemir_collab/data/mcool/SMF-12-23_walk_9_MDM2_containing.fa.mcool::resolutions/5000\n",
      "ModelSeq(chr='12', start=65000, end=985000, label=None)\n",
      "(16653,)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/pollard/home/shzhang/miniconda3/envs/akita-working/lib/python3.9/site-packages/cooltools/lib/numutils.py:1376: RuntimeWarning: invalid value encountered in true_divide\n",
      "  val_cur = ar_cur / armask_cur\n",
      "/tmp/ipykernel_6359/2476405126.py:32: RuntimeWarning: invalid value encountered in true_divide\n",
      "  target_matrix=(target_matrix/np.nanmax(target_matrix))*np.nanmax(akita_pred_matrix)\n",
      "/tmp/ipykernel_6359/2476405126.py:32: RuntimeWarning: All-NaN slice encountered\n",
      "  target_matrix=(target_matrix/np.nanmax(target_matrix))*np.nanmax(akita_pred_matrix)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "plotting SMF-12-23_walk_9_MDM2_containing\n",
      "region ('12', 65000.0, 985000.0)\n",
      "SMF-12-23_walk_8_MDM2_containing.fa\n",
      "ending at rightmost sequence \n",
      "/pollard/home/shzhang/akita_variant_scoring/akdemir_collab/data/mcool/SMF-12-23_walk_8_MDM2_containing.fa.mcool::resolutions/5000\n",
      "ModelSeq(chr='12', start=3340000, end=4265000, label=None)\n",
      "(16836,)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/pollard/home/shzhang/miniconda3/envs/akita-working/lib/python3.9/site-packages/cooltools/lib/numutils.py:1376: RuntimeWarning: invalid value encountered in true_divide\n",
      "  val_cur = ar_cur / armask_cur\n",
      "/tmp/ipykernel_6359/2476405126.py:32: RuntimeWarning: invalid value encountered in true_divide\n",
      "  target_matrix=(target_matrix/np.nanmax(target_matrix))*np.nanmax(akita_pred_matrix)\n",
      "/tmp/ipykernel_6359/2476405126.py:32: RuntimeWarning: All-NaN slice encountered\n",
      "  target_matrix=(target_matrix/np.nanmax(target_matrix))*np.nanmax(akita_pred_matrix)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "plotting SMF-12-23_walk_8_MDM2_containing\n",
      "region ('12', 3340000.0, 4265000.0)\n",
      "SMF-12-12_walk_3_KRAS_containing.fa\n",
      "1931914\n",
      "2980490\n",
      "4110595\n",
      "/pollard/home/shzhang/akita_variant_scoring/akdemir_collab/data/mcool/SMF-12-12_walk_3_KRAS_containing.fa.mcool::resolutions/5000\n",
      "ModelSeq(chr='12', start=1995000, end=2915000, label=None)\n",
      "WARNING: /pollard/home/shzhang/akita_variant_scoring/akdemir_collab/data/mcool/SMF-12-12_walk_3_KRAS_containing.fa.mcool::resolutions/5000 >50% bins filtered, check:  12:1995000-2915000. \n",
      "(16653,)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/pollard/home/shzhang/miniconda3/envs/akita-working/lib/python3.9/site-packages/cooler/api.py:744: RuntimeWarning: divide by zero encountered in true_divide\n",
      "  bias1 = 1 / bias1\n",
      "/pollard/home/shzhang/miniconda3/envs/akita-working/lib/python3.9/site-packages/cooler/api.py:745: RuntimeWarning: divide by zero encountered in true_divide\n",
      "  bias2 = 1 / bias2\n",
      "/pollard/home/shzhang/miniconda3/envs/akita-working/lib/python3.9/site-packages/cooler/api.py:746: RuntimeWarning: invalid value encountered in multiply\n",
      "  arr = arr * np.outer(bias1, bias2)\n",
      "/pollard/home/shzhang/miniconda3/envs/akita-working/lib/python3.9/site-packages/cooltools/lib/numutils.py:1376: RuntimeWarning: invalid value encountered in true_divide\n",
      "  val_cur = ar_cur / armask_cur\n",
      "/tmp/ipykernel_6359/3696775899.py:114: RuntimeWarning: divide by zero encountered in log\n",
      "  seq_hic_obsexp = np.log(seq_hic_obsexp)\n",
      "/tmp/ipykernel_6359/2476405126.py:32: RuntimeWarning: invalid value encountered in true_divide\n",
      "  target_matrix=(target_matrix/np.nanmax(target_matrix))*np.nanmax(akita_pred_matrix)\n",
      "/tmp/ipykernel_6359/2476405126.py:32: RuntimeWarning: All-NaN slice encountered\n",
      "  target_matrix=(target_matrix/np.nanmax(target_matrix))*np.nanmax(akita_pred_matrix)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "plotting SMF-12-12_walk_3_KRAS_containing\n",
      "region ('12', 1995000.0, 2915000.0)\n",
      "SMF-12-23_walk_9_GLI1_containing.fa\n",
      "2814341\n",
      "3862917\n",
      "4298594\n",
      "/pollard/home/shzhang/akita_variant_scoring/akdemir_collab/data/mcool/SMF-12-23_walk_9_GLI1_containing.fa.mcool::resolutions/5000\n",
      "ModelSeq(chr='12', start=2875000, end=3800000, label=None)\n",
      "(16836,)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/pollard/home/shzhang/miniconda3/envs/akita-working/lib/python3.9/site-packages/cooltools/lib/numutils.py:1376: RuntimeWarning: invalid value encountered in true_divide\n",
      "  val_cur = ar_cur / armask_cur\n",
      "/tmp/ipykernel_6359/2476405126.py:32: RuntimeWarning: invalid value encountered in true_divide\n",
      "  target_matrix=(target_matrix/np.nanmax(target_matrix))*np.nanmax(akita_pred_matrix)\n",
      "/tmp/ipykernel_6359/2476405126.py:32: RuntimeWarning: All-NaN slice encountered\n",
      "  target_matrix=(target_matrix/np.nanmax(target_matrix))*np.nanmax(akita_pred_matrix)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "plotting SMF-12-23_walk_9_GLI1_containing\n",
      "region ('12', 2875000.0, 3800000.0)\n",
      "SMF-12-12_walk_4_KRAS_containing.fa\n",
      "starting from left most sequence\n",
      "/pollard/home/shzhang/akita_variant_scoring/akdemir_collab/data/mcool/SMF-12-12_walk_4_KRAS_containing.fa.mcool::resolutions/5000\n",
      "ModelSeq(chr='12', start=65000, end=985000, label=None)\n",
      "WARNING: /pollard/home/shzhang/akita_variant_scoring/akdemir_collab/data/mcool/SMF-12-12_walk_4_KRAS_containing.fa.mcool::resolutions/5000 >50% bins filtered, check:  12:65000-985000. \n",
      "(16653,)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/pollard/home/shzhang/miniconda3/envs/akita-working/lib/python3.9/site-packages/cooler/api.py:744: RuntimeWarning: divide by zero encountered in true_divide\n",
      "  bias1 = 1 / bias1\n",
      "/pollard/home/shzhang/miniconda3/envs/akita-working/lib/python3.9/site-packages/cooler/api.py:745: RuntimeWarning: divide by zero encountered in true_divide\n",
      "  bias2 = 1 / bias2\n",
      "/pollard/home/shzhang/miniconda3/envs/akita-working/lib/python3.9/site-packages/cooler/api.py:746: RuntimeWarning: invalid value encountered in multiply\n",
      "  arr = arr * np.outer(bias1, bias2)\n",
      "/pollard/home/shzhang/miniconda3/envs/akita-working/lib/python3.9/site-packages/numpy/lib/nanfunctions.py:1119: RuntimeWarning: All-NaN slice encountered\n",
      "  r, k = function_base._ureduce(a, func=_nanmedian, axis=axis, out=out,\n",
      "/pollard/home/shzhang/miniconda3/envs/akita-working/lib/python3.9/site-packages/cooltools/lib/numutils.py:1376: RuntimeWarning: invalid value encountered in true_divide\n",
      "  val_cur = ar_cur / armask_cur\n",
      "/tmp/ipykernel_6359/2476405126.py:32: RuntimeWarning: invalid value encountered in true_divide\n",
      "  target_matrix=(target_matrix/np.nanmax(target_matrix))*np.nanmax(akita_pred_matrix)\n",
      "/tmp/ipykernel_6359/2476405126.py:32: RuntimeWarning: All-NaN slice encountered\n",
      "  target_matrix=(target_matrix/np.nanmax(target_matrix))*np.nanmax(akita_pred_matrix)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "plotting SMF-12-12_walk_4_KRAS_containing\n",
      "region ('12', 65000.0, 985000.0)\n",
      "SMF-12-12_walk_5_KRAS_containing.fa\n",
      "starting from left most sequence\n",
      "error!!! in SMF-12-12_walk_5_KRAS_containing.fa.mcool\n",
      "SMF-12-17_walk_7_MDM2_containing.fa\n",
      "starting from left most sequence\n",
      "/pollard/home/shzhang/akita_variant_scoring/akdemir_collab/data/mcool/SMF-12-17_walk_7_MDM2_containing.fa.mcool::resolutions/5000\n",
      "ModelSeq(chr='12', start=65000, end=985000, label=None)\n",
      "(16653,)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/pollard/home/shzhang/miniconda3/envs/akita-working/lib/python3.9/site-packages/cooltools/lib/numutils.py:1376: RuntimeWarning: invalid value encountered in true_divide\n",
      "  val_cur = ar_cur / armask_cur\n",
      "/tmp/ipykernel_6359/2476405126.py:32: RuntimeWarning: invalid value encountered in true_divide\n",
      "  target_matrix=(target_matrix/np.nanmax(target_matrix))*np.nanmax(akita_pred_matrix)\n",
      "/tmp/ipykernel_6359/2476405126.py:32: RuntimeWarning: All-NaN slice encountered\n",
      "  target_matrix=(target_matrix/np.nanmax(target_matrix))*np.nanmax(akita_pred_matrix)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "plotting SMF-12-17_walk_7_MDM2_containing\n",
      "region ('12', 65000.0, 985000.0)\n",
      "SMF-12-23_walk_11_GLI1_containing.fa\n",
      "623024\n",
      "1671600\n",
      "1716777\n",
      "/pollard/home/shzhang/akita_variant_scoring/akdemir_collab/data/mcool/SMF-12-23_walk_11_GLI1_containing.fa.mcool::resolutions/5000\n",
      "ModelSeq(chr='12', start=685000, end=1610000, label=None)\n",
      "(16836,)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/pollard/home/shzhang/miniconda3/envs/akita-working/lib/python3.9/site-packages/cooltools/lib/numutils.py:1376: RuntimeWarning: invalid value encountered in true_divide\n",
      "  val_cur = ar_cur / armask_cur\n",
      "/tmp/ipykernel_6359/2476405126.py:32: RuntimeWarning: invalid value encountered in true_divide\n",
      "  target_matrix=(target_matrix/np.nanmax(target_matrix))*np.nanmax(akita_pred_matrix)\n",
      "/tmp/ipykernel_6359/2476405126.py:32: RuntimeWarning: All-NaN slice encountered\n",
      "  target_matrix=(target_matrix/np.nanmax(target_matrix))*np.nanmax(akita_pred_matrix)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "plotting SMF-12-23_walk_11_GLI1_containing\n",
      "region ('12', 685000.0, 1610000.0)\n",
      "SMF-12-12_walk_4_GLI1_containing.fa\n",
      "ending at rightmost sequence \n",
      "/pollard/home/shzhang/akita_variant_scoring/akdemir_collab/data/mcool/SMF-12-12_walk_4_GLI1_containing.fa.mcool::resolutions/5000\n",
      "ModelSeq(chr='12', start=595000, end=1515000, label=None)\n",
      "(16653,)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/pollard/home/shzhang/miniconda3/envs/akita-working/lib/python3.9/site-packages/cooltools/lib/numutils.py:1376: RuntimeWarning: invalid value encountered in true_divide\n",
      "  val_cur = ar_cur / armask_cur\n",
      "/tmp/ipykernel_6359/2476405126.py:32: RuntimeWarning: invalid value encountered in true_divide\n",
      "  target_matrix=(target_matrix/np.nanmax(target_matrix))*np.nanmax(akita_pred_matrix)\n",
      "/tmp/ipykernel_6359/2476405126.py:32: RuntimeWarning: All-NaN slice encountered\n",
      "  target_matrix=(target_matrix/np.nanmax(target_matrix))*np.nanmax(akita_pred_matrix)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "plotting SMF-12-12_walk_4_GLI1_containing\n",
      "region ('12', 595000.0, 1515000.0)\n",
      "SMF-12-17_walk_4_GLI1_containing.fa\n",
      "starting from left most sequence\n",
      "/pollard/home/shzhang/akita_variant_scoring/akdemir_collab/data/mcool/SMF-12-17_walk_4_GLI1_containing.fa.mcool::resolutions/5000\n",
      "ModelSeq(chr='12', start=65000, end=985000, label=None)\n",
      "(16653,)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/pollard/home/shzhang/miniconda3/envs/akita-working/lib/python3.9/site-packages/cooltools/lib/numutils.py:1376: RuntimeWarning: invalid value encountered in true_divide\n",
      "  val_cur = ar_cur / armask_cur\n",
      "/tmp/ipykernel_6359/2476405126.py:32: RuntimeWarning: invalid value encountered in true_divide\n",
      "  target_matrix=(target_matrix/np.nanmax(target_matrix))*np.nanmax(akita_pred_matrix)\n",
      "/tmp/ipykernel_6359/2476405126.py:32: RuntimeWarning: All-NaN slice encountered\n",
      "  target_matrix=(target_matrix/np.nanmax(target_matrix))*np.nanmax(akita_pred_matrix)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "plotting SMF-12-17_walk_4_GLI1_containing\n",
      "region ('12', 65000.0, 985000.0)\n",
      "SMF-12-23_walk_11_MDM2_containing.fa\n",
      "starting from left most sequence\n",
      "error!!! in SMF-12-23_walk_11_MDM2_containing.fa.mcool\n",
      "SMF-12-17_walk_9_MDM2_containing.fa\n",
      "starting from left most sequence\n",
      "/pollard/home/shzhang/akita_variant_scoring/akdemir_collab/data/mcool/SMF-12-17_walk_9_MDM2_containing.fa.mcool::resolutions/5000\n",
      "ModelSeq(chr='12', start=65000, end=985000, label=None)\n",
      "(16653,)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/pollard/home/shzhang/miniconda3/envs/akita-working/lib/python3.9/site-packages/cooltools/lib/numutils.py:1376: RuntimeWarning: invalid value encountered in true_divide\n",
      "  val_cur = ar_cur / armask_cur\n",
      "/tmp/ipykernel_6359/2476405126.py:32: RuntimeWarning: invalid value encountered in true_divide\n",
      "  target_matrix=(target_matrix/np.nanmax(target_matrix))*np.nanmax(akita_pred_matrix)\n",
      "/tmp/ipykernel_6359/2476405126.py:32: RuntimeWarning: All-NaN slice encountered\n",
      "  target_matrix=(target_matrix/np.nanmax(target_matrix))*np.nanmax(akita_pred_matrix)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "plotting SMF-12-17_walk_9_MDM2_containing\n",
      "region ('12', 65000.0, 985000.0)\n",
      "SMF-12-23_walk_10_MDM2_containing.fa\n",
      "starting from left most sequence\n",
      "/pollard/home/shzhang/akita_variant_scoring/akdemir_collab/data/mcool/SMF-12-23_walk_10_MDM2_containing.fa.mcool::resolutions/5000\n",
      "ModelSeq(chr='12', start=65000, end=985000, label=None)\n",
      "(16653,)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/pollard/home/shzhang/miniconda3/envs/akita-working/lib/python3.9/site-packages/cooltools/lib/numutils.py:1376: RuntimeWarning: invalid value encountered in true_divide\n",
      "  val_cur = ar_cur / armask_cur\n",
      "/tmp/ipykernel_6359/2476405126.py:32: RuntimeWarning: invalid value encountered in true_divide\n",
      "  target_matrix=(target_matrix/np.nanmax(target_matrix))*np.nanmax(akita_pred_matrix)\n",
      "/tmp/ipykernel_6359/2476405126.py:32: RuntimeWarning: All-NaN slice encountered\n",
      "  target_matrix=(target_matrix/np.nanmax(target_matrix))*np.nanmax(akita_pred_matrix)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "plotting SMF-12-23_walk_10_MDM2_containing\n",
      "region ('12', 65000.0, 985000.0)\n",
      "SMF-12-17_walk_8_MDM2_containing.fa\n",
      "starting from left most sequence\n",
      "/pollard/home/shzhang/akita_variant_scoring/akdemir_collab/data/mcool/SMF-12-17_walk_8_MDM2_containing.fa.mcool::resolutions/5000\n",
      "ModelSeq(chr='12', start=65000, end=985000, label=None)\n",
      "(16653,)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/pollard/home/shzhang/miniconda3/envs/akita-working/lib/python3.9/site-packages/cooltools/lib/numutils.py:1376: RuntimeWarning: invalid value encountered in true_divide\n",
      "  val_cur = ar_cur / armask_cur\n",
      "/tmp/ipykernel_6359/2476405126.py:32: RuntimeWarning: invalid value encountered in true_divide\n",
      "  target_matrix=(target_matrix/np.nanmax(target_matrix))*np.nanmax(akita_pred_matrix)\n",
      "/tmp/ipykernel_6359/2476405126.py:32: RuntimeWarning: All-NaN slice encountered\n",
      "  target_matrix=(target_matrix/np.nanmax(target_matrix))*np.nanmax(akita_pred_matrix)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "plotting SMF-12-17_walk_8_MDM2_containing\n",
      "region ('12', 65000.0, 985000.0)\n",
      "SMF-12-17_walk_5_GLI1_containing.fa\n",
      "starting from left most sequence\n",
      "/pollard/home/shzhang/akita_variant_scoring/akdemir_collab/data/mcool/SMF-12-17_walk_5_GLI1_containing.fa.mcool::resolutions/5000\n",
      "ModelSeq(chr='12', start=65000, end=985000, label=None)\n",
      "(16653,)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/pollard/home/shzhang/miniconda3/envs/akita-working/lib/python3.9/site-packages/cooler/api.py:744: RuntimeWarning: divide by zero encountered in true_divide\n",
      "  bias1 = 1 / bias1\n",
      "/pollard/home/shzhang/miniconda3/envs/akita-working/lib/python3.9/site-packages/cooler/api.py:745: RuntimeWarning: divide by zero encountered in true_divide\n",
      "  bias2 = 1 / bias2\n",
      "/pollard/home/shzhang/miniconda3/envs/akita-working/lib/python3.9/site-packages/cooler/api.py:746: RuntimeWarning: invalid value encountered in multiply\n",
      "  arr = arr * np.outer(bias1, bias2)\n",
      "/pollard/home/shzhang/miniconda3/envs/akita-working/lib/python3.9/site-packages/cooltools/lib/numutils.py:1376: RuntimeWarning: invalid value encountered in true_divide\n",
      "  val_cur = ar_cur / armask_cur\n",
      "/tmp/ipykernel_6359/2476405126.py:32: RuntimeWarning: invalid value encountered in true_divide\n",
      "  target_matrix=(target_matrix/np.nanmax(target_matrix))*np.nanmax(akita_pred_matrix)\n",
      "/tmp/ipykernel_6359/2476405126.py:32: RuntimeWarning: All-NaN slice encountered\n",
      "  target_matrix=(target_matrix/np.nanmax(target_matrix))*np.nanmax(akita_pred_matrix)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "plotting SMF-12-17_walk_5_GLI1_containing\n",
      "region ('12', 65000.0, 985000.0)\n",
      "SMF-12-17_walk_3_GLI1_containing.fa\n",
      "917253\n",
      "1965829\n",
      "2201610\n",
      "/pollard/home/shzhang/akita_variant_scoring/akdemir_collab/data/mcool/SMF-12-17_walk_3_GLI1_containing.fa.mcool::resolutions/5000\n",
      "ModelSeq(chr='12', start=980000, end=1905000, label=None)\n",
      "(16836,)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/pollard/home/shzhang/miniconda3/envs/akita-working/lib/python3.9/site-packages/cooler/api.py:744: RuntimeWarning: divide by zero encountered in true_divide\n",
      "  bias1 = 1 / bias1\n",
      "/pollard/home/shzhang/miniconda3/envs/akita-working/lib/python3.9/site-packages/cooler/api.py:745: RuntimeWarning: divide by zero encountered in true_divide\n",
      "  bias2 = 1 / bias2\n",
      "/pollard/home/shzhang/miniconda3/envs/akita-working/lib/python3.9/site-packages/cooler/api.py:746: RuntimeWarning: invalid value encountered in multiply\n",
      "  arr = arr * np.outer(bias1, bias2)\n",
      "/pollard/home/shzhang/miniconda3/envs/akita-working/lib/python3.9/site-packages/cooltools/lib/numutils.py:1376: RuntimeWarning: invalid value encountered in true_divide\n",
      "  val_cur = ar_cur / armask_cur\n",
      "/tmp/ipykernel_6359/2476405126.py:32: RuntimeWarning: invalid value encountered in true_divide\n",
      "  target_matrix=(target_matrix/np.nanmax(target_matrix))*np.nanmax(akita_pred_matrix)\n",
      "/tmp/ipykernel_6359/2476405126.py:32: RuntimeWarning: All-NaN slice encountered\n",
      "  target_matrix=(target_matrix/np.nanmax(target_matrix))*np.nanmax(akita_pred_matrix)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "plotting SMF-12-17_walk_3_GLI1_containing\n",
      "region ('12', 980000.0, 1905000.0)\n"
     ]
    }
   ],
   "source": [
    "for mcool_name in list_mcool:\n",
    "    \n",
    "    try:\n",
    "    \n",
    "        #make akita prediction\n",
    "        REF_mat, ALT_mat, R_label, A_label, gene_start, gene_end, gene_len, CHR, seq_start, seq_end = akita_pred(mcool_name)\n",
    "       \n",
    "        #need to crop the ends because akita crops the ends \n",
    "        crop_ends=((512-448)*2048)/2\n",
    "        seqs_info=(CHR[3:], myfloor(seq_start+crop_ends), myceil(seq_end-crop_ends))\n",
    "        hic_region=seqs_info\n",
    "        file_path=f'{data_dir}mcool/{mcool_name}'\n",
    "        \n",
    "        #normalize hic map\n",
    "        norm_hic_map=normalize_hic(file_path, seqs_info, resolution, clip, crop_bp, kernel_stddev, pool_width, \n",
    "                  no_log, crop, as_obsexp, global_obsexp, diagonal_offset, norm_type, output_file)\n",
    "        #print('target shape', norm_hic_map.shape)\n",
    "        \n",
    "        \n",
    "        #kernel for akita prediction        \n",
    "        ALT_mat_condensed=condense_matrices(target_size=100, original_size=ALT_mat.shape[0], akita_pred_matrix=ALT_mat)\n",
    "\n",
    "        \n",
    "        #kernel for hic map tbd \n",
    "               \n",
    "        \n",
    "        #plot all plots \n",
    "        plot_complex_pred(REF_mat,ALT_mat,R_label,A_label, gene_start, \n",
    "                          gene_end, gene_len, hic_region, file_path, norm_hic_map, ALT_mat_condensed)\n",
    "        \n",
    "        \n",
    "        \n",
    "        \n",
    "    except:\n",
    "        print('error!!! in ' + mcool_name)\n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8d1f73ba",
   "metadata": {},
   "outputs": [],
   "source": [
    "2980490-1931914"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3bce50b2",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "08779164",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "    original_size = 448\n",
    "    target_size = 100\n",
    "\n",
    "    # Define the weight for cropping the last row and column by 60%\n",
    "    crop_weight = 0.4 \n",
    "\n",
    "\n",
    "    # Define the kernel bandwidth\n",
    "    bandwidth = 10  # Adjust as needed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cc2845d2",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dff394ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pdg"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1e9611fa",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "54ddbf31",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7838b1e7",
   "metadata": {},
   "outputs": [],
   "source": [
    "#let's plot\n",
    "#original, unwhatevered mcool\n",
    "#mcool with normalization\n",
    "#akita prediction, with lines for genes DONE \n",
    "\n",
    "#underneath, plot windowed prediction with same number of bins \n",
    "#mcool\n",
    "#akita prediction "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b0e5d6f7",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6f1860bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2f341b95",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "337ef3d2",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
