{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "9362b7b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd \n",
    "import sys, os, psutil, io\n",
    "import gzip\n",
    "import time \n",
    "import warnings\n",
    "\n",
    "from pybedtools import BedTool\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "warnings.filterwarnings('ignore')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "e85d3362",
   "metadata": {},
   "outputs": [],
   "source": [
    "synteny_condensed=pd.read_csv('/pollard/home/shzhang/data/synteny/hg38_mm10_condensed.txt', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "c3792fd8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>hg_chr</th>\n",
       "      <th>hg_start</th>\n",
       "      <th>hg_end</th>\n",
       "      <th>hg_len</th>\n",
       "      <th>hg_seq</th>\n",
       "      <th>mm_chr</th>\n",
       "      <th>mm_start</th>\n",
       "      <th>mm_end</th>\n",
       "      <th>mm_len</th>\n",
       "      <th>mm_seq</th>\n",
       "      <th>total_align_len</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>56554</td>\n",
       "      <td>58923</td>\n",
       "      <td>2370</td>\n",
       "      <td>CACATTATTCTGCAGTTTTTATCTTGCTTACGTATTTTTGGAATGT...</td>\n",
       "      <td>chr2</td>\n",
       "      <td>111473169</td>\n",
       "      <td>111475499</td>\n",
       "      <td>2331</td>\n",
       "      <td>cacattgtattgta--ctctattatgttt-tattgtattatattat...</td>\n",
       "      <td>2518</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>60916</td>\n",
       "      <td>61124</td>\n",
       "      <td>209</td>\n",
       "      <td>TTAATATTTCCCACTTCATCATTACTAACCTCACTCACTCTGAACT...</td>\n",
       "      <td>chr2</td>\n",
       "      <td>111476095</td>\n",
       "      <td>111476298</td>\n",
       "      <td>204</td>\n",
       "      <td>ttAATAT--CCCAGTGCACC-TCACTACTCTGGCTCACGCTGAGCC...</td>\n",
       "      <td>209</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>61125</td>\n",
       "      <td>61835</td>\n",
       "      <td>711</td>\n",
       "      <td>AGTTCAAGACGTTTAATCAAGACCCTTCACTTACTTAACTATACCT...</td>\n",
       "      <td>chr2</td>\n",
       "      <td>111477065</td>\n",
       "      <td>111477685</td>\n",
       "      <td>621</td>\n",
       "      <td>aGCCCAAGAAAATTAAACAAATCCCTTTACTAATTTAAATGTATCT...</td>\n",
       "      <td>731</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>61943</td>\n",
       "      <td>61971</td>\n",
       "      <td>29</td>\n",
       "      <td>ctccgcctcccggggtcaagctattctcc</td>\n",
       "      <td>chr2</td>\n",
       "      <td>111477686</td>\n",
       "      <td>111477713</td>\n",
       "      <td>28</td>\n",
       "      <td>CTCTGCTTTGTGTGGAGAAG-TATTGTCC</td>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>62160</td>\n",
       "      <td>62467</td>\n",
       "      <td>308</td>\n",
       "      <td>tgtaaaactttttcctaatttaacagaaaaataatagtattatatt...</td>\n",
       "      <td>chr2</td>\n",
       "      <td>111477762</td>\n",
       "      <td>111478104</td>\n",
       "      <td>343</td>\n",
       "      <td>tgtaa----tttttcttatttaaCA-AGAAATGAGA-------ATT...</td>\n",
       "      <td>372</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1229834</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57107075</td>\n",
       "      <td>57107468</td>\n",
       "      <td>394</td>\n",
       "      <td>TGATTATGTTCGTTGTGACTGACAGCTGGATTGtagtggttaagaa...</td>\n",
       "      <td>chrX_GL456233_random</td>\n",
       "      <td>316168</td>\n",
       "      <td>316535</td>\n",
       "      <td>368</td>\n",
       "      <td>TAATGATGTTGATTGTGACTAGCAACAAGACTGtagtgattaagag...</td>\n",
       "      <td>397</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1229835</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57125268</td>\n",
       "      <td>57126152</td>\n",
       "      <td>885</td>\n",
       "      <td>aaagctgcagtgagtcgtaattgtgccactgcactccagcttgggc...</td>\n",
       "      <td>chrX_GL456233_random</td>\n",
       "      <td>321571</td>\n",
       "      <td>322437</td>\n",
       "      <td>867</td>\n",
       "      <td>aaggctgtatcg---cctaatagtgccacttcccatgagccaagca...</td>\n",
       "      <td>954</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1229836</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57126153</td>\n",
       "      <td>57126418</td>\n",
       "      <td>266</td>\n",
       "      <td>AGGTTGTTACTGCAAATCATTCATGTAGTCAGTGCCAGGGGTTAAA...</td>\n",
       "      <td>chrX_GL456233_random</td>\n",
       "      <td>322699</td>\n",
       "      <td>322963</td>\n",
       "      <td>265</td>\n",
       "      <td>ATATTGTTACTACAAATCATTCATATAGTCAGTGTGAAGAATAAAA...</td>\n",
       "      <td>266</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1229837</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57126419</td>\n",
       "      <td>57126561</td>\n",
       "      <td>143</td>\n",
       "      <td>CTTTTAATAGTGAAACACATGACCAGGCAGTGGC-TTTCCTATGAA...</td>\n",
       "      <td>chrX_GL456233_random</td>\n",
       "      <td>323295</td>\n",
       "      <td>323435</td>\n",
       "      <td>141</td>\n",
       "      <td>CTTTTCATCATAAAGTACATGACCAGGAAATATCTTTTCTTGTGAA...</td>\n",
       "      <td>150</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1229838</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57128357</td>\n",
       "      <td>57130391</td>\n",
       "      <td>2035</td>\n",
       "      <td>TATGATTGATTCAAGAAAATAAAATTGCTCTCCTCGTCCCTCCAGG...</td>\n",
       "      <td>chrX_GL456233_random</td>\n",
       "      <td>324344</td>\n",
       "      <td>326292</td>\n",
       "      <td>1949</td>\n",
       "      <td>TGTGATAGAGTCATGGAAGTAAAATTGCTTTCTTTCTCCTTCCAGG...</td>\n",
       "      <td>2097</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1229839 rows Ã— 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        hg_chr  hg_start    hg_end  hg_len  \\\n",
       "0         chr1     56554     58923    2370   \n",
       "1         chr1     60916     61124     209   \n",
       "2         chr1     61125     61835     711   \n",
       "3         chr1     61943     61971      29   \n",
       "4         chr1     62160     62467     308   \n",
       "...        ...       ...       ...     ...   \n",
       "1229834   chrY  57107075  57107468     394   \n",
       "1229835   chrY  57125268  57126152     885   \n",
       "1229836   chrY  57126153  57126418     266   \n",
       "1229837   chrY  57126419  57126561     143   \n",
       "1229838   chrY  57128357  57130391    2035   \n",
       "\n",
       "                                                    hg_seq  \\\n",
       "0        CACATTATTCTGCAGTTTTTATCTTGCTTACGTATTTTTGGAATGT...   \n",
       "1        TTAATATTTCCCACTTCATCATTACTAACCTCACTCACTCTGAACT...   \n",
       "2        AGTTCAAGACGTTTAATCAAGACCCTTCACTTACTTAACTATACCT...   \n",
       "3                            ctccgcctcccggggtcaagctattctcc   \n",
       "4        tgtaaaactttttcctaatttaacagaaaaataatagtattatatt...   \n",
       "...                                                    ...   \n",
       "1229834  TGATTATGTTCGTTGTGACTGACAGCTGGATTGtagtggttaagaa...   \n",
       "1229835  aaagctgcagtgagtcgtaattgtgccactgcactccagcttgggc...   \n",
       "1229836  AGGTTGTTACTGCAAATCATTCATGTAGTCAGTGCCAGGGGTTAAA...   \n",
       "1229837  CTTTTAATAGTGAAACACATGACCAGGCAGTGGC-TTTCCTATGAA...   \n",
       "1229838  TATGATTGATTCAAGAAAATAAAATTGCTCTCCTCGTCCCTCCAGG...   \n",
       "\n",
       "                       mm_chr   mm_start     mm_end  mm_len  \\\n",
       "0                        chr2  111473169  111475499    2331   \n",
       "1                        chr2  111476095  111476298     204   \n",
       "2                        chr2  111477065  111477685     621   \n",
       "3                        chr2  111477686  111477713      28   \n",
       "4                        chr2  111477762  111478104     343   \n",
       "...                       ...        ...        ...     ...   \n",
       "1229834  chrX_GL456233_random     316168     316535     368   \n",
       "1229835  chrX_GL456233_random     321571     322437     867   \n",
       "1229836  chrX_GL456233_random     322699     322963     265   \n",
       "1229837  chrX_GL456233_random     323295     323435     141   \n",
       "1229838  chrX_GL456233_random     324344     326292    1949   \n",
       "\n",
       "                                                    mm_seq  total_align_len  \n",
       "0        cacattgtattgta--ctctattatgttt-tattgtattatattat...             2518  \n",
       "1        ttAATAT--CCCAGTGCACC-TCACTACTCTGGCTCACGCTGAGCC...              209  \n",
       "2        aGCCCAAGAAAATTAAACAAATCCCTTTACTAATTTAAATGTATCT...              731  \n",
       "3                            CTCTGCTTTGTGTGGAGAAG-TATTGTCC               29  \n",
       "4        tgtaa----tttttcttatttaaCA-AGAAATGAGA-------ATT...              372  \n",
       "...                                                    ...              ...  \n",
       "1229834  TAATGATGTTGATTGTGACTAGCAACAAGACTGtagtgattaagag...              397  \n",
       "1229835  aaggctgtatcg---cctaatagtgccacttcccatgagccaagca...              954  \n",
       "1229836  ATATTGTTACTACAAATCATTCATATAGTCAGTGTGAAGAATAAAA...              266  \n",
       "1229837  CTTTTCATCATAAAGTACATGACCAGGAAATATCTTTTCTTGTGAA...              150  \n",
       "1229838  TGTGATAGAGTCATGGAAGTAAAATTGCTTTCTTTCTCCTTCCAGG...             2097  \n",
       "\n",
       "[1229839 rows x 11 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "synteny_condensed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "908d361e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n",
      "True\n"
     ]
    }
   ],
   "source": [
    "#check that lengths of seqs and alignments are correct  \n",
    "synteny_condensed['hg_dash_n']=synteny_condensed['hg_seq'].str.count('-')\n",
    "synteny_condensed['mm_dash_n']=synteny_condensed['mm_seq'].str.count('-')\n",
    "\n",
    "synteny_condensed['hg_percent']=1-(synteny_condensed['hg_dash_n']/synteny_condensed['total_align_len'])\n",
    "synteny_condensed['mm_percent']=1-(synteny_condensed['mm_dash_n']/synteny_condensed['total_align_len'])\n",
    "\n",
    "\n",
    "print(((synteny_condensed['hg_len'] + synteny_condensed['hg_dash_n']) == synteny_condensed['total_align_len']).all())\n",
    "print(((synteny_condensed['mm_len'] + synteny_condensed['mm_dash_n']) == synteny_condensed['total_align_len']).all())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "a950d778",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.9612903225806452"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.median(synteny_condensed['hg_percent'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "994e441a",
   "metadata": {},
   "outputs": [],
   "source": [
    "chrom_len=pd.read_csv('/pollard/home/shzhang/akita_variant_scoring/data/chrom_lengths_hg38', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "0855f907",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>var_index</th>\n",
       "      <th>mse_mean</th>\n",
       "      <th>corr_mean</th>\n",
       "      <th>oneminus_corrmean</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>5.914621e-03</td>\n",
       "      <td>0.979139</td>\n",
       "      <td>2.086145e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>1.629834e-07</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>4.470926e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>1.445605e-08</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>3.242280e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>2.073410e-08</td>\n",
       "      <td>0.999999</td>\n",
       "      <td>9.900652e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>214039</th>\n",
       "      <td>214039</td>\n",
       "      <td>4.707555e-08</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>9.518438e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>214040</th>\n",
       "      <td>214040</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>214041</th>\n",
       "      <td>214041</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>214042</th>\n",
       "      <td>214042</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>214043</th>\n",
       "      <td>214043</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>214044 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        var_index      mse_mean  corr_mean  oneminus_corrmean\n",
       "0               0           NaN        NaN                NaN\n",
       "1               1  5.914621e-03   0.979139       2.086145e-02\n",
       "2               2  1.629834e-07   1.000000       4.470926e-07\n",
       "3               3  1.445605e-08   1.000000       3.242280e-08\n",
       "4               4  2.073410e-08   0.999999       9.900652e-07\n",
       "...           ...           ...        ...                ...\n",
       "214039     214039  4.707555e-08   1.000000       9.518438e-08\n",
       "214040     214040           NaN        NaN                NaN\n",
       "214041     214041           NaN        NaN                NaN\n",
       "214042     214042           NaN        NaN                NaN\n",
       "214043     214043           NaN        NaN                NaN\n",
       "\n",
       "[214044 rows x 4 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_path='/pollard/data/projects/kgjoni/CBTN_collab/CBTN_data/'\n",
    "scores=pd.read_csv(data_path + 'simple-variants_230620/simple_variants_scored/simple_variants_all_scores', sep='\\t')\n",
    "scores['oneminus_corrmean']=1-scores['corr_mean']\n",
    "scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a7ffcaa6",
   "metadata": {},
   "outputs": [],
   "source": [
    "#get mse and corr idx \n",
    "mse_sort=scores.sort_values(by='mse_mean', ascending=False)\n",
    "top_1000_mse = mse_sort.index[:500]\n",
    "top_1000_mse_idx = mse_sort['var_index'].loc[top_1000_mse].tolist()\n",
    "\n",
    "corr_sort=scores.sort_values(by='oneminus_corrmean', ascending=False)\n",
    "top_1000_corr = corr_sort.index[:500]\n",
    "top_1000_corr_idx = corr_sort['var_index'].loc[top_1000_corr].tolist()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "6c75304d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>overall_index</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>74790</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>129853</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>130218</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>130728</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>130730</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>137</th>\n",
       "      <td>182222</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>138</th>\n",
       "      <td>183088</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139</th>\n",
       "      <td>108979</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>140</th>\n",
       "      <td>109077</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>141</th>\n",
       "      <td>109374</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>142 rows Ã— 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     overall_index\n",
       "0            74790\n",
       "1           129853\n",
       "2           130218\n",
       "3           130728\n",
       "4           130730\n",
       "..             ...\n",
       "137         182222\n",
       "138         183088\n",
       "139         108979\n",
       "140         109077\n",
       "141         109374\n",
       "\n",
       "[142 rows x 1 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "top_variants=pd.read_csv('/pollard/home/shzhang/top_simple_variants', sep='\\t')\n",
    "top_variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fc35b661",
   "metadata": {},
   "outputs": [],
   "source": [
    "idx_union= list(set(top_1000_mse_idx).union(top_1000_corr_idx))\n",
    "len(idx_union)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "0dc61c57",
   "metadata": {},
   "outputs": [],
   "source": [
    "simple_df=pd.read_csv(data_path + 'simple-variants_230620/simple_variants_all', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "e05eff22",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>INFO</th>\n",
       "      <th>FORMAT</th>\n",
       "      <th>BiospecimenID</th>\n",
       "      <th>sample_var_index</th>\n",
       "      <th>overall_index</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2425</th>\n",
       "      <td>chr10</td>\n",
       "      <td>122357800</td>\n",
       "      <td>.</td>\n",
       "      <td>CCA</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>CAL=Strelka2,VarDict;CSQ=-|intergenic_variant|...</td>\n",
       "      <td>GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS</td>\n",
       "      <td>BS_19EJ85F8</td>\n",
       "      <td>1452</td>\n",
       "      <td>2425</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2815</th>\n",
       "      <td>chr15</td>\n",
       "      <td>28432799</td>\n",
       "      <td>.</td>\n",
       "      <td>CAT</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>CAL=Strelka2,VarDict;CSQ=-|intergenic_variant|...</td>\n",
       "      <td>GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS</td>\n",
       "      <td>BS_19EJ85F8</td>\n",
       "      <td>1842</td>\n",
       "      <td>2815</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4993</th>\n",
       "      <td>chr2</td>\n",
       "      <td>32439175</td>\n",
       "      <td>.</td>\n",
       "      <td>TTG</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>CAL=Strelka2,VarDict;CSQ=-|intron_variant|MODI...</td>\n",
       "      <td>GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS</td>\n",
       "      <td>BS_A94A7EJ4</td>\n",
       "      <td>1501</td>\n",
       "      <td>4993</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6422</th>\n",
       "      <td>chr3</td>\n",
       "      <td>17630126</td>\n",
       "      <td>.</td>\n",
       "      <td>AAGAGG</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>CAL=Strelka2,Mutect2,VarDict,Lancet;CSQ=-|intr...</td>\n",
       "      <td>GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS</td>\n",
       "      <td>BS_A94A7EJ4</td>\n",
       "      <td>2930</td>\n",
       "      <td>6422</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10525</th>\n",
       "      <td>chr7</td>\n",
       "      <td>158180223</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>CAT</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>CAL=Strelka2,Mutect2,Lancet;CSQ=AT|intron_vari...</td>\n",
       "      <td>GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS</td>\n",
       "      <td>BS_A94A7EJ4</td>\n",
       "      <td>7033</td>\n",
       "      <td>10525</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>206438</th>\n",
       "      <td>chr9</td>\n",
       "      <td>88136504</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>CTT</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>CAL=Strelka2,VarDict;CSQ=TT|intron_variant|MOD...</td>\n",
       "      <td>GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS</td>\n",
       "      <td>BS_G6V64VWY</td>\n",
       "      <td>1338</td>\n",
       "      <td>206438</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>210546</th>\n",
       "      <td>chr10</td>\n",
       "      <td>46535047</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>GCC</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>CAL=Strelka2,Mutect2,VarDict;CSQ=CC|intron_var...</td>\n",
       "      <td>GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS</td>\n",
       "      <td>BS_QR84QJN4</td>\n",
       "      <td>3072</td>\n",
       "      <td>210546</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>211562</th>\n",
       "      <td>chr15</td>\n",
       "      <td>28646249</td>\n",
       "      <td>.</td>\n",
       "      <td>ACTTTTC</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>CAL=Mutect2,VarDict;CSQ=-|downstream_gene_vari...</td>\n",
       "      <td>GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS</td>\n",
       "      <td>BS_QR84QJN4</td>\n",
       "      <td>4088</td>\n",
       "      <td>211562</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>212556</th>\n",
       "      <td>chrX</td>\n",
       "      <td>85899908</td>\n",
       "      <td>.</td>\n",
       "      <td>TAACAAGTGTTGATGAGGATACAGAAAAAAGCAGACCCTTTACACT...</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>CAL=Mutect2,VarDict;CSQ=-|intron_variant|MODIF...</td>\n",
       "      <td>GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS</td>\n",
       "      <td>BS_QR84QJN4</td>\n",
       "      <td>5082</td>\n",
       "      <td>212556</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>212776</th>\n",
       "      <td>chr1</td>\n",
       "      <td>154423260</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>AATATATATATATATATATATATATAT</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>CAL=Mutect2,VarDict;CSQ=ATATATATATATATATATATAT...</td>\n",
       "      <td>GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS</td>\n",
       "      <td>BS_A86314EJ</td>\n",
       "      <td>83</td>\n",
       "      <td>212776</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>142 rows Ã— 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        CHROM        POS ID  \\\n",
       "2425    chr10  122357800  .   \n",
       "2815    chr15   28432799  .   \n",
       "4993     chr2   32439175  .   \n",
       "6422     chr3   17630126  .   \n",
       "10525    chr7  158180223  .   \n",
       "...       ...        ... ..   \n",
       "206438   chr9   88136504  .   \n",
       "210546  chr10   46535047  .   \n",
       "211562  chr15   28646249  .   \n",
       "212556   chrX   85899908  .   \n",
       "212776   chr1  154423260  .   \n",
       "\n",
       "                                                      REF  \\\n",
       "2425                                                  CCA   \n",
       "2815                                                  CAT   \n",
       "4993                                                  TTG   \n",
       "6422                                               AAGAGG   \n",
       "10525                                                   C   \n",
       "...                                                   ...   \n",
       "206438                                                  C   \n",
       "210546                                                  G   \n",
       "211562                                            ACTTTTC   \n",
       "212556  TAACAAGTGTTGATGAGGATACAGAAAAAAGCAGACCCTTTACACT...   \n",
       "212776                                                  A   \n",
       "\n",
       "                                ALT QUAL FILTER  \\\n",
       "2425                              C    .   PASS   \n",
       "2815                              C    .   PASS   \n",
       "4993                              T    .   PASS   \n",
       "6422                              A    .   PASS   \n",
       "10525                           CAT    .   PASS   \n",
       "...                             ...  ...    ...   \n",
       "206438                          CTT    .   PASS   \n",
       "210546                          GCC    .   PASS   \n",
       "211562                            A    .   PASS   \n",
       "212556                            T    .   PASS   \n",
       "212776  AATATATATATATATATATATATATAT    .   PASS   \n",
       "\n",
       "                                                     INFO  \\\n",
       "2425    CAL=Strelka2,VarDict;CSQ=-|intergenic_variant|...   \n",
       "2815    CAL=Strelka2,VarDict;CSQ=-|intergenic_variant|...   \n",
       "4993    CAL=Strelka2,VarDict;CSQ=-|intron_variant|MODI...   \n",
       "6422    CAL=Strelka2,Mutect2,VarDict,Lancet;CSQ=-|intr...   \n",
       "10525   CAL=Strelka2,Mutect2,Lancet;CSQ=AT|intron_vari...   \n",
       "...                                                   ...   \n",
       "206438  CAL=Strelka2,VarDict;CSQ=TT|intron_variant|MOD...   \n",
       "210546  CAL=Strelka2,Mutect2,VarDict;CSQ=CC|intron_var...   \n",
       "211562  CAL=Mutect2,VarDict;CSQ=-|downstream_gene_vari...   \n",
       "212556  CAL=Mutect2,VarDict;CSQ=-|intron_variant|MODIF...   \n",
       "212776  CAL=Mutect2,VarDict;CSQ=ATATATATATATATATATATAT...   \n",
       "\n",
       "                                                   FORMAT BiospecimenID  \\\n",
       "2425    GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS   BS_19EJ85F8   \n",
       "2815    GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS   BS_19EJ85F8   \n",
       "4993    GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS   BS_A94A7EJ4   \n",
       "6422    GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS   BS_A94A7EJ4   \n",
       "10525   GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS   BS_A94A7EJ4   \n",
       "...                                                   ...           ...   \n",
       "206438  GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS   BS_G6V64VWY   \n",
       "210546  GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS   BS_QR84QJN4   \n",
       "211562  GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS   BS_QR84QJN4   \n",
       "212556  GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS   BS_QR84QJN4   \n",
       "212776  GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS   BS_A86314EJ   \n",
       "\n",
       "        sample_var_index  overall_index  \n",
       "2425                1452           2425  \n",
       "2815                1842           2815  \n",
       "4993                1501           4993  \n",
       "6422                2930           6422  \n",
       "10525               7033          10525  \n",
       "...                  ...            ...  \n",
       "206438              1338         206438  \n",
       "210546              3072         210546  \n",
       "211562              4088         211562  \n",
       "212556              5082         212556  \n",
       "212776                83         212776  \n",
       "\n",
       "[142 rows x 12 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#keep only those in top mse and corr\n",
    "#simple_top = simple_df[simple_df['sample_var_index'].isin(idx_union)]\n",
    "simple_top=simple_df[simple_df.index.isin(top_variants['overall_index'])]\n",
    "simple_top['overall_index']=simple_top.index\n",
    "simple_top"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "11fa1788",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>INFO</th>\n",
       "      <th>FORMAT</th>\n",
       "      <th>BiospecimenID</th>\n",
       "      <th>sample_var_index</th>\n",
       "      <th>overall_index</th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr10</td>\n",
       "      <td>122357800</td>\n",
       "      <td>.</td>\n",
       "      <td>CCA</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>CAL=Strelka2,VarDict;CSQ=-|intergenic_variant|...</td>\n",
       "      <td>GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS</td>\n",
       "      <td>BS_19EJ85F8</td>\n",
       "      <td>1452</td>\n",
       "      <td>2425</td>\n",
       "      <td>chr10</td>\n",
       "      <td>133797422</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr15</td>\n",
       "      <td>28432799</td>\n",
       "      <td>.</td>\n",
       "      <td>CAT</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>CAL=Strelka2,VarDict;CSQ=-|intergenic_variant|...</td>\n",
       "      <td>GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS</td>\n",
       "      <td>BS_19EJ85F8</td>\n",
       "      <td>1842</td>\n",
       "      <td>2815</td>\n",
       "      <td>chr15</td>\n",
       "      <td>101991189</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr2</td>\n",
       "      <td>32439175</td>\n",
       "      <td>.</td>\n",
       "      <td>TTG</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>CAL=Strelka2,VarDict;CSQ=-|intron_variant|MODI...</td>\n",
       "      <td>GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS</td>\n",
       "      <td>BS_A94A7EJ4</td>\n",
       "      <td>1501</td>\n",
       "      <td>4993</td>\n",
       "      <td>chr2</td>\n",
       "      <td>242193529</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr3</td>\n",
       "      <td>17630126</td>\n",
       "      <td>.</td>\n",
       "      <td>AAGAGG</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>CAL=Strelka2,Mutect2,VarDict,Lancet;CSQ=-|intr...</td>\n",
       "      <td>GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS</td>\n",
       "      <td>BS_A94A7EJ4</td>\n",
       "      <td>2930</td>\n",
       "      <td>6422</td>\n",
       "      <td>chr3</td>\n",
       "      <td>198295559</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr7</td>\n",
       "      <td>158180223</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>CAT</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>CAL=Strelka2,Mutect2,Lancet;CSQ=AT|intron_vari...</td>\n",
       "      <td>GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS</td>\n",
       "      <td>BS_A94A7EJ4</td>\n",
       "      <td>7033</td>\n",
       "      <td>10525</td>\n",
       "      <td>chr7</td>\n",
       "      <td>159345973</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>137</th>\n",
       "      <td>chr9</td>\n",
       "      <td>88136504</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>CTT</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>CAL=Strelka2,VarDict;CSQ=TT|intron_variant|MOD...</td>\n",
       "      <td>GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS</td>\n",
       "      <td>BS_G6V64VWY</td>\n",
       "      <td>1338</td>\n",
       "      <td>206438</td>\n",
       "      <td>chr9</td>\n",
       "      <td>138394717</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>138</th>\n",
       "      <td>chr10</td>\n",
       "      <td>46535047</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>GCC</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>CAL=Strelka2,Mutect2,VarDict;CSQ=CC|intron_var...</td>\n",
       "      <td>GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS</td>\n",
       "      <td>BS_QR84QJN4</td>\n",
       "      <td>3072</td>\n",
       "      <td>210546</td>\n",
       "      <td>chr10</td>\n",
       "      <td>133797422</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139</th>\n",
       "      <td>chr15</td>\n",
       "      <td>28646249</td>\n",
       "      <td>.</td>\n",
       "      <td>ACTTTTC</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>CAL=Mutect2,VarDict;CSQ=-|downstream_gene_vari...</td>\n",
       "      <td>GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS</td>\n",
       "      <td>BS_QR84QJN4</td>\n",
       "      <td>4088</td>\n",
       "      <td>211562</td>\n",
       "      <td>chr15</td>\n",
       "      <td>101991189</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>140</th>\n",
       "      <td>chrX</td>\n",
       "      <td>85899908</td>\n",
       "      <td>.</td>\n",
       "      <td>TAACAAGTGTTGATGAGGATACAGAAAAAAGCAGACCCTTTACACT...</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>CAL=Mutect2,VarDict;CSQ=-|intron_variant|MODIF...</td>\n",
       "      <td>GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS</td>\n",
       "      <td>BS_QR84QJN4</td>\n",
       "      <td>5082</td>\n",
       "      <td>212556</td>\n",
       "      <td>chrX</td>\n",
       "      <td>156040895</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>141</th>\n",
       "      <td>chr1</td>\n",
       "      <td>154423260</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>AATATATATATATATATATATATATAT</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>CAL=Mutect2,VarDict;CSQ=ATATATATATATATATATATAT...</td>\n",
       "      <td>GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS</td>\n",
       "      <td>BS_A86314EJ</td>\n",
       "      <td>83</td>\n",
       "      <td>212776</td>\n",
       "      <td>chr1</td>\n",
       "      <td>248956422</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>142 rows Ã— 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     CHROM        POS ID                                                REF  \\\n",
       "0    chr10  122357800  .                                                CCA   \n",
       "1    chr15   28432799  .                                                CAT   \n",
       "2     chr2   32439175  .                                                TTG   \n",
       "3     chr3   17630126  .                                             AAGAGG   \n",
       "4     chr7  158180223  .                                                  C   \n",
       "..     ...        ... ..                                                ...   \n",
       "137   chr9   88136504  .                                                  C   \n",
       "138  chr10   46535047  .                                                  G   \n",
       "139  chr15   28646249  .                                            ACTTTTC   \n",
       "140   chrX   85899908  .  TAACAAGTGTTGATGAGGATACAGAAAAAAGCAGACCCTTTACACT...   \n",
       "141   chr1  154423260  .                                                  A   \n",
       "\n",
       "                             ALT QUAL FILTER  \\\n",
       "0                              C    .   PASS   \n",
       "1                              C    .   PASS   \n",
       "2                              T    .   PASS   \n",
       "3                              A    .   PASS   \n",
       "4                            CAT    .   PASS   \n",
       "..                           ...  ...    ...   \n",
       "137                          CTT    .   PASS   \n",
       "138                          GCC    .   PASS   \n",
       "139                            A    .   PASS   \n",
       "140                            T    .   PASS   \n",
       "141  AATATATATATATATATATATATATAT    .   PASS   \n",
       "\n",
       "                                                  INFO  \\\n",
       "0    CAL=Strelka2,VarDict;CSQ=-|intergenic_variant|...   \n",
       "1    CAL=Strelka2,VarDict;CSQ=-|intergenic_variant|...   \n",
       "2    CAL=Strelka2,VarDict;CSQ=-|intron_variant|MODI...   \n",
       "3    CAL=Strelka2,Mutect2,VarDict,Lancet;CSQ=-|intr...   \n",
       "4    CAL=Strelka2,Mutect2,Lancet;CSQ=AT|intron_vari...   \n",
       "..                                                 ...   \n",
       "137  CAL=Strelka2,VarDict;CSQ=TT|intron_variant|MOD...   \n",
       "138  CAL=Strelka2,Mutect2,VarDict;CSQ=CC|intron_var...   \n",
       "139  CAL=Mutect2,VarDict;CSQ=-|downstream_gene_vari...   \n",
       "140  CAL=Mutect2,VarDict;CSQ=-|intron_variant|MODIF...   \n",
       "141  CAL=Mutect2,VarDict;CSQ=ATATATATATATATATATATAT...   \n",
       "\n",
       "                                                FORMAT BiospecimenID  \\\n",
       "0    GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS   BS_19EJ85F8   \n",
       "1    GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS   BS_19EJ85F8   \n",
       "2    GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS   BS_A94A7EJ4   \n",
       "3    GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS   BS_A94A7EJ4   \n",
       "4    GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS   BS_A94A7EJ4   \n",
       "..                                                 ...           ...   \n",
       "137  GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS   BS_G6V64VWY   \n",
       "138  GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS   BS_QR84QJN4   \n",
       "139  GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS   BS_QR84QJN4   \n",
       "140  GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS   BS_QR84QJN4   \n",
       "141  GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS   BS_A86314EJ   \n",
       "\n",
       "     sample_var_index  overall_index      0          1  \n",
       "0                1452           2425  chr10  133797422  \n",
       "1                1842           2815  chr15  101991189  \n",
       "2                1501           4993   chr2  242193529  \n",
       "3                2930           6422   chr3  198295559  \n",
       "4                7033          10525   chr7  159345973  \n",
       "..                ...            ...    ...        ...  \n",
       "137              1338         206438   chr9  138394717  \n",
       "138              3072         210546  chr10  133797422  \n",
       "139              4088         211562  chr15  101991189  \n",
       "140              5082         212556   chrX  156040895  \n",
       "141                83         212776   chr1  248956422  \n",
       "\n",
       "[142 rows x 14 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#merge with chr len information\n",
    "chrom_len=pd.read_csv('/pollard/home/shzhang/akita_variant_scoring/data/chrom_lengths_hg38', sep='\\t', header=None)\n",
    "chrom_len[0]='chr' + chrom_len[0]\n",
    "simple=simple_top.merge(chrom_len, right_on=0, left_on='CHROM', how='left')\n",
    "simple = simple.reset_index(drop=True)\n",
    "\n",
    "simple"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "2b27b213",
   "metadata": {},
   "outputs": [],
   "source": [
    "simple = simple.rename(columns={1: 'chr_len'})\n",
    "\n",
    "\n",
    "simple['POS_300KB']=np.floor(np.maximum(0, simple['POS']-150e3))\n",
    "simple['END_300KB']=np.floor(np.minimum(simple['chr_len'], simple['POS']+150e3))\n",
    "simple['POS_1MB']=np.floor(np.maximum(0, simple['POS']-5e5))\n",
    "simple['END_1MB']=np.floor(np.minimum(simple['chr_len'], simple['POS']+ 5e5))\n",
    "\n",
    "\n",
    "simple['postoend_1mb']=simple['END_1MB']-simple['POS_1MB']\n",
    "simple['postoend_300kb']=simple['END_300KB']-simple['POS_300KB']\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "8e7716d3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>INFO</th>\n",
       "      <th>FORMAT</th>\n",
       "      <th>BiospecimenID</th>\n",
       "      <th>sample_var_index</th>\n",
       "      <th>overall_index</th>\n",
       "      <th>0</th>\n",
       "      <th>chr_len</th>\n",
       "      <th>POS_300KB</th>\n",
       "      <th>END_300KB</th>\n",
       "      <th>POS_1MB</th>\n",
       "      <th>END_1MB</th>\n",
       "      <th>postoend_1mb</th>\n",
       "      <th>postoend_300kb</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr10</td>\n",
       "      <td>122357800</td>\n",
       "      <td>.</td>\n",
       "      <td>CCA</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>CAL=Strelka2,VarDict;CSQ=-|intergenic_variant|...</td>\n",
       "      <td>GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS</td>\n",
       "      <td>BS_19EJ85F8</td>\n",
       "      <td>1452</td>\n",
       "      <td>2425</td>\n",
       "      <td>chr10</td>\n",
       "      <td>133797422</td>\n",
       "      <td>122207800.0</td>\n",
       "      <td>122507800.0</td>\n",
       "      <td>121857800.0</td>\n",
       "      <td>122857800.0</td>\n",
       "      <td>1000000.0</td>\n",
       "      <td>300000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr15</td>\n",
       "      <td>28432799</td>\n",
       "      <td>.</td>\n",
       "      <td>CAT</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>CAL=Strelka2,VarDict;CSQ=-|intergenic_variant|...</td>\n",
       "      <td>GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS</td>\n",
       "      <td>BS_19EJ85F8</td>\n",
       "      <td>1842</td>\n",
       "      <td>2815</td>\n",
       "      <td>chr15</td>\n",
       "      <td>101991189</td>\n",
       "      <td>28282799.0</td>\n",
       "      <td>28582799.0</td>\n",
       "      <td>27932799.0</td>\n",
       "      <td>28932799.0</td>\n",
       "      <td>1000000.0</td>\n",
       "      <td>300000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr2</td>\n",
       "      <td>32439175</td>\n",
       "      <td>.</td>\n",
       "      <td>TTG</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>CAL=Strelka2,VarDict;CSQ=-|intron_variant|MODI...</td>\n",
       "      <td>GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS</td>\n",
       "      <td>BS_A94A7EJ4</td>\n",
       "      <td>1501</td>\n",
       "      <td>4993</td>\n",
       "      <td>chr2</td>\n",
       "      <td>242193529</td>\n",
       "      <td>32289175.0</td>\n",
       "      <td>32589175.0</td>\n",
       "      <td>31939175.0</td>\n",
       "      <td>32939175.0</td>\n",
       "      <td>1000000.0</td>\n",
       "      <td>300000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr3</td>\n",
       "      <td>17630126</td>\n",
       "      <td>.</td>\n",
       "      <td>AAGAGG</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>CAL=Strelka2,Mutect2,VarDict,Lancet;CSQ=-|intr...</td>\n",
       "      <td>GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS</td>\n",
       "      <td>BS_A94A7EJ4</td>\n",
       "      <td>2930</td>\n",
       "      <td>6422</td>\n",
       "      <td>chr3</td>\n",
       "      <td>198295559</td>\n",
       "      <td>17480126.0</td>\n",
       "      <td>17780126.0</td>\n",
       "      <td>17130126.0</td>\n",
       "      <td>18130126.0</td>\n",
       "      <td>1000000.0</td>\n",
       "      <td>300000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr7</td>\n",
       "      <td>158180223</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>CAT</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>CAL=Strelka2,Mutect2,Lancet;CSQ=AT|intron_vari...</td>\n",
       "      <td>GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS</td>\n",
       "      <td>BS_A94A7EJ4</td>\n",
       "      <td>7033</td>\n",
       "      <td>10525</td>\n",
       "      <td>chr7</td>\n",
       "      <td>159345973</td>\n",
       "      <td>158030223.0</td>\n",
       "      <td>158330223.0</td>\n",
       "      <td>157680223.0</td>\n",
       "      <td>158680223.0</td>\n",
       "      <td>1000000.0</td>\n",
       "      <td>300000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>137</th>\n",
       "      <td>chr9</td>\n",
       "      <td>88136504</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>CTT</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>CAL=Strelka2,VarDict;CSQ=TT|intron_variant|MOD...</td>\n",
       "      <td>GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS</td>\n",
       "      <td>BS_G6V64VWY</td>\n",
       "      <td>1338</td>\n",
       "      <td>206438</td>\n",
       "      <td>chr9</td>\n",
       "      <td>138394717</td>\n",
       "      <td>87986504.0</td>\n",
       "      <td>88286504.0</td>\n",
       "      <td>87636504.0</td>\n",
       "      <td>88636504.0</td>\n",
       "      <td>1000000.0</td>\n",
       "      <td>300000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>138</th>\n",
       "      <td>chr10</td>\n",
       "      <td>46535047</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>GCC</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>CAL=Strelka2,Mutect2,VarDict;CSQ=CC|intron_var...</td>\n",
       "      <td>GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS</td>\n",
       "      <td>BS_QR84QJN4</td>\n",
       "      <td>3072</td>\n",
       "      <td>210546</td>\n",
       "      <td>chr10</td>\n",
       "      <td>133797422</td>\n",
       "      <td>46385047.0</td>\n",
       "      <td>46685047.0</td>\n",
       "      <td>46035047.0</td>\n",
       "      <td>47035047.0</td>\n",
       "      <td>1000000.0</td>\n",
       "      <td>300000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139</th>\n",
       "      <td>chr15</td>\n",
       "      <td>28646249</td>\n",
       "      <td>.</td>\n",
       "      <td>ACTTTTC</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>CAL=Mutect2,VarDict;CSQ=-|downstream_gene_vari...</td>\n",
       "      <td>GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS</td>\n",
       "      <td>BS_QR84QJN4</td>\n",
       "      <td>4088</td>\n",
       "      <td>211562</td>\n",
       "      <td>chr15</td>\n",
       "      <td>101991189</td>\n",
       "      <td>28496249.0</td>\n",
       "      <td>28796249.0</td>\n",
       "      <td>28146249.0</td>\n",
       "      <td>29146249.0</td>\n",
       "      <td>1000000.0</td>\n",
       "      <td>300000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>140</th>\n",
       "      <td>chrX</td>\n",
       "      <td>85899908</td>\n",
       "      <td>.</td>\n",
       "      <td>TAACAAGTGTTGATGAGGATACAGAAAAAAGCAGACCCTTTACACT...</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>CAL=Mutect2,VarDict;CSQ=-|intron_variant|MODIF...</td>\n",
       "      <td>GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS</td>\n",
       "      <td>BS_QR84QJN4</td>\n",
       "      <td>5082</td>\n",
       "      <td>212556</td>\n",
       "      <td>chrX</td>\n",
       "      <td>156040895</td>\n",
       "      <td>85749908.0</td>\n",
       "      <td>86049908.0</td>\n",
       "      <td>85399908.0</td>\n",
       "      <td>86399908.0</td>\n",
       "      <td>1000000.0</td>\n",
       "      <td>300000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>141</th>\n",
       "      <td>chr1</td>\n",
       "      <td>154423260</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>AATATATATATATATATATATATATAT</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>CAL=Mutect2,VarDict;CSQ=ATATATATATATATATATATAT...</td>\n",
       "      <td>GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS</td>\n",
       "      <td>BS_A86314EJ</td>\n",
       "      <td>83</td>\n",
       "      <td>212776</td>\n",
       "      <td>chr1</td>\n",
       "      <td>248956422</td>\n",
       "      <td>154273260.0</td>\n",
       "      <td>154573260.0</td>\n",
       "      <td>153923260.0</td>\n",
       "      <td>154923260.0</td>\n",
       "      <td>1000000.0</td>\n",
       "      <td>300000.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>142 rows Ã— 20 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     CHROM        POS ID                                                REF  \\\n",
       "0    chr10  122357800  .                                                CCA   \n",
       "1    chr15   28432799  .                                                CAT   \n",
       "2     chr2   32439175  .                                                TTG   \n",
       "3     chr3   17630126  .                                             AAGAGG   \n",
       "4     chr7  158180223  .                                                  C   \n",
       "..     ...        ... ..                                                ...   \n",
       "137   chr9   88136504  .                                                  C   \n",
       "138  chr10   46535047  .                                                  G   \n",
       "139  chr15   28646249  .                                            ACTTTTC   \n",
       "140   chrX   85899908  .  TAACAAGTGTTGATGAGGATACAGAAAAAAGCAGACCCTTTACACT...   \n",
       "141   chr1  154423260  .                                                  A   \n",
       "\n",
       "                             ALT QUAL FILTER  \\\n",
       "0                              C    .   PASS   \n",
       "1                              C    .   PASS   \n",
       "2                              T    .   PASS   \n",
       "3                              A    .   PASS   \n",
       "4                            CAT    .   PASS   \n",
       "..                           ...  ...    ...   \n",
       "137                          CTT    .   PASS   \n",
       "138                          GCC    .   PASS   \n",
       "139                            A    .   PASS   \n",
       "140                            T    .   PASS   \n",
       "141  AATATATATATATATATATATATATAT    .   PASS   \n",
       "\n",
       "                                                  INFO  \\\n",
       "0    CAL=Strelka2,VarDict;CSQ=-|intergenic_variant|...   \n",
       "1    CAL=Strelka2,VarDict;CSQ=-|intergenic_variant|...   \n",
       "2    CAL=Strelka2,VarDict;CSQ=-|intron_variant|MODI...   \n",
       "3    CAL=Strelka2,Mutect2,VarDict,Lancet;CSQ=-|intr...   \n",
       "4    CAL=Strelka2,Mutect2,Lancet;CSQ=AT|intron_vari...   \n",
       "..                                                 ...   \n",
       "137  CAL=Strelka2,VarDict;CSQ=TT|intron_variant|MOD...   \n",
       "138  CAL=Strelka2,Mutect2,VarDict;CSQ=CC|intron_var...   \n",
       "139  CAL=Mutect2,VarDict;CSQ=-|downstream_gene_vari...   \n",
       "140  CAL=Mutect2,VarDict;CSQ=-|intron_variant|MODIF...   \n",
       "141  CAL=Mutect2,VarDict;CSQ=ATATATATATATATATATATAT...   \n",
       "\n",
       "                                                FORMAT BiospecimenID  \\\n",
       "0    GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS   BS_19EJ85F8   \n",
       "1    GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS   BS_19EJ85F8   \n",
       "2    GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS   BS_A94A7EJ4   \n",
       "3    GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS   BS_A94A7EJ4   \n",
       "4    GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS   BS_A94A7EJ4   \n",
       "..                                                 ...           ...   \n",
       "137  GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS   BS_G6V64VWY   \n",
       "138  GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS   BS_QR84QJN4   \n",
       "139  GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS   BS_QR84QJN4   \n",
       "140  GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS   BS_QR84QJN4   \n",
       "141  GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS   BS_A86314EJ   \n",
       "\n",
       "     sample_var_index  overall_index      0    chr_len    POS_300KB  \\\n",
       "0                1452           2425  chr10  133797422  122207800.0   \n",
       "1                1842           2815  chr15  101991189   28282799.0   \n",
       "2                1501           4993   chr2  242193529   32289175.0   \n",
       "3                2930           6422   chr3  198295559   17480126.0   \n",
       "4                7033          10525   chr7  159345973  158030223.0   \n",
       "..                ...            ...    ...        ...          ...   \n",
       "137              1338         206438   chr9  138394717   87986504.0   \n",
       "138              3072         210546  chr10  133797422   46385047.0   \n",
       "139              4088         211562  chr15  101991189   28496249.0   \n",
       "140              5082         212556   chrX  156040895   85749908.0   \n",
       "141                83         212776   chr1  248956422  154273260.0   \n",
       "\n",
       "       END_300KB      POS_1MB      END_1MB  postoend_1mb  postoend_300kb  \n",
       "0    122507800.0  121857800.0  122857800.0     1000000.0        300000.0  \n",
       "1     28582799.0   27932799.0   28932799.0     1000000.0        300000.0  \n",
       "2     32589175.0   31939175.0   32939175.0     1000000.0        300000.0  \n",
       "3     17780126.0   17130126.0   18130126.0     1000000.0        300000.0  \n",
       "4    158330223.0  157680223.0  158680223.0     1000000.0        300000.0  \n",
       "..           ...          ...          ...           ...             ...  \n",
       "137   88286504.0   87636504.0   88636504.0     1000000.0        300000.0  \n",
       "138   46685047.0   46035047.0   47035047.0     1000000.0        300000.0  \n",
       "139   28796249.0   28146249.0   29146249.0     1000000.0        300000.0  \n",
       "140   86049908.0   85399908.0   86399908.0     1000000.0        300000.0  \n",
       "141  154573260.0  153923260.0  154923260.0     1000000.0        300000.0  \n",
       "\n",
       "[142 rows x 20 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "simple\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "89c1160a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "142\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS_300KB</th>\n",
       "      <th>END_300KB</th>\n",
       "      <th>overall_index</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr10</td>\n",
       "      <td>122207800</td>\n",
       "      <td>122507800</td>\n",
       "      <td>2425</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr15</td>\n",
       "      <td>28282799</td>\n",
       "      <td>28582799</td>\n",
       "      <td>2815</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr2</td>\n",
       "      <td>32289175</td>\n",
       "      <td>32589175</td>\n",
       "      <td>4993</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr3</td>\n",
       "      <td>17480126</td>\n",
       "      <td>17780126</td>\n",
       "      <td>6422</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr7</td>\n",
       "      <td>158030223</td>\n",
       "      <td>158330223</td>\n",
       "      <td>10525</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   CHROM  POS_300KB  END_300KB  overall_index\n",
       "0  chr10  122207800  122507800           2425\n",
       "1  chr15   28282799   28582799           2815\n",
       "2   chr2   32289175   32589175           4993\n",
       "3   chr3   17480126   17780126           6422\n",
       "4   chr7  158030223  158330223          10525"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "variants_300kb=simple\n",
    "variants_300kb=variants_300kb[['CHROM', 'POS_300KB', 'END_300KB', 'overall_index']]\n",
    "variants_300kb['POS_300KB']=variants_300kb['POS_300KB'].astype(int)\n",
    "variants_300kb['END_300KB']=variants_300kb['END_300KB'].astype(int)\n",
    "print(len(variants_300kb))\n",
    "variants_300kb.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "86ca1360",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "142\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS_1MB</th>\n",
       "      <th>END_1MB</th>\n",
       "      <th>overall_index</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr10</td>\n",
       "      <td>121857800</td>\n",
       "      <td>122857800</td>\n",
       "      <td>2425</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr15</td>\n",
       "      <td>27932799</td>\n",
       "      <td>28932799</td>\n",
       "      <td>2815</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr2</td>\n",
       "      <td>31939175</td>\n",
       "      <td>32939175</td>\n",
       "      <td>4993</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr3</td>\n",
       "      <td>17130126</td>\n",
       "      <td>18130126</td>\n",
       "      <td>6422</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr7</td>\n",
       "      <td>157680223</td>\n",
       "      <td>158680223</td>\n",
       "      <td>10525</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   CHROM    POS_1MB    END_1MB  overall_index\n",
       "0  chr10  121857800  122857800           2425\n",
       "1  chr15   27932799   28932799           2815\n",
       "2   chr2   31939175   32939175           4993\n",
       "3   chr3   17130126   18130126           6422\n",
       "4   chr7  157680223  158680223          10525"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "variants_1mb=simple\n",
    "variants_1mb=variants_1mb[['CHROM', 'POS_1MB', 'END_1MB', 'overall_index']]\n",
    "variants_1mb['POS_1MB']=variants_1mb['POS_1MB'].astype(int)\n",
    "variants_1mb['END_1MB']=variants_1mb['END_1MB'].astype(int)\n",
    "print(len(variants_1mb))\n",
    "variants_1mb.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "79451772",
   "metadata": {},
   "outputs": [],
   "source": [
    "variant_300kb_BED = BedTool.from_dataframe(variants_300kb)\n",
    "variant_1mb_BED = BedTool.from_dataframe(variants_1mb)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "02948f8c",
   "metadata": {},
   "outputs": [],
   "source": [
    "synteny_BED = BedTool.from_dataframe(synteny_condensed)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "33566315",
   "metadata": {},
   "outputs": [],
   "source": [
    "#reports only SVs where there is overlap with  gene\n",
    "#-F ensures that at least 0.5 of gene has overlap with SV \n",
    "#use loj in this case, because many SVs overlap multiple syntenic regions  \n",
    "variant_300kb_synt=variant_300kb_BED.intersect(synteny_BED, loj=True, wo=True).to_dataframe(header=None)\n",
    "variant_1mb_synt=variant_1mb_BED.intersect(synteny_BED, loj=True, wo=True).to_dataframe(header=None)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "49681822",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "16547\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS_300KB</th>\n",
       "      <th>END_300KB</th>\n",
       "      <th>overall_index</th>\n",
       "      <th>hg_chr</th>\n",
       "      <th>hg_start</th>\n",
       "      <th>hg_end</th>\n",
       "      <th>hg_len</th>\n",
       "      <th>hg_seq</th>\n",
       "      <th>mm_chr</th>\n",
       "      <th>mm_start</th>\n",
       "      <th>mm_end</th>\n",
       "      <th>mm_len</th>\n",
       "      <th>mm_seq</th>\n",
       "      <th>total_align_len</th>\n",
       "      <th>hg_dash_n</th>\n",
       "      <th>mm_dash_n</th>\n",
       "      <th>hg_percent</th>\n",
       "      <th>mm_percent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr10</td>\n",
       "      <td>122207800</td>\n",
       "      <td>122507800</td>\n",
       "      <td>2425</td>\n",
       "      <td>chr10</td>\n",
       "      <td>122207147</td>\n",
       "      <td>122208150</td>\n",
       "      <td>1004</td>\n",
       "      <td>aattagccaggcgtggtggcacacacctgtaatcccagctactcgg...</td>\n",
       "      <td>chr7</td>\n",
       "      <td>130725901</td>\n",
       "      <td>130726680</td>\n",
       "      <td>780</td>\n",
       "      <td>AACTAGATGGGCATGGTTGC---------------------CTTG-...</td>\n",
       "      <td>1032</td>\n",
       "      <td>28</td>\n",
       "      <td>252</td>\n",
       "      <td>0.9728682170542635</td>\n",
       "      <td>0.7558139534883721</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr10</td>\n",
       "      <td>122207800</td>\n",
       "      <td>122507800</td>\n",
       "      <td>2425</td>\n",
       "      <td>chr10</td>\n",
       "      <td>122209989</td>\n",
       "      <td>122212651</td>\n",
       "      <td>2663</td>\n",
       "      <td>ccctggtggtttattcgctatcttcccctgaagaatgtgacttcca...</td>\n",
       "      <td>chr7</td>\n",
       "      <td>130728155</td>\n",
       "      <td>130730902</td>\n",
       "      <td>2748</td>\n",
       "      <td>CTTTGGTGGC-------------------AATGACAATGGCTGCTA...</td>\n",
       "      <td>2902</td>\n",
       "      <td>239</td>\n",
       "      <td>154</td>\n",
       "      <td>0.9176430048242591</td>\n",
       "      <td>0.9469331495520331</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr10</td>\n",
       "      <td>122207800</td>\n",
       "      <td>122507800</td>\n",
       "      <td>2425</td>\n",
       "      <td>chr10</td>\n",
       "      <td>122212878</td>\n",
       "      <td>122214639</td>\n",
       "      <td>1762</td>\n",
       "      <td>AACTTGATTTCAAGCCTGGAATAGGAGAAGTCTGCAATGGCTAGAC...</td>\n",
       "      <td>chr7</td>\n",
       "      <td>130731103</td>\n",
       "      <td>130732753</td>\n",
       "      <td>1651</td>\n",
       "      <td>AACTTTATTTCAAAAGTGGAATAAT---ACTCACAAATGTCTACCC...</td>\n",
       "      <td>1833</td>\n",
       "      <td>71</td>\n",
       "      <td>182</td>\n",
       "      <td>0.96126568466994</td>\n",
       "      <td>0.900709219858156</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr10</td>\n",
       "      <td>122207800</td>\n",
       "      <td>122507800</td>\n",
       "      <td>2425</td>\n",
       "      <td>chr10</td>\n",
       "      <td>122214713</td>\n",
       "      <td>122215469</td>\n",
       "      <td>757</td>\n",
       "      <td>CAGCCCCGCTA-TTCATATGGAGG----------AGAAAAAGGGAG...</td>\n",
       "      <td>chr7</td>\n",
       "      <td>130732889</td>\n",
       "      <td>130733584</td>\n",
       "      <td>696</td>\n",
       "      <td>CACCTCCATTACCTTGTATgaaagaacaaaagaaagaaaaaggaag...</td>\n",
       "      <td>797</td>\n",
       "      <td>40</td>\n",
       "      <td>101</td>\n",
       "      <td>0.9498117942283564</td>\n",
       "      <td>0.8732747804265998</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr10</td>\n",
       "      <td>122207800</td>\n",
       "      <td>122507800</td>\n",
       "      <td>2425</td>\n",
       "      <td>chr10</td>\n",
       "      <td>122216399</td>\n",
       "      <td>122217427</td>\n",
       "      <td>1029</td>\n",
       "      <td>CATGGTCATTCCCTGAAACAAACCATCATAAAAGGAGATTCATAGA...</td>\n",
       "      <td>chr7</td>\n",
       "      <td>130734765</td>\n",
       "      <td>130735709</td>\n",
       "      <td>945</td>\n",
       "      <td>CATGGTCACTCCCTGAAATCAACCATCAGTAAAGGAGATCCATAGA...</td>\n",
       "      <td>1060</td>\n",
       "      <td>31</td>\n",
       "      <td>115</td>\n",
       "      <td>0.970754716981132</td>\n",
       "      <td>0.8915094339622641</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   CHROM  POS_300KB  END_300KB  overall_index hg_chr   hg_start     hg_end  \\\n",
       "0  chr10  122207800  122507800           2425  chr10  122207147  122208150   \n",
       "1  chr10  122207800  122507800           2425  chr10  122209989  122212651   \n",
       "2  chr10  122207800  122507800           2425  chr10  122212878  122214639   \n",
       "3  chr10  122207800  122507800           2425  chr10  122214713  122215469   \n",
       "4  chr10  122207800  122507800           2425  chr10  122216399  122217427   \n",
       "\n",
       "  hg_len                                             hg_seq mm_chr   mm_start  \\\n",
       "0   1004  aattagccaggcgtggtggcacacacctgtaatcccagctactcgg...   chr7  130725901   \n",
       "1   2663  ccctggtggtttattcgctatcttcccctgaagaatgtgacttcca...   chr7  130728155   \n",
       "2   1762  AACTTGATTTCAAGCCTGGAATAGGAGAAGTCTGCAATGGCTAGAC...   chr7  130731103   \n",
       "3    757  CAGCCCCGCTA-TTCATATGGAGG----------AGAAAAAGGGAG...   chr7  130732889   \n",
       "4   1029  CATGGTCATTCCCTGAAACAAACCATCATAAAAGGAGATTCATAGA...   chr7  130734765   \n",
       "\n",
       "      mm_end mm_len                                             mm_seq  \\\n",
       "0  130726680    780  AACTAGATGGGCATGGTTGC---------------------CTTG-...   \n",
       "1  130730902   2748  CTTTGGTGGC-------------------AATGACAATGGCTGCTA...   \n",
       "2  130732753   1651  AACTTTATTTCAAAAGTGGAATAAT---ACTCACAAATGTCTACCC...   \n",
       "3  130733584    696  CACCTCCATTACCTTGTATgaaagaacaaaagaaagaaaaaggaag...   \n",
       "4  130735709    945  CATGGTCACTCCCTGAAATCAACCATCAGTAAAGGAGATCCATAGA...   \n",
       "\n",
       "  total_align_len hg_dash_n mm_dash_n          hg_percent          mm_percent  \n",
       "0            1032        28       252  0.9728682170542635  0.7558139534883721  \n",
       "1            2902       239       154  0.9176430048242591  0.9469331495520331  \n",
       "2            1833        71       182    0.96126568466994   0.900709219858156  \n",
       "3             797        40       101  0.9498117942283564  0.8732747804265998  \n",
       "4            1060        31       115   0.970754716981132  0.8915094339622641  "
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "variant_300kb_synt.columns = list(variants_300kb.columns) + list(synteny_condensed.columns)\n",
    "variant_300kb_synt = variant_300kb_synt[variant_300kb_synt['hg_chr'] != '.']\n",
    "print(len(variant_300kb_synt))\n",
    "variant_300kb_synt.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "f899372a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "54233\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS_1MB</th>\n",
       "      <th>END_1MB</th>\n",
       "      <th>overall_index</th>\n",
       "      <th>hg_chr</th>\n",
       "      <th>hg_start</th>\n",
       "      <th>hg_end</th>\n",
       "      <th>hg_len</th>\n",
       "      <th>hg_seq</th>\n",
       "      <th>mm_chr</th>\n",
       "      <th>mm_start</th>\n",
       "      <th>mm_end</th>\n",
       "      <th>mm_len</th>\n",
       "      <th>mm_seq</th>\n",
       "      <th>total_align_len</th>\n",
       "      <th>hg_dash_n</th>\n",
       "      <th>mm_dash_n</th>\n",
       "      <th>hg_percent</th>\n",
       "      <th>mm_percent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr10</td>\n",
       "      <td>121857800</td>\n",
       "      <td>122857800</td>\n",
       "      <td>2425</td>\n",
       "      <td>chr10</td>\n",
       "      <td>121860593</td>\n",
       "      <td>121861411</td>\n",
       "      <td>819</td>\n",
       "      <td>TGAGAAAGGAAGTACAGTAAATGTTTTTAATAAGT-TGGACAGAAA...</td>\n",
       "      <td>chr7</td>\n",
       "      <td>130472471</td>\n",
       "      <td>130473302</td>\n",
       "      <td>832</td>\n",
       "      <td>TGATAAAGGACTTACCACAAATGGTTTTAATAAATATATACAGAAA...</td>\n",
       "      <td>872</td>\n",
       "      <td>53</td>\n",
       "      <td>40</td>\n",
       "      <td>0.939220</td>\n",
       "      <td>0.954128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr10</td>\n",
       "      <td>121857800</td>\n",
       "      <td>122857800</td>\n",
       "      <td>2425</td>\n",
       "      <td>chr10</td>\n",
       "      <td>121861835</td>\n",
       "      <td>121862235</td>\n",
       "      <td>401</td>\n",
       "      <td>CCCTCATGACTAGGGTTAACAACAACAATGTACTCCTATTTTCTGC...</td>\n",
       "      <td>chr7</td>\n",
       "      <td>130473648</td>\n",
       "      <td>130474021</td>\n",
       "      <td>374</td>\n",
       "      <td>TCCTTACATATACAGCTACAAGAAACAAAACCCTGCTCCTCTCTGG...</td>\n",
       "      <td>401</td>\n",
       "      <td>0</td>\n",
       "      <td>27</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.932668</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr10</td>\n",
       "      <td>121857800</td>\n",
       "      <td>122857800</td>\n",
       "      <td>2425</td>\n",
       "      <td>chr10</td>\n",
       "      <td>121862273</td>\n",
       "      <td>121862391</td>\n",
       "      <td>119</td>\n",
       "      <td>ACAGCCATTCCTTCTGAATTCTAGCCATTGGAGAGACCTTGATATC...</td>\n",
       "      <td>chr7</td>\n",
       "      <td>130474043</td>\n",
       "      <td>130474146</td>\n",
       "      <td>104</td>\n",
       "      <td>ACCACTTCTCTTTCCAAATTCTGGCCGTTAGCAATACCTTACCACA...</td>\n",
       "      <td>119</td>\n",
       "      <td>0</td>\n",
       "      <td>15</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.873950</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr10</td>\n",
       "      <td>121857800</td>\n",
       "      <td>122857800</td>\n",
       "      <td>2425</td>\n",
       "      <td>chr10</td>\n",
       "      <td>121864539</td>\n",
       "      <td>121866810</td>\n",
       "      <td>2272</td>\n",
       "      <td>CAGAAACAGAAATTACCATGAAGCAGAGTTTTCACttagaccattt...</td>\n",
       "      <td>chr7</td>\n",
       "      <td>130475811</td>\n",
       "      <td>130477803</td>\n",
       "      <td>1993</td>\n",
       "      <td>CAAAAATAAAATTTCCCATGCAGCCAAGTTTG-ATTTAGGCT-TTT...</td>\n",
       "      <td>2382</td>\n",
       "      <td>110</td>\n",
       "      <td>389</td>\n",
       "      <td>0.953820</td>\n",
       "      <td>0.836692</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr10</td>\n",
       "      <td>121857800</td>\n",
       "      <td>122857800</td>\n",
       "      <td>2425</td>\n",
       "      <td>chr10</td>\n",
       "      <td>121866931</td>\n",
       "      <td>121867299</td>\n",
       "      <td>369</td>\n",
       "      <td>AACTGAGAGCATCCGAAGATTACAAAGGACAAAGCTCTCAATCTTA...</td>\n",
       "      <td>chr7</td>\n",
       "      <td>130477804</td>\n",
       "      <td>130478171</td>\n",
       "      <td>368</td>\n",
       "      <td>AACTAATAACatccaaaggt--aaaaagacagagcttctggtctca...</td>\n",
       "      <td>403</td>\n",
       "      <td>34</td>\n",
       "      <td>35</td>\n",
       "      <td>0.915633</td>\n",
       "      <td>0.913151</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   CHROM    POS_1MB    END_1MB  overall_index hg_chr   hg_start     hg_end  \\\n",
       "0  chr10  121857800  122857800           2425  chr10  121860593  121861411   \n",
       "1  chr10  121857800  122857800           2425  chr10  121861835  121862235   \n",
       "2  chr10  121857800  122857800           2425  chr10  121862273  121862391   \n",
       "3  chr10  121857800  122857800           2425  chr10  121864539  121866810   \n",
       "4  chr10  121857800  122857800           2425  chr10  121866931  121867299   \n",
       "\n",
       "   hg_len                                             hg_seq mm_chr  \\\n",
       "0     819  TGAGAAAGGAAGTACAGTAAATGTTTTTAATAAGT-TGGACAGAAA...   chr7   \n",
       "1     401  CCCTCATGACTAGGGTTAACAACAACAATGTACTCCTATTTTCTGC...   chr7   \n",
       "2     119  ACAGCCATTCCTTCTGAATTCTAGCCATTGGAGAGACCTTGATATC...   chr7   \n",
       "3    2272  CAGAAACAGAAATTACCATGAAGCAGAGTTTTCACttagaccattt...   chr7   \n",
       "4     369  AACTGAGAGCATCCGAAGATTACAAAGGACAAAGCTCTCAATCTTA...   chr7   \n",
       "\n",
       "    mm_start     mm_end  mm_len  \\\n",
       "0  130472471  130473302     832   \n",
       "1  130473648  130474021     374   \n",
       "2  130474043  130474146     104   \n",
       "3  130475811  130477803    1993   \n",
       "4  130477804  130478171     368   \n",
       "\n",
       "                                              mm_seq  total_align_len  \\\n",
       "0  TGATAAAGGACTTACCACAAATGGTTTTAATAAATATATACAGAAA...              872   \n",
       "1  TCCTTACATATACAGCTACAAGAAACAAAACCCTGCTCCTCTCTGG...              401   \n",
       "2  ACCACTTCTCTTTCCAAATTCTGGCCGTTAGCAATACCTTACCACA...              119   \n",
       "3  CAAAAATAAAATTTCCCATGCAGCCAAGTTTG-ATTTAGGCT-TTT...             2382   \n",
       "4  AACTAATAACatccaaaggt--aaaaagacagagcttctggtctca...              403   \n",
       "\n",
       "   hg_dash_n  mm_dash_n  hg_percent  mm_percent  \n",
       "0         53         40    0.939220    0.954128  \n",
       "1          0         27    1.000000    0.932668  \n",
       "2          0         15    1.000000    0.873950  \n",
       "3        110        389    0.953820    0.836692  \n",
       "4         34         35    0.915633    0.913151  "
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "variant_1mb_synt.columns = list(variants_1mb.columns) + list(synteny_condensed.columns)\n",
    "variant_1mb_synt = variant_1mb_synt[variant_1mb_synt['hg_chr'] != '.']\n",
    "print(len(variant_1mb_synt))\n",
    "variant_1mb_synt.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "80feb0fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_index(hg_seq, idx):\n",
    "    count_letters = 0  # To count the letters encountered in the combined string\n",
    "    idx_new = 0  # Corresponding index in combined string\n",
    "    \n",
    "    for char in hg_seq:\n",
    "        if count_letters == (idx+1):\n",
    "            break\n",
    "        if char.isalpha():\n",
    "            count_letters += 1\n",
    "        idx_new += 1\n",
    "    \n",
    "    return idx_new\n",
    "\n",
    "#given two aligned strings with start and end idx\n",
    "#count up amount of same characters within the alignment subset \n",
    "def count_str(hg_seq, mm_seq, start_idx, end_idx):\n",
    "    #print('start_idx', start_idx)\n",
    "    #print('end_idx', end_idx)\n",
    "    hg_subset = hg_seq[start_idx:end_idx]\n",
    "    mm_subset = mm_seq[start_idx:end_idx]\n",
    "\n",
    "    count_same = sum(1 for a, b in zip(hg_subset, mm_subset) if a == b)\n",
    "    \n",
    "#     if end_idx>start_idx:\n",
    "#         prop=count_same/(end_idx-start_idx)\n",
    "#     else:\n",
    "#         prop=0\n",
    "    prop=count_same/(end_idx-start_idx)\n",
    "\n",
    "    return(prop)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "66a007f2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "index  0\n",
      "index  10\n",
      "index  20\n",
      "index  30\n",
      "index  40\n",
      "index  50\n",
      "index  60\n",
      "index  70\n",
      "index  80\n",
      "index  90\n",
      "index  100\n",
      "index  110\n",
      "index  120\n",
      "index  130\n",
      "index  140\n",
      "Time taken: 24.575964212417603 seconds\n"
     ]
    }
   ],
   "source": [
    "start=time.time()\n",
    "for i in simple.index:\n",
    "   \n",
    "    if i % 10 == 0:\n",
    "        print('index ', i)\n",
    "    \n",
    "    \n",
    "    #THIS ONE FOR 1MB SCORE \n",
    "    try:\n",
    "        #print(i)\n",
    "        variant=simple.loc[i]\n",
    "        idx=variant.overall_index\n",
    "        int_df=variant_1mb_synt[variant_1mb_synt['overall_index'] == idx]\n",
    "\n",
    "        #print(len(int_df))\n",
    "    \n",
    "    \n",
    "        if len(int_df)>0:\n",
    "\n",
    "\n",
    "            int_df['start_idx']=(np.maximum(int_df['POS_1MB'], int_df['hg_start'])) - int_df['hg_start'] \n",
    "            int_df['end_idx']=(np.minimum(int_df['END_1MB'], int_df['hg_end'])) - int_df['hg_start'] \n",
    "            int_df['n_overlap']=int_df['end_idx']-int_df['start_idx']\n",
    "\n",
    "            \n",
    "            #get alignment indices\n",
    "            int_df['start_align_idx'] = int_df.apply(lambda row: find_index(row['hg_seq'], row['start_idx']), axis=1)\n",
    "            int_df['end_align_idx'] = int_df.apply(lambda row: find_index(row['hg_seq'], row['end_idx']), axis=1)\n",
    "            int_df['n_align_overlap']=int_df['end_align_idx']-int_df['start_align_idx']\n",
    "        \n",
    "        \n",
    "            int_df['count_prop'] = int_df.apply(lambda row: count_str(row['hg_seq'], row['mm_seq'], row['start_align_idx'], row['end_align_idx']), axis=1)\n",
    "\n",
    "\n",
    "            #proportion of SV that covers a syntenic region  \n",
    "            simple.loc[i, 'synt_prop_overlap_1MB'] = int(np.sum(int_df['n_overlap'])) /abs(simple.loc[i, 'postoend_1mb'])\n",
    "            #weighted proportion of SV in syntenic region, and how similar those regions are \n",
    "            simple.loc[i, 'align_synt_prop_1MB'] =  (np.sum(int_df['count_prop'] *int_df['n_overlap'])) /abs(simple.loc[i, 'postoend_1mb'])\n",
    "\n",
    "\n",
    "\n",
    "        else:\n",
    "            simple.loc[i, 'synt_prop_overlap_1MB'] = 0\n",
    "            simple.loc[i, 'align_synt_prop_1MB'] = 0\n",
    "            \n",
    "    except:\n",
    "        print('error in index', i)\n",
    "        \n",
    "        \n",
    "    #THIS ONE FOR NORMAL SV LENGTH \n",
    "    try:\n",
    "        #print(i)\n",
    "        variant=simple.loc[i]\n",
    "        idx=variant.overall_index\n",
    "        int_df=variant_300kb_synt[variant_300kb_synt['overall_index'] == idx]\n",
    "        #print(len(int_df))\n",
    "    \n",
    "    \n",
    "        if len(int_df)>0:\n",
    "\n",
    "\n",
    "            int_df['start_idx']=(np.maximum(int_df['POS_300KB'], int_df['hg_start'])) - int_df['hg_start'] \n",
    "            int_df['end_idx']=(np.minimum(int_df['END_300KB'], int_df['hg_end'])) - int_df['hg_start'] \n",
    "            int_df['n_overlap']=int_df['end_idx']-int_df['start_idx']\n",
    "\n",
    "            \n",
    "            #get alignment indices\n",
    "            int_df['start_align_idx'] = int_df.apply(lambda row: find_index(row['hg_seq'], row['start_idx']), axis=1)\n",
    "            int_df['end_align_idx'] = int_df.apply(lambda row: find_index(row['hg_seq'], row['end_idx']), axis=1)\n",
    "            int_df['n_align_overlap']=int_df['end_align_idx']-int_df['start_align_idx']\n",
    "        \n",
    "        \n",
    "            int_df['count_prop'] = int_df.apply(lambda row: count_str(row['hg_seq'], row['mm_seq'], row['start_align_idx'], row['end_align_idx']), axis=1)\n",
    "\n",
    "\n",
    "            #proportion of SV that covers a syntenic region  \n",
    "            simple.loc[i, 'synt_prop_overlap_300KB'] = int(np.sum(int_df['n_overlap'])) /abs(simple.loc[i, 'postoend_300kb'])\n",
    "            #weighted proportion of SV in syntenic region, and how similar those regions are \n",
    "            simple.loc[i, 'align_synt_prop_300KB'] =  (np.sum(int_df['count_prop'] *int_df['n_overlap'])) /abs(simple.loc[i, 'postoend_300kb'])\n",
    "\n",
    "\n",
    "\n",
    "        else:\n",
    "            simple.loc[i, 'synt_prop_overlap_300KB'] = 0\n",
    "            simple.loc[i, 'align_synt_prop_300KB'] = 0\n",
    "            \n",
    "    except:\n",
    "        print('error 1mb in index', i)\n",
    "        \n",
    "        \n",
    "end=time.time()\n",
    "print(f\"Time taken: {end-start} seconds\")\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3c5f01c2",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "b0e307cd",
   "metadata": {},
   "outputs": [],
   "source": [
    "simple.to_csv('/pollard/home/shzhang/20231024_simple_topvariants_synteny_combined.txt', sep='\\t', index=False)  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a753bf46",
   "metadata": {},
   "outputs": [],
   "source": [
    "simple"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "04f411ec",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "21a6002e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1e88143b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c961dfa4",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
