{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "9362b7b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd \n",
    "import sys, os, psutil, io\n",
    "import gzip\n",
    "import time \n",
    "import warnings\n",
    "\n",
    "from pybedtools import BedTool\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "warnings.filterwarnings('ignore')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "e85d3362",
   "metadata": {},
   "outputs": [],
   "source": [
    "synteny_condensed=pd.read_csv('/pollard/home/shzhang/data/synteny/hg38_mm10_condensed.txt', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "c3792fd8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>hg_chr</th>\n",
       "      <th>hg_start</th>\n",
       "      <th>hg_end</th>\n",
       "      <th>hg_len</th>\n",
       "      <th>hg_seq</th>\n",
       "      <th>mm_chr</th>\n",
       "      <th>mm_start</th>\n",
       "      <th>mm_end</th>\n",
       "      <th>mm_len</th>\n",
       "      <th>mm_seq</th>\n",
       "      <th>total_align_len</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>56554</td>\n",
       "      <td>58923</td>\n",
       "      <td>2370</td>\n",
       "      <td>CACATTATTCTGCAGTTTTTATCTTGCTTACGTATTTTTGGAATGT...</td>\n",
       "      <td>chr2</td>\n",
       "      <td>111473169</td>\n",
       "      <td>111475499</td>\n",
       "      <td>2331</td>\n",
       "      <td>cacattgtattgta--ctctattatgttt-tattgtattatattat...</td>\n",
       "      <td>2518</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>60916</td>\n",
       "      <td>61124</td>\n",
       "      <td>209</td>\n",
       "      <td>TTAATATTTCCCACTTCATCATTACTAACCTCACTCACTCTGAACT...</td>\n",
       "      <td>chr2</td>\n",
       "      <td>111476095</td>\n",
       "      <td>111476298</td>\n",
       "      <td>204</td>\n",
       "      <td>ttAATAT--CCCAGTGCACC-TCACTACTCTGGCTCACGCTGAGCC...</td>\n",
       "      <td>209</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>61125</td>\n",
       "      <td>61835</td>\n",
       "      <td>711</td>\n",
       "      <td>AGTTCAAGACGTTTAATCAAGACCCTTCACTTACTTAACTATACCT...</td>\n",
       "      <td>chr2</td>\n",
       "      <td>111477065</td>\n",
       "      <td>111477685</td>\n",
       "      <td>621</td>\n",
       "      <td>aGCCCAAGAAAATTAAACAAATCCCTTTACTAATTTAAATGTATCT...</td>\n",
       "      <td>731</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>61943</td>\n",
       "      <td>61971</td>\n",
       "      <td>29</td>\n",
       "      <td>ctccgcctcccggggtcaagctattctcc</td>\n",
       "      <td>chr2</td>\n",
       "      <td>111477686</td>\n",
       "      <td>111477713</td>\n",
       "      <td>28</td>\n",
       "      <td>CTCTGCTTTGTGTGGAGAAG-TATTGTCC</td>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>62160</td>\n",
       "      <td>62467</td>\n",
       "      <td>308</td>\n",
       "      <td>tgtaaaactttttcctaatttaacagaaaaataatagtattatatt...</td>\n",
       "      <td>chr2</td>\n",
       "      <td>111477762</td>\n",
       "      <td>111478104</td>\n",
       "      <td>343</td>\n",
       "      <td>tgtaa----tttttcttatttaaCA-AGAAATGAGA-------ATT...</td>\n",
       "      <td>372</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1229834</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57107075</td>\n",
       "      <td>57107468</td>\n",
       "      <td>394</td>\n",
       "      <td>TGATTATGTTCGTTGTGACTGACAGCTGGATTGtagtggttaagaa...</td>\n",
       "      <td>chrX_GL456233_random</td>\n",
       "      <td>316168</td>\n",
       "      <td>316535</td>\n",
       "      <td>368</td>\n",
       "      <td>TAATGATGTTGATTGTGACTAGCAACAAGACTGtagtgattaagag...</td>\n",
       "      <td>397</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1229835</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57125268</td>\n",
       "      <td>57126152</td>\n",
       "      <td>885</td>\n",
       "      <td>aaagctgcagtgagtcgtaattgtgccactgcactccagcttgggc...</td>\n",
       "      <td>chrX_GL456233_random</td>\n",
       "      <td>321571</td>\n",
       "      <td>322437</td>\n",
       "      <td>867</td>\n",
       "      <td>aaggctgtatcg---cctaatagtgccacttcccatgagccaagca...</td>\n",
       "      <td>954</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1229836</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57126153</td>\n",
       "      <td>57126418</td>\n",
       "      <td>266</td>\n",
       "      <td>AGGTTGTTACTGCAAATCATTCATGTAGTCAGTGCCAGGGGTTAAA...</td>\n",
       "      <td>chrX_GL456233_random</td>\n",
       "      <td>322699</td>\n",
       "      <td>322963</td>\n",
       "      <td>265</td>\n",
       "      <td>ATATTGTTACTACAAATCATTCATATAGTCAGTGTGAAGAATAAAA...</td>\n",
       "      <td>266</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1229837</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57126419</td>\n",
       "      <td>57126561</td>\n",
       "      <td>143</td>\n",
       "      <td>CTTTTAATAGTGAAACACATGACCAGGCAGTGGC-TTTCCTATGAA...</td>\n",
       "      <td>chrX_GL456233_random</td>\n",
       "      <td>323295</td>\n",
       "      <td>323435</td>\n",
       "      <td>141</td>\n",
       "      <td>CTTTTCATCATAAAGTACATGACCAGGAAATATCTTTTCTTGTGAA...</td>\n",
       "      <td>150</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1229838</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57128357</td>\n",
       "      <td>57130391</td>\n",
       "      <td>2035</td>\n",
       "      <td>TATGATTGATTCAAGAAAATAAAATTGCTCTCCTCGTCCCTCCAGG...</td>\n",
       "      <td>chrX_GL456233_random</td>\n",
       "      <td>324344</td>\n",
       "      <td>326292</td>\n",
       "      <td>1949</td>\n",
       "      <td>TGTGATAGAGTCATGGAAGTAAAATTGCTTTCTTTCTCCTTCCAGG...</td>\n",
       "      <td>2097</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1229839 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        hg_chr  hg_start    hg_end  hg_len  \\\n",
       "0         chr1     56554     58923    2370   \n",
       "1         chr1     60916     61124     209   \n",
       "2         chr1     61125     61835     711   \n",
       "3         chr1     61943     61971      29   \n",
       "4         chr1     62160     62467     308   \n",
       "...        ...       ...       ...     ...   \n",
       "1229834   chrY  57107075  57107468     394   \n",
       "1229835   chrY  57125268  57126152     885   \n",
       "1229836   chrY  57126153  57126418     266   \n",
       "1229837   chrY  57126419  57126561     143   \n",
       "1229838   chrY  57128357  57130391    2035   \n",
       "\n",
       "                                                    hg_seq  \\\n",
       "0        CACATTATTCTGCAGTTTTTATCTTGCTTACGTATTTTTGGAATGT...   \n",
       "1        TTAATATTTCCCACTTCATCATTACTAACCTCACTCACTCTGAACT...   \n",
       "2        AGTTCAAGACGTTTAATCAAGACCCTTCACTTACTTAACTATACCT...   \n",
       "3                            ctccgcctcccggggtcaagctattctcc   \n",
       "4        tgtaaaactttttcctaatttaacagaaaaataatagtattatatt...   \n",
       "...                                                    ...   \n",
       "1229834  TGATTATGTTCGTTGTGACTGACAGCTGGATTGtagtggttaagaa...   \n",
       "1229835  aaagctgcagtgagtcgtaattgtgccactgcactccagcttgggc...   \n",
       "1229836  AGGTTGTTACTGCAAATCATTCATGTAGTCAGTGCCAGGGGTTAAA...   \n",
       "1229837  CTTTTAATAGTGAAACACATGACCAGGCAGTGGC-TTTCCTATGAA...   \n",
       "1229838  TATGATTGATTCAAGAAAATAAAATTGCTCTCCTCGTCCCTCCAGG...   \n",
       "\n",
       "                       mm_chr   mm_start     mm_end  mm_len  \\\n",
       "0                        chr2  111473169  111475499    2331   \n",
       "1                        chr2  111476095  111476298     204   \n",
       "2                        chr2  111477065  111477685     621   \n",
       "3                        chr2  111477686  111477713      28   \n",
       "4                        chr2  111477762  111478104     343   \n",
       "...                       ...        ...        ...     ...   \n",
       "1229834  chrX_GL456233_random     316168     316535     368   \n",
       "1229835  chrX_GL456233_random     321571     322437     867   \n",
       "1229836  chrX_GL456233_random     322699     322963     265   \n",
       "1229837  chrX_GL456233_random     323295     323435     141   \n",
       "1229838  chrX_GL456233_random     324344     326292    1949   \n",
       "\n",
       "                                                    mm_seq  total_align_len  \n",
       "0        cacattgtattgta--ctctattatgttt-tattgtattatattat...             2518  \n",
       "1        ttAATAT--CCCAGTGCACC-TCACTACTCTGGCTCACGCTGAGCC...              209  \n",
       "2        aGCCCAAGAAAATTAAACAAATCCCTTTACTAATTTAAATGTATCT...              731  \n",
       "3                            CTCTGCTTTGTGTGGAGAAG-TATTGTCC               29  \n",
       "4        tgtaa----tttttcttatttaaCA-AGAAATGAGA-------ATT...              372  \n",
       "...                                                    ...              ...  \n",
       "1229834  TAATGATGTTGATTGTGACTAGCAACAAGACTGtagtgattaagag...              397  \n",
       "1229835  aaggctgtatcg---cctaatagtgccacttcccatgagccaagca...              954  \n",
       "1229836  ATATTGTTACTACAAATCATTCATATAGTCAGTGTGAAGAATAAAA...              266  \n",
       "1229837  CTTTTCATCATAAAGTACATGACCAGGAAATATCTTTTCTTGTGAA...              150  \n",
       "1229838  TGTGATAGAGTCATGGAAGTAAAATTGCTTTCTTTCTCCTTCCAGG...             2097  \n",
       "\n",
       "[1229839 rows x 11 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "synteny_condensed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "908d361e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n",
      "True\n"
     ]
    }
   ],
   "source": [
    "# #check that lengths of seqs and alignments are correct  \n",
    "# synteny_condensed['hg_dash_n']=synteny_condensed['hg_seq'].str.count('-')\n",
    "# synteny_condensed['mm_dash_n']=synteny_condensed['mm_seq'].str.count('-')\n",
    "\n",
    "# synteny_condensed['hg_percent']=1-(synteny_condensed['hg_dash_n']/synteny_condensed['total_align_len'])\n",
    "# synteny_condensed['mm_percent']=1-(synteny_condensed['mm_dash_n']/synteny_condensed['total_align_len'])\n",
    "\n",
    "\n",
    "# print(((synteny_condensed['hg_len'] + synteny_condensed['hg_dash_n']) == synteny_condensed['total_align_len']).all())\n",
    "# print(((synteny_condensed['mm_len'] + synteny_condensed['mm_dash_n']) == synteny_condensed['total_align_len']).all())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "a950d778",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.9612903225806452"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.median(synteny_condensed['hg_percent'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ee317551",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "994e441a",
   "metadata": {},
   "outputs": [],
   "source": [
    "chrom_len=pd.read_csv('/pollard/home/shzhang/akita_variant_scoring/data/chrom_lengths_hg38', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "0855f907",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>var_index</th>\n",
       "      <th>mse_mean</th>\n",
       "      <th>corr_mean</th>\n",
       "      <th>oneminus_corrmean</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>5.914621e-03</td>\n",
       "      <td>0.979139</td>\n",
       "      <td>2.086145e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>1.629834e-07</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>4.470926e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>1.445605e-08</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>3.242280e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>2.073410e-08</td>\n",
       "      <td>0.999999</td>\n",
       "      <td>9.900652e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>214039</th>\n",
       "      <td>214039</td>\n",
       "      <td>4.707555e-08</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>9.518438e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>214040</th>\n",
       "      <td>214040</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>214041</th>\n",
       "      <td>214041</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>214042</th>\n",
       "      <td>214042</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>214043</th>\n",
       "      <td>214043</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>214044 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        var_index      mse_mean  corr_mean  oneminus_corrmean\n",
       "0               0           NaN        NaN                NaN\n",
       "1               1  5.914621e-03   0.979139       2.086145e-02\n",
       "2               2  1.629834e-07   1.000000       4.470926e-07\n",
       "3               3  1.445605e-08   1.000000       3.242280e-08\n",
       "4               4  2.073410e-08   0.999999       9.900652e-07\n",
       "...           ...           ...        ...                ...\n",
       "214039     214039  4.707555e-08   1.000000       9.518438e-08\n",
       "214040     214040           NaN        NaN                NaN\n",
       "214041     214041           NaN        NaN                NaN\n",
       "214042     214042           NaN        NaN                NaN\n",
       "214043     214043           NaN        NaN                NaN\n",
       "\n",
       "[214044 rows x 4 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_path='/pollard/data/projects/kgjoni/CBTN_collab/CBTN_data/'\n",
    "scores=pd.read_csv(data_path + 'simple-variants_230620/simple_variants_scored/simple_variants_all_scores', sep='\\t')\n",
    "scores['oneminus_corrmean']=1-scores['corr_mean']\n",
    "scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "6df00472",
   "metadata": {},
   "outputs": [],
   "source": [
    "#get mse and corr idx \n",
    "mse_sort=scores.sort_values(by='mse_mean', ascending=False)\n",
    "top_1000_mse = mse_sort.index[:500]\n",
    "top_1000_mse_idx = mse_sort['var_index'].loc[top_1000_mse].tolist()\n",
    "\n",
    "corr_sort=scores.sort_values(by='oneminus_corrmean', ascending=False)\n",
    "top_1000_corr = corr_sort.index[:500]\n",
    "top_1000_corr_idx = corr_sort['var_index'].loc[top_1000_corr].tolist()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "13eb3ed1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "988"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "idx_union= list(set(top_1000_mse_idx).union(top_1000_corr_idx))\n",
    "len(idx_union)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "0dc61c57",
   "metadata": {},
   "outputs": [],
   "source": [
    "simple_df=pd.read_csv(data_path + 'simple-variants_230620/simple_variants_all', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "e05eff22",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>INFO</th>\n",
       "      <th>FORMAT</th>\n",
       "      <th>BiospecimenID</th>\n",
       "      <th>sample_var_index</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>133</th>\n",
       "      <td>chr5</td>\n",
       "      <td>20909748</td>\n",
       "      <td>.</td>\n",
       "      <td>TCTC</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>CAL=Strelka2,VarDict;CSQ=-|intron_variant&amp;non_...</td>\n",
       "      <td>GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS</td>\n",
       "      <td>BS_0YWFFRYV</td>\n",
       "      <td>133</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1009</th>\n",
       "      <td>chr1</td>\n",
       "      <td>28546633</td>\n",
       "      <td>.</td>\n",
       "      <td>CACGGTGGCTCACGCCTGTAATCCCAGCATTTTGGGAAGCCGAGGT...</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>CAL=VarDict,Lancet;CSQ=-|downstream_gene_varia...</td>\n",
       "      <td>GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS</td>\n",
       "      <td>BS_19EJ85F8</td>\n",
       "      <td>36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1103</th>\n",
       "      <td>chr1</td>\n",
       "      <td>155342274</td>\n",
       "      <td>.</td>\n",
       "      <td>TCA</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>CAL=Strelka2,Mutect2,VarDict,Lancet;CSQ=-|intr...</td>\n",
       "      <td>GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS</td>\n",
       "      <td>BS_19EJ85F8</td>\n",
       "      <td>130</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1147</th>\n",
       "      <td>chr1</td>\n",
       "      <td>213640280</td>\n",
       "      <td>.</td>\n",
       "      <td>TTC</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>CAL=Strelka2,VarDict;CSQ=-|intron_variant&amp;non_...</td>\n",
       "      <td>GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS</td>\n",
       "      <td>BS_19EJ85F8</td>\n",
       "      <td>174</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1739</th>\n",
       "      <td>chr5</td>\n",
       "      <td>105062135</td>\n",
       "      <td>.</td>\n",
       "      <td>CGT</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>CAL=Strelka2,VarDict;CSQ=-|intron_variant&amp;non_...</td>\n",
       "      <td>GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS</td>\n",
       "      <td>BS_19EJ85F8</td>\n",
       "      <td>766</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>212617</th>\n",
       "      <td>chrX</td>\n",
       "      <td>121433158</td>\n",
       "      <td>.</td>\n",
       "      <td>ATTACTATCT</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>CAL=Strelka2,Mutect2,VarDict;CSQ=-|intron_vari...</td>\n",
       "      <td>GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS</td>\n",
       "      <td>BS_QR84QJN4</td>\n",
       "      <td>5143</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>212786</th>\n",
       "      <td>chr1</td>\n",
       "      <td>187435115</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>TTA</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>CAL=Strelka2,Mutect2,VarDict,Lancet;CSQ=TA|int...</td>\n",
       "      <td>GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS</td>\n",
       "      <td>BS_A86314EJ</td>\n",
       "      <td>93</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>212829</th>\n",
       "      <td>chr2</td>\n",
       "      <td>77613461</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>TA</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>CAL=Strelka2,Mutect2,VarDict,Lancet;CSQ=A|inte...</td>\n",
       "      <td>GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS</td>\n",
       "      <td>BS_A86314EJ</td>\n",
       "      <td>136</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>212914</th>\n",
       "      <td>chr3</td>\n",
       "      <td>6395149</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>GCGCCCCCCCC</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>CAL=Mutect2,VarDict;CSQ=CGCCCCCCCC|intergenic_...</td>\n",
       "      <td>GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS</td>\n",
       "      <td>BS_A86314EJ</td>\n",
       "      <td>221</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>213854</th>\n",
       "      <td>chr18</td>\n",
       "      <td>53010569</td>\n",
       "      <td>.</td>\n",
       "      <td>ATGTACATAGG</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>CAL=Strelka2,VarDict,Lancet;CSQ=-|intron_varia...</td>\n",
       "      <td>GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS</td>\n",
       "      <td>BS_A86314EJ</td>\n",
       "      <td>1161</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>988 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        CHROM        POS ID  \\\n",
       "133      chr5   20909748  .   \n",
       "1009     chr1   28546633  .   \n",
       "1103     chr1  155342274  .   \n",
       "1147     chr1  213640280  .   \n",
       "1739     chr5  105062135  .   \n",
       "...       ...        ... ..   \n",
       "212617   chrX  121433158  .   \n",
       "212786   chr1  187435115  .   \n",
       "212829   chr2   77613461  .   \n",
       "212914   chr3    6395149  .   \n",
       "213854  chr18   53010569  .   \n",
       "\n",
       "                                                      REF          ALT QUAL  \\\n",
       "133                                                  TCTC            T    .   \n",
       "1009    CACGGTGGCTCACGCCTGTAATCCCAGCATTTTGGGAAGCCGAGGT...            C    .   \n",
       "1103                                                  TCA            T    .   \n",
       "1147                                                  TTC            T    .   \n",
       "1739                                                  CGT            C    .   \n",
       "...                                                   ...          ...  ...   \n",
       "212617                                         ATTACTATCT            A    .   \n",
       "212786                                                  T          TTA    .   \n",
       "212829                                                  T           TA    .   \n",
       "212914                                                  G  GCGCCCCCCCC    .   \n",
       "213854                                        ATGTACATAGG            A    .   \n",
       "\n",
       "       FILTER                                               INFO  \\\n",
       "133      PASS  CAL=Strelka2,VarDict;CSQ=-|intron_variant&non_...   \n",
       "1009     PASS  CAL=VarDict,Lancet;CSQ=-|downstream_gene_varia...   \n",
       "1103     PASS  CAL=Strelka2,Mutect2,VarDict,Lancet;CSQ=-|intr...   \n",
       "1147     PASS  CAL=Strelka2,VarDict;CSQ=-|intron_variant&non_...   \n",
       "1739     PASS  CAL=Strelka2,VarDict;CSQ=-|intron_variant&non_...   \n",
       "...       ...                                                ...   \n",
       "212617   PASS  CAL=Strelka2,Mutect2,VarDict;CSQ=-|intron_vari...   \n",
       "212786   PASS  CAL=Strelka2,Mutect2,VarDict,Lancet;CSQ=TA|int...   \n",
       "212829   PASS  CAL=Strelka2,Mutect2,VarDict,Lancet;CSQ=A|inte...   \n",
       "212914   PASS  CAL=Mutect2,VarDict;CSQ=CGCCCCCCCC|intergenic_...   \n",
       "213854   PASS  CAL=Strelka2,VarDict,Lancet;CSQ=-|intron_varia...   \n",
       "\n",
       "                                                   FORMAT BiospecimenID  \\\n",
       "133     GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS   BS_0YWFFRYV   \n",
       "1009    GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS   BS_19EJ85F8   \n",
       "1103    GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS   BS_19EJ85F8   \n",
       "1147    GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS   BS_19EJ85F8   \n",
       "1739    GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS   BS_19EJ85F8   \n",
       "...                                                   ...           ...   \n",
       "212617  GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS   BS_QR84QJN4   \n",
       "212786  GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS   BS_A86314EJ   \n",
       "212829  GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS   BS_A86314EJ   \n",
       "212914  GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS   BS_A86314EJ   \n",
       "213854  GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS   BS_A86314EJ   \n",
       "\n",
       "        sample_var_index  \n",
       "133                  133  \n",
       "1009                  36  \n",
       "1103                 130  \n",
       "1147                 174  \n",
       "1739                 766  \n",
       "...                  ...  \n",
       "212617              5143  \n",
       "212786                93  \n",
       "212829               136  \n",
       "212914               221  \n",
       "213854              1161  \n",
       "\n",
       "[988 rows x 11 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#keep only those in top mse and corr\n",
    "#simple_top = simple_df[simple_df['sample_var_index'].isin(idx_union)]\n",
    "simple_top=simple_df[simple_df.index.isin(idx_union)]\n",
    "simple_top"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "11fa1788",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>INFO</th>\n",
       "      <th>FORMAT</th>\n",
       "      <th>BiospecimenID</th>\n",
       "      <th>sample_var_index</th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr5</td>\n",
       "      <td>20909748</td>\n",
       "      <td>.</td>\n",
       "      <td>TCTC</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>CAL=Strelka2,VarDict;CSQ=-|intron_variant&amp;non_...</td>\n",
       "      <td>GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS</td>\n",
       "      <td>BS_0YWFFRYV</td>\n",
       "      <td>133</td>\n",
       "      <td>chr5</td>\n",
       "      <td>181538259</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>28546633</td>\n",
       "      <td>.</td>\n",
       "      <td>CACGGTGGCTCACGCCTGTAATCCCAGCATTTTGGGAAGCCGAGGT...</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>CAL=VarDict,Lancet;CSQ=-|downstream_gene_varia...</td>\n",
       "      <td>GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS</td>\n",
       "      <td>BS_19EJ85F8</td>\n",
       "      <td>36</td>\n",
       "      <td>chr1</td>\n",
       "      <td>248956422</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>155342274</td>\n",
       "      <td>.</td>\n",
       "      <td>TCA</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>CAL=Strelka2,Mutect2,VarDict,Lancet;CSQ=-|intr...</td>\n",
       "      <td>GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS</td>\n",
       "      <td>BS_19EJ85F8</td>\n",
       "      <td>130</td>\n",
       "      <td>chr1</td>\n",
       "      <td>248956422</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>213640280</td>\n",
       "      <td>.</td>\n",
       "      <td>TTC</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>CAL=Strelka2,VarDict;CSQ=-|intron_variant&amp;non_...</td>\n",
       "      <td>GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS</td>\n",
       "      <td>BS_19EJ85F8</td>\n",
       "      <td>174</td>\n",
       "      <td>chr1</td>\n",
       "      <td>248956422</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr5</td>\n",
       "      <td>105062135</td>\n",
       "      <td>.</td>\n",
       "      <td>CGT</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>CAL=Strelka2,VarDict;CSQ=-|intron_variant&amp;non_...</td>\n",
       "      <td>GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS</td>\n",
       "      <td>BS_19EJ85F8</td>\n",
       "      <td>766</td>\n",
       "      <td>chr5</td>\n",
       "      <td>181538259</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>983</th>\n",
       "      <td>chrX</td>\n",
       "      <td>121433158</td>\n",
       "      <td>.</td>\n",
       "      <td>ATTACTATCT</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>CAL=Strelka2,Mutect2,VarDict;CSQ=-|intron_vari...</td>\n",
       "      <td>GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS</td>\n",
       "      <td>BS_QR84QJN4</td>\n",
       "      <td>5143</td>\n",
       "      <td>chrX</td>\n",
       "      <td>156040895</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>984</th>\n",
       "      <td>chr1</td>\n",
       "      <td>187435115</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>TTA</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>CAL=Strelka2,Mutect2,VarDict,Lancet;CSQ=TA|int...</td>\n",
       "      <td>GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS</td>\n",
       "      <td>BS_A86314EJ</td>\n",
       "      <td>93</td>\n",
       "      <td>chr1</td>\n",
       "      <td>248956422</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>985</th>\n",
       "      <td>chr2</td>\n",
       "      <td>77613461</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>TA</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>CAL=Strelka2,Mutect2,VarDict,Lancet;CSQ=A|inte...</td>\n",
       "      <td>GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS</td>\n",
       "      <td>BS_A86314EJ</td>\n",
       "      <td>136</td>\n",
       "      <td>chr2</td>\n",
       "      <td>242193529</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>986</th>\n",
       "      <td>chr3</td>\n",
       "      <td>6395149</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>GCGCCCCCCCC</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>CAL=Mutect2,VarDict;CSQ=CGCCCCCCCC|intergenic_...</td>\n",
       "      <td>GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS</td>\n",
       "      <td>BS_A86314EJ</td>\n",
       "      <td>221</td>\n",
       "      <td>chr3</td>\n",
       "      <td>198295559</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>987</th>\n",
       "      <td>chr18</td>\n",
       "      <td>53010569</td>\n",
       "      <td>.</td>\n",
       "      <td>ATGTACATAGG</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>CAL=Strelka2,VarDict,Lancet;CSQ=-|intron_varia...</td>\n",
       "      <td>GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS</td>\n",
       "      <td>BS_A86314EJ</td>\n",
       "      <td>1161</td>\n",
       "      <td>chr18</td>\n",
       "      <td>80373285</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>988 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     CHROM        POS ID                                                REF  \\\n",
       "0     chr5   20909748  .                                               TCTC   \n",
       "1     chr1   28546633  .  CACGGTGGCTCACGCCTGTAATCCCAGCATTTTGGGAAGCCGAGGT...   \n",
       "2     chr1  155342274  .                                                TCA   \n",
       "3     chr1  213640280  .                                                TTC   \n",
       "4     chr5  105062135  .                                                CGT   \n",
       "..     ...        ... ..                                                ...   \n",
       "983   chrX  121433158  .                                         ATTACTATCT   \n",
       "984   chr1  187435115  .                                                  T   \n",
       "985   chr2   77613461  .                                                  T   \n",
       "986   chr3    6395149  .                                                  G   \n",
       "987  chr18   53010569  .                                        ATGTACATAGG   \n",
       "\n",
       "             ALT QUAL FILTER  \\\n",
       "0              T    .   PASS   \n",
       "1              C    .   PASS   \n",
       "2              T    .   PASS   \n",
       "3              T    .   PASS   \n",
       "4              C    .   PASS   \n",
       "..           ...  ...    ...   \n",
       "983            A    .   PASS   \n",
       "984          TTA    .   PASS   \n",
       "985           TA    .   PASS   \n",
       "986  GCGCCCCCCCC    .   PASS   \n",
       "987            A    .   PASS   \n",
       "\n",
       "                                                  INFO  \\\n",
       "0    CAL=Strelka2,VarDict;CSQ=-|intron_variant&non_...   \n",
       "1    CAL=VarDict,Lancet;CSQ=-|downstream_gene_varia...   \n",
       "2    CAL=Strelka2,Mutect2,VarDict,Lancet;CSQ=-|intr...   \n",
       "3    CAL=Strelka2,VarDict;CSQ=-|intron_variant&non_...   \n",
       "4    CAL=Strelka2,VarDict;CSQ=-|intron_variant&non_...   \n",
       "..                                                 ...   \n",
       "983  CAL=Strelka2,Mutect2,VarDict;CSQ=-|intron_vari...   \n",
       "984  CAL=Strelka2,Mutect2,VarDict,Lancet;CSQ=TA|int...   \n",
       "985  CAL=Strelka2,Mutect2,VarDict,Lancet;CSQ=A|inte...   \n",
       "986  CAL=Mutect2,VarDict;CSQ=CGCCCCCCCC|intergenic_...   \n",
       "987  CAL=Strelka2,VarDict,Lancet;CSQ=-|intron_varia...   \n",
       "\n",
       "                                                FORMAT BiospecimenID  \\\n",
       "0    GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS   BS_0YWFFRYV   \n",
       "1    GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS   BS_19EJ85F8   \n",
       "2    GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS   BS_19EJ85F8   \n",
       "3    GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS   BS_19EJ85F8   \n",
       "4    GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS   BS_19EJ85F8   \n",
       "..                                                 ...           ...   \n",
       "983  GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS   BS_QR84QJN4   \n",
       "984  GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS   BS_A86314EJ   \n",
       "985  GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS   BS_A86314EJ   \n",
       "986  GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS   BS_A86314EJ   \n",
       "987  GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS   BS_A86314EJ   \n",
       "\n",
       "     sample_var_index      0          1  \n",
       "0                 133   chr5  181538259  \n",
       "1                  36   chr1  248956422  \n",
       "2                 130   chr1  248956422  \n",
       "3                 174   chr1  248956422  \n",
       "4                 766   chr5  181538259  \n",
       "..                ...    ...        ...  \n",
       "983              5143   chrX  156040895  \n",
       "984                93   chr1  248956422  \n",
       "985               136   chr2  242193529  \n",
       "986               221   chr3  198295559  \n",
       "987              1161  chr18   80373285  \n",
       "\n",
       "[988 rows x 13 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#merge with chr len information\n",
    "chrom_len=pd.read_csv('/pollard/home/shzhang/akita_variant_scoring/data/chrom_lengths_hg38', sep='\\t', header=None)\n",
    "chrom_len[0]='chr' + chrom_len[0]\n",
    "simple=simple_top.merge(chrom_len, right_on=0, left_on='CHROM', how='left')\n",
    "simple = simple.reset_index(drop=True)\n",
    "\n",
    "simple"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "2b27b213",
   "metadata": {},
   "outputs": [],
   "source": [
    "simple = simple.rename(columns={1: 'chr_len'})\n",
    "\n",
    "\n",
    "simple['POS_300KB']=np.floor(np.maximum(0, simple['POS']-150e3))\n",
    "simple['END_300KB']=np.floor(np.minimum(simple['chr_len'], simple['POS']+150e3))\n",
    "simple['POS_1MB']=np.floor(np.maximum(0, simple['POS']-5e5))\n",
    "simple['END_1MB']=np.floor(np.minimum(simple['chr_len'], simple['POS']+ 5e5))\n",
    "\n",
    "\n",
    "simple['postoend_1mb']=simple['END_1MB']-simple['POS_1MB']\n",
    "simple['postoend_300kb']=simple['END_300KB']-simple['POS_300KB']\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "ed0309ca",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>INFO</th>\n",
       "      <th>FORMAT</th>\n",
       "      <th>BiospecimenID</th>\n",
       "      <th>sample_var_index</th>\n",
       "      <th>0</th>\n",
       "      <th>chr_len</th>\n",
       "      <th>POS_300KB</th>\n",
       "      <th>END_300KB</th>\n",
       "      <th>POS_1MB</th>\n",
       "      <th>END_1MB</th>\n",
       "      <th>postoend_1mb</th>\n",
       "      <th>postoend_300kb</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr5</td>\n",
       "      <td>20909748</td>\n",
       "      <td>.</td>\n",
       "      <td>TCTC</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>CAL=Strelka2,VarDict;CSQ=-|intron_variant&amp;non_...</td>\n",
       "      <td>GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS</td>\n",
       "      <td>BS_0YWFFRYV</td>\n",
       "      <td>133</td>\n",
       "      <td>chr5</td>\n",
       "      <td>181538259</td>\n",
       "      <td>20759748.0</td>\n",
       "      <td>21059748.0</td>\n",
       "      <td>20409748.0</td>\n",
       "      <td>21409748.0</td>\n",
       "      <td>1000000.0</td>\n",
       "      <td>300000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>28546633</td>\n",
       "      <td>.</td>\n",
       "      <td>CACGGTGGCTCACGCCTGTAATCCCAGCATTTTGGGAAGCCGAGGT...</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>CAL=VarDict,Lancet;CSQ=-|downstream_gene_varia...</td>\n",
       "      <td>GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS</td>\n",
       "      <td>BS_19EJ85F8</td>\n",
       "      <td>36</td>\n",
       "      <td>chr1</td>\n",
       "      <td>248956422</td>\n",
       "      <td>28396633.0</td>\n",
       "      <td>28696633.0</td>\n",
       "      <td>28046633.0</td>\n",
       "      <td>29046633.0</td>\n",
       "      <td>1000000.0</td>\n",
       "      <td>300000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>155342274</td>\n",
       "      <td>.</td>\n",
       "      <td>TCA</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>CAL=Strelka2,Mutect2,VarDict,Lancet;CSQ=-|intr...</td>\n",
       "      <td>GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS</td>\n",
       "      <td>BS_19EJ85F8</td>\n",
       "      <td>130</td>\n",
       "      <td>chr1</td>\n",
       "      <td>248956422</td>\n",
       "      <td>155192274.0</td>\n",
       "      <td>155492274.0</td>\n",
       "      <td>154842274.0</td>\n",
       "      <td>155842274.0</td>\n",
       "      <td>1000000.0</td>\n",
       "      <td>300000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>213640280</td>\n",
       "      <td>.</td>\n",
       "      <td>TTC</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>CAL=Strelka2,VarDict;CSQ=-|intron_variant&amp;non_...</td>\n",
       "      <td>GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS</td>\n",
       "      <td>BS_19EJ85F8</td>\n",
       "      <td>174</td>\n",
       "      <td>chr1</td>\n",
       "      <td>248956422</td>\n",
       "      <td>213490280.0</td>\n",
       "      <td>213790280.0</td>\n",
       "      <td>213140280.0</td>\n",
       "      <td>214140280.0</td>\n",
       "      <td>1000000.0</td>\n",
       "      <td>300000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr5</td>\n",
       "      <td>105062135</td>\n",
       "      <td>.</td>\n",
       "      <td>CGT</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>CAL=Strelka2,VarDict;CSQ=-|intron_variant&amp;non_...</td>\n",
       "      <td>GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS</td>\n",
       "      <td>BS_19EJ85F8</td>\n",
       "      <td>766</td>\n",
       "      <td>chr5</td>\n",
       "      <td>181538259</td>\n",
       "      <td>104912135.0</td>\n",
       "      <td>105212135.0</td>\n",
       "      <td>104562135.0</td>\n",
       "      <td>105562135.0</td>\n",
       "      <td>1000000.0</td>\n",
       "      <td>300000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>983</th>\n",
       "      <td>chrX</td>\n",
       "      <td>121433158</td>\n",
       "      <td>.</td>\n",
       "      <td>ATTACTATCT</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>CAL=Strelka2,Mutect2,VarDict;CSQ=-|intron_vari...</td>\n",
       "      <td>GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS</td>\n",
       "      <td>BS_QR84QJN4</td>\n",
       "      <td>5143</td>\n",
       "      <td>chrX</td>\n",
       "      <td>156040895</td>\n",
       "      <td>121283158.0</td>\n",
       "      <td>121583158.0</td>\n",
       "      <td>120933158.0</td>\n",
       "      <td>121933158.0</td>\n",
       "      <td>1000000.0</td>\n",
       "      <td>300000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>984</th>\n",
       "      <td>chr1</td>\n",
       "      <td>187435115</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>TTA</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>CAL=Strelka2,Mutect2,VarDict,Lancet;CSQ=TA|int...</td>\n",
       "      <td>GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS</td>\n",
       "      <td>BS_A86314EJ</td>\n",
       "      <td>93</td>\n",
       "      <td>chr1</td>\n",
       "      <td>248956422</td>\n",
       "      <td>187285115.0</td>\n",
       "      <td>187585115.0</td>\n",
       "      <td>186935115.0</td>\n",
       "      <td>187935115.0</td>\n",
       "      <td>1000000.0</td>\n",
       "      <td>300000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>985</th>\n",
       "      <td>chr2</td>\n",
       "      <td>77613461</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>TA</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>CAL=Strelka2,Mutect2,VarDict,Lancet;CSQ=A|inte...</td>\n",
       "      <td>GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS</td>\n",
       "      <td>BS_A86314EJ</td>\n",
       "      <td>136</td>\n",
       "      <td>chr2</td>\n",
       "      <td>242193529</td>\n",
       "      <td>77463461.0</td>\n",
       "      <td>77763461.0</td>\n",
       "      <td>77113461.0</td>\n",
       "      <td>78113461.0</td>\n",
       "      <td>1000000.0</td>\n",
       "      <td>300000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>986</th>\n",
       "      <td>chr3</td>\n",
       "      <td>6395149</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>GCGCCCCCCCC</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>CAL=Mutect2,VarDict;CSQ=CGCCCCCCCC|intergenic_...</td>\n",
       "      <td>GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS</td>\n",
       "      <td>BS_A86314EJ</td>\n",
       "      <td>221</td>\n",
       "      <td>chr3</td>\n",
       "      <td>198295559</td>\n",
       "      <td>6245149.0</td>\n",
       "      <td>6545149.0</td>\n",
       "      <td>5895149.0</td>\n",
       "      <td>6895149.0</td>\n",
       "      <td>1000000.0</td>\n",
       "      <td>300000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>987</th>\n",
       "      <td>chr18</td>\n",
       "      <td>53010569</td>\n",
       "      <td>.</td>\n",
       "      <td>ATGTACATAGG</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>CAL=Strelka2,VarDict,Lancet;CSQ=-|intron_varia...</td>\n",
       "      <td>GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS</td>\n",
       "      <td>BS_A86314EJ</td>\n",
       "      <td>1161</td>\n",
       "      <td>chr18</td>\n",
       "      <td>80373285</td>\n",
       "      <td>52860569.0</td>\n",
       "      <td>53160569.0</td>\n",
       "      <td>52510569.0</td>\n",
       "      <td>53510569.0</td>\n",
       "      <td>1000000.0</td>\n",
       "      <td>300000.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>988 rows × 19 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     CHROM        POS ID                                                REF  \\\n",
       "0     chr5   20909748  .                                               TCTC   \n",
       "1     chr1   28546633  .  CACGGTGGCTCACGCCTGTAATCCCAGCATTTTGGGAAGCCGAGGT...   \n",
       "2     chr1  155342274  .                                                TCA   \n",
       "3     chr1  213640280  .                                                TTC   \n",
       "4     chr5  105062135  .                                                CGT   \n",
       "..     ...        ... ..                                                ...   \n",
       "983   chrX  121433158  .                                         ATTACTATCT   \n",
       "984   chr1  187435115  .                                                  T   \n",
       "985   chr2   77613461  .                                                  T   \n",
       "986   chr3    6395149  .                                                  G   \n",
       "987  chr18   53010569  .                                        ATGTACATAGG   \n",
       "\n",
       "             ALT QUAL FILTER  \\\n",
       "0              T    .   PASS   \n",
       "1              C    .   PASS   \n",
       "2              T    .   PASS   \n",
       "3              T    .   PASS   \n",
       "4              C    .   PASS   \n",
       "..           ...  ...    ...   \n",
       "983            A    .   PASS   \n",
       "984          TTA    .   PASS   \n",
       "985           TA    .   PASS   \n",
       "986  GCGCCCCCCCC    .   PASS   \n",
       "987            A    .   PASS   \n",
       "\n",
       "                                                  INFO  \\\n",
       "0    CAL=Strelka2,VarDict;CSQ=-|intron_variant&non_...   \n",
       "1    CAL=VarDict,Lancet;CSQ=-|downstream_gene_varia...   \n",
       "2    CAL=Strelka2,Mutect2,VarDict,Lancet;CSQ=-|intr...   \n",
       "3    CAL=Strelka2,VarDict;CSQ=-|intron_variant&non_...   \n",
       "4    CAL=Strelka2,VarDict;CSQ=-|intron_variant&non_...   \n",
       "..                                                 ...   \n",
       "983  CAL=Strelka2,Mutect2,VarDict;CSQ=-|intron_vari...   \n",
       "984  CAL=Strelka2,Mutect2,VarDict,Lancet;CSQ=TA|int...   \n",
       "985  CAL=Strelka2,Mutect2,VarDict,Lancet;CSQ=A|inte...   \n",
       "986  CAL=Mutect2,VarDict;CSQ=CGCCCCCCCC|intergenic_...   \n",
       "987  CAL=Strelka2,VarDict,Lancet;CSQ=-|intron_varia...   \n",
       "\n",
       "                                                FORMAT BiospecimenID  \\\n",
       "0    GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS   BS_0YWFFRYV   \n",
       "1    GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS   BS_19EJ85F8   \n",
       "2    GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS   BS_19EJ85F8   \n",
       "3    GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS   BS_19EJ85F8   \n",
       "4    GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS   BS_19EJ85F8   \n",
       "..                                                 ...           ...   \n",
       "983  GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS   BS_QR84QJN4   \n",
       "984  GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS   BS_A86314EJ   \n",
       "985  GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS   BS_A86314EJ   \n",
       "986  GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS   BS_A86314EJ   \n",
       "987  GT:AD:ADC:ADR:AF:AFC:AFR:DP:DPC:DPR:GTC:GT_STATUS   BS_A86314EJ   \n",
       "\n",
       "     sample_var_index      0    chr_len    POS_300KB    END_300KB  \\\n",
       "0                 133   chr5  181538259   20759748.0   21059748.0   \n",
       "1                  36   chr1  248956422   28396633.0   28696633.0   \n",
       "2                 130   chr1  248956422  155192274.0  155492274.0   \n",
       "3                 174   chr1  248956422  213490280.0  213790280.0   \n",
       "4                 766   chr5  181538259  104912135.0  105212135.0   \n",
       "..                ...    ...        ...          ...          ...   \n",
       "983              5143   chrX  156040895  121283158.0  121583158.0   \n",
       "984                93   chr1  248956422  187285115.0  187585115.0   \n",
       "985               136   chr2  242193529   77463461.0   77763461.0   \n",
       "986               221   chr3  198295559    6245149.0    6545149.0   \n",
       "987              1161  chr18   80373285   52860569.0   53160569.0   \n",
       "\n",
       "         POS_1MB      END_1MB  postoend_1mb  postoend_300kb  \n",
       "0     20409748.0   21409748.0     1000000.0        300000.0  \n",
       "1     28046633.0   29046633.0     1000000.0        300000.0  \n",
       "2    154842274.0  155842274.0     1000000.0        300000.0  \n",
       "3    213140280.0  214140280.0     1000000.0        300000.0  \n",
       "4    104562135.0  105562135.0     1000000.0        300000.0  \n",
       "..           ...          ...           ...             ...  \n",
       "983  120933158.0  121933158.0     1000000.0        300000.0  \n",
       "984  186935115.0  187935115.0     1000000.0        300000.0  \n",
       "985   77113461.0   78113461.0     1000000.0        300000.0  \n",
       "986    5895149.0    6895149.0     1000000.0        300000.0  \n",
       "987   52510569.0   53510569.0     1000000.0        300000.0  \n",
       "\n",
       "[988 rows x 19 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "simple\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "0e620baf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "988\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS_300KB</th>\n",
       "      <th>END_300KB</th>\n",
       "      <th>ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr5</td>\n",
       "      <td>20759748</td>\n",
       "      <td>21059748</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>28396633</td>\n",
       "      <td>28696633</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>155192274</td>\n",
       "      <td>155492274</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>213490280</td>\n",
       "      <td>213790280</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr5</td>\n",
       "      <td>104912135</td>\n",
       "      <td>105212135</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  CHROM  POS_300KB  END_300KB ID\n",
       "0  chr5   20759748   21059748  .\n",
       "1  chr1   28396633   28696633  .\n",
       "2  chr1  155192274  155492274  .\n",
       "3  chr1  213490280  213790280  .\n",
       "4  chr5  104912135  105212135  ."
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "variants_300kb=simple\n",
    "variants_300kb=variants_300kb[['CHROM', 'POS_300KB', 'END_300KB', 'ID']]\n",
    "variants_300kb['POS_300KB']=variants_300kb['POS_300KB'].astype(int)\n",
    "variants_300kb['END_300KB']=variants_300kb['END_300KB'].astype(int)\n",
    "print(len(variants_300kb))\n",
    "variants_300kb.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "98cee638",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "988\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS_1MB</th>\n",
       "      <th>END_1MB</th>\n",
       "      <th>ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr5</td>\n",
       "      <td>20409748</td>\n",
       "      <td>21409748</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>28046633</td>\n",
       "      <td>29046633</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>154842274</td>\n",
       "      <td>155842274</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>213140280</td>\n",
       "      <td>214140280</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr5</td>\n",
       "      <td>104562135</td>\n",
       "      <td>105562135</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  CHROM    POS_1MB    END_1MB ID\n",
       "0  chr5   20409748   21409748  .\n",
       "1  chr1   28046633   29046633  .\n",
       "2  chr1  154842274  155842274  .\n",
       "3  chr1  213140280  214140280  .\n",
       "4  chr5  104562135  105562135  ."
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "variants_1mb=simple\n",
    "variants_1mb=variants_1mb[['CHROM', 'POS_1MB', 'END_1MB', 'ID']]\n",
    "variants_1mb['POS_1MB']=variants_1mb['POS_1MB'].astype(int)\n",
    "variants_1mb['END_1MB']=variants_1mb['END_1MB'].astype(int)\n",
    "print(len(variants_1mb))\n",
    "variants_1mb.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "2f5191fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "variant_300kb_BED = BedTool.from_dataframe(variants_300kb)\n",
    "variant_1mb_BED = BedTool.from_dataframe(variants_1mb)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "631fe2b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "synteny_BED = BedTool.from_dataframe(synteny_condensed)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "469600c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "#reports only SVs where there is overlap with  gene\n",
    "#-F ensures that at least 0.5 of gene has overlap with SV \n",
    "#use loj in this case, because many SVs overlap multiple syntenic regions  \n",
    "variant_300kb_synt=variant_300kb_BED.intersect(synteny_BED, loj=True, wo=True).to_dataframe(header=None)\n",
    "variant_1mb_synt=variant_1mb_BED.intersect(synteny_BED, loj=True, wo=True).to_dataframe(header=None)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "49091322",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "113678\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS_300KB</th>\n",
       "      <th>END_300KB</th>\n",
       "      <th>ID</th>\n",
       "      <th>hg_chr</th>\n",
       "      <th>hg_start</th>\n",
       "      <th>hg_end</th>\n",
       "      <th>hg_len</th>\n",
       "      <th>hg_seq</th>\n",
       "      <th>mm_chr</th>\n",
       "      <th>mm_start</th>\n",
       "      <th>mm_end</th>\n",
       "      <th>mm_len</th>\n",
       "      <th>mm_seq</th>\n",
       "      <th>total_align_len</th>\n",
       "      <th>hg_dash_n</th>\n",
       "      <th>mm_dash_n</th>\n",
       "      <th>hg_percent</th>\n",
       "      <th>mm_percent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr5</td>\n",
       "      <td>20759748</td>\n",
       "      <td>21059748</td>\n",
       "      <td>.</td>\n",
       "      <td>chr5</td>\n",
       "      <td>20766763</td>\n",
       "      <td>20767471</td>\n",
       "      <td>709</td>\n",
       "      <td>TCTGTAAGCACAATACTTTGTAT---TTTTTGACCAACCTACCTTA...</td>\n",
       "      <td>chr15</td>\n",
       "      <td>81698495</td>\n",
       "      <td>81699204</td>\n",
       "      <td>710</td>\n",
       "      <td>TCTGTAC-CACAATAATCTTTATACCTTTTCGAGCTACCAACTTTC...</td>\n",
       "      <td>744</td>\n",
       "      <td>35</td>\n",
       "      <td>34</td>\n",
       "      <td>0.9529569892473119</td>\n",
       "      <td>0.9543010752688172</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr5</td>\n",
       "      <td>20759748</td>\n",
       "      <td>21059748</td>\n",
       "      <td>.</td>\n",
       "      <td>chr5</td>\n",
       "      <td>20767971</td>\n",
       "      <td>20769891</td>\n",
       "      <td>1921</td>\n",
       "      <td>CTGGTTGGTAGATTGCTACCTCCTtgattaacagtggtaaatattt...</td>\n",
       "      <td>chr15</td>\n",
       "      <td>81699205</td>\n",
       "      <td>81701043</td>\n",
       "      <td>1839</td>\n",
       "      <td>CTGGTTACTGGAGTGCTGCCTTCTTGATTAACAGAGTTGAATATTT...</td>\n",
       "      <td>1973</td>\n",
       "      <td>52</td>\n",
       "      <td>134</td>\n",
       "      <td>0.9736441966548404</td>\n",
       "      <td>0.9320831221490117</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr5</td>\n",
       "      <td>20759748</td>\n",
       "      <td>21059748</td>\n",
       "      <td>.</td>\n",
       "      <td>chr5</td>\n",
       "      <td>20769910</td>\n",
       "      <td>20769953</td>\n",
       "      <td>44</td>\n",
       "      <td>GATATAACACTTTCATACAAATGTCATGAAATAGCAGACATTTT</td>\n",
       "      <td>chr15</td>\n",
       "      <td>81701273</td>\n",
       "      <td>81701316</td>\n",
       "      <td>44</td>\n",
       "      <td>GGTAAGATATGCTTATATAAATGTCATGTAATAACAAATGTTTT</td>\n",
       "      <td>44</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr5</td>\n",
       "      <td>20759748</td>\n",
       "      <td>21059748</td>\n",
       "      <td>.</td>\n",
       "      <td>chr5</td>\n",
       "      <td>20778823</td>\n",
       "      <td>20779823</td>\n",
       "      <td>1001</td>\n",
       "      <td>AATAGACTTGGGAAATGCGATGTTTTCTAGGCATGATCCTGAAATC...</td>\n",
       "      <td>chr15</td>\n",
       "      <td>81708469</td>\n",
       "      <td>81709456</td>\n",
       "      <td>988</td>\n",
       "      <td>AAGAGAGTTAGGAAATATGGT-CTTTCTTGGAGTAAAGCTGAATTA...</td>\n",
       "      <td>1061</td>\n",
       "      <td>60</td>\n",
       "      <td>73</td>\n",
       "      <td>0.943449575871819</td>\n",
       "      <td>0.9311969839773798</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr5</td>\n",
       "      <td>20759748</td>\n",
       "      <td>21059748</td>\n",
       "      <td>.</td>\n",
       "      <td>chr5</td>\n",
       "      <td>20780510</td>\n",
       "      <td>20781236</td>\n",
       "      <td>727</td>\n",
       "      <td>aactgtagctgactgATCTACAGAAAGGTTTCATTCACTC-AGCGT...</td>\n",
       "      <td>chr15</td>\n",
       "      <td>81710628</td>\n",
       "      <td>81711251</td>\n",
       "      <td>624</td>\n",
       "      <td>AACTCCACCAGGCTGAC--ATAATCAGGTTTACTTTATTCTGGCAT...</td>\n",
       "      <td>748</td>\n",
       "      <td>21</td>\n",
       "      <td>124</td>\n",
       "      <td>0.9719251336898396</td>\n",
       "      <td>0.8342245989304813</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  CHROM  POS_300KB  END_300KB ID hg_chr  hg_start    hg_end hg_len  \\\n",
       "0  chr5   20759748   21059748  .   chr5  20766763  20767471    709   \n",
       "1  chr5   20759748   21059748  .   chr5  20767971  20769891   1921   \n",
       "2  chr5   20759748   21059748  .   chr5  20769910  20769953     44   \n",
       "3  chr5   20759748   21059748  .   chr5  20778823  20779823   1001   \n",
       "4  chr5   20759748   21059748  .   chr5  20780510  20781236    727   \n",
       "\n",
       "                                              hg_seq mm_chr  mm_start  \\\n",
       "0  TCTGTAAGCACAATACTTTGTAT---TTTTTGACCAACCTACCTTA...  chr15  81698495   \n",
       "1  CTGGTTGGTAGATTGCTACCTCCTtgattaacagtggtaaatattt...  chr15  81699205   \n",
       "2       GATATAACACTTTCATACAAATGTCATGAAATAGCAGACATTTT  chr15  81701273   \n",
       "3  AATAGACTTGGGAAATGCGATGTTTTCTAGGCATGATCCTGAAATC...  chr15  81708469   \n",
       "4  aactgtagctgactgATCTACAGAAAGGTTTCATTCACTC-AGCGT...  chr15  81710628   \n",
       "\n",
       "     mm_end mm_len                                             mm_seq  \\\n",
       "0  81699204    710  TCTGTAC-CACAATAATCTTTATACCTTTTCGAGCTACCAACTTTC...   \n",
       "1  81701043   1839  CTGGTTACTGGAGTGCTGCCTTCTTGATTAACAGAGTTGAATATTT...   \n",
       "2  81701316     44       GGTAAGATATGCTTATATAAATGTCATGTAATAACAAATGTTTT   \n",
       "3  81709456    988  AAGAGAGTTAGGAAATATGGT-CTTTCTTGGAGTAAAGCTGAATTA...   \n",
       "4  81711251    624  AACTCCACCAGGCTGAC--ATAATCAGGTTTACTTTATTCTGGCAT...   \n",
       "\n",
       "  total_align_len hg_dash_n mm_dash_n          hg_percent          mm_percent  \n",
       "0             744        35        34  0.9529569892473119  0.9543010752688172  \n",
       "1            1973        52       134  0.9736441966548404  0.9320831221490117  \n",
       "2              44         0         0                 1.0                 1.0  \n",
       "3            1061        60        73   0.943449575871819  0.9311969839773798  \n",
       "4             748        21       124  0.9719251336898396  0.8342245989304813  "
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "variant_300kb_synt.columns = list(variants_300kb.columns) + list(synteny_condensed.columns)\n",
    "variant_300kb_synt = variant_300kb_synt[variant_300kb_synt['hg_chr'] != '.']\n",
    "print(len(variant_300kb_synt))\n",
    "variant_300kb_synt.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "1d2062bb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "372089\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS_1MB</th>\n",
       "      <th>END_1MB</th>\n",
       "      <th>ID</th>\n",
       "      <th>hg_chr</th>\n",
       "      <th>hg_start</th>\n",
       "      <th>hg_end</th>\n",
       "      <th>hg_len</th>\n",
       "      <th>hg_seq</th>\n",
       "      <th>mm_chr</th>\n",
       "      <th>mm_start</th>\n",
       "      <th>mm_end</th>\n",
       "      <th>mm_len</th>\n",
       "      <th>mm_seq</th>\n",
       "      <th>total_align_len</th>\n",
       "      <th>hg_dash_n</th>\n",
       "      <th>mm_dash_n</th>\n",
       "      <th>hg_percent</th>\n",
       "      <th>mm_percent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr5</td>\n",
       "      <td>20409748</td>\n",
       "      <td>21409748</td>\n",
       "      <td>.</td>\n",
       "      <td>chr5</td>\n",
       "      <td>20427363</td>\n",
       "      <td>20427405</td>\n",
       "      <td>43</td>\n",
       "      <td>GAATAAACATCACATATTATCTTAAAAGGCTTATTTTACTCAA</td>\n",
       "      <td>chr15</td>\n",
       "      <td>81368472</td>\n",
       "      <td>81368514</td>\n",
       "      <td>43</td>\n",
       "      <td>GAGTAAACACTGAACATAACCTTAAAAGATTTATTTGGCCCAA</td>\n",
       "      <td>43</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr5</td>\n",
       "      <td>20409748</td>\n",
       "      <td>21409748</td>\n",
       "      <td>.</td>\n",
       "      <td>chr5</td>\n",
       "      <td>20428715</td>\n",
       "      <td>20429577</td>\n",
       "      <td>863</td>\n",
       "      <td>atgtttgtGACCCAATGTGTTTTACAATAAGACGTATATAAGCAAA...</td>\n",
       "      <td>chr15</td>\n",
       "      <td>81369922</td>\n",
       "      <td>81370730</td>\n",
       "      <td>809</td>\n",
       "      <td>ATATTTTTTATGCACTATGCTTTT--ACAGAATGCATATATGCATA...</td>\n",
       "      <td>875</td>\n",
       "      <td>12</td>\n",
       "      <td>66</td>\n",
       "      <td>0.9862857142857143</td>\n",
       "      <td>0.9245714285714286</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr5</td>\n",
       "      <td>20409748</td>\n",
       "      <td>21409748</td>\n",
       "      <td>.</td>\n",
       "      <td>chr5</td>\n",
       "      <td>20430941</td>\n",
       "      <td>20431130</td>\n",
       "      <td>190</td>\n",
       "      <td>AAAGAATGACAATACATTCCTGCATCTATTTCTCCTAATATATTTA...</td>\n",
       "      <td>chr15</td>\n",
       "      <td>81374457</td>\n",
       "      <td>81374661</td>\n",
       "      <td>205</td>\n",
       "      <td>AAAGAGT---AAGACATTTCT--ATTTAT--CTGTTAATATGTTTG...</td>\n",
       "      <td>219</td>\n",
       "      <td>29</td>\n",
       "      <td>14</td>\n",
       "      <td>0.8675799086757991</td>\n",
       "      <td>0.9360730593607306</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr5</td>\n",
       "      <td>20409748</td>\n",
       "      <td>21409748</td>\n",
       "      <td>.</td>\n",
       "      <td>chr5</td>\n",
       "      <td>20431528</td>\n",
       "      <td>20432741</td>\n",
       "      <td>1214</td>\n",
       "      <td>AAAGAAAAgaacctcagtcaatcacaaagagccaaccagcctattg...</td>\n",
       "      <td>chr15</td>\n",
       "      <td>81374698</td>\n",
       "      <td>81375849</td>\n",
       "      <td>1152</td>\n",
       "      <td>AAAACAATGAACCTTGTCAATTTTTGAACAGTAAACCAAT-TGTCA...</td>\n",
       "      <td>1290</td>\n",
       "      <td>76</td>\n",
       "      <td>138</td>\n",
       "      <td>0.9410852713178295</td>\n",
       "      <td>0.8930232558139535</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr5</td>\n",
       "      <td>20409748</td>\n",
       "      <td>21409748</td>\n",
       "      <td>.</td>\n",
       "      <td>chr5</td>\n",
       "      <td>20434929</td>\n",
       "      <td>20435479</td>\n",
       "      <td>551</td>\n",
       "      <td>CTTGGAAGGGGTATGCATACATTCAGATGGATAGCCACTGCTTTAT...</td>\n",
       "      <td>chr15</td>\n",
       "      <td>81376313</td>\n",
       "      <td>81376851</td>\n",
       "      <td>539</td>\n",
       "      <td>CTTCAGGAGGAAATGCATTCTGTAAGATGTATTAGCATTGGCCTGC...</td>\n",
       "      <td>574</td>\n",
       "      <td>23</td>\n",
       "      <td>35</td>\n",
       "      <td>0.9599303135888502</td>\n",
       "      <td>0.9390243902439024</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  CHROM   POS_1MB   END_1MB ID hg_chr  hg_start    hg_end hg_len  \\\n",
       "0  chr5  20409748  21409748  .   chr5  20427363  20427405     43   \n",
       "1  chr5  20409748  21409748  .   chr5  20428715  20429577    863   \n",
       "2  chr5  20409748  21409748  .   chr5  20430941  20431130    190   \n",
       "3  chr5  20409748  21409748  .   chr5  20431528  20432741   1214   \n",
       "4  chr5  20409748  21409748  .   chr5  20434929  20435479    551   \n",
       "\n",
       "                                              hg_seq mm_chr  mm_start  \\\n",
       "0        GAATAAACATCACATATTATCTTAAAAGGCTTATTTTACTCAA  chr15  81368472   \n",
       "1  atgtttgtGACCCAATGTGTTTTACAATAAGACGTATATAAGCAAA...  chr15  81369922   \n",
       "2  AAAGAATGACAATACATTCCTGCATCTATTTCTCCTAATATATTTA...  chr15  81374457   \n",
       "3  AAAGAAAAgaacctcagtcaatcacaaagagccaaccagcctattg...  chr15  81374698   \n",
       "4  CTTGGAAGGGGTATGCATACATTCAGATGGATAGCCACTGCTTTAT...  chr15  81376313   \n",
       "\n",
       "     mm_end mm_len                                             mm_seq  \\\n",
       "0  81368514     43        GAGTAAACACTGAACATAACCTTAAAAGATTTATTTGGCCCAA   \n",
       "1  81370730    809  ATATTTTTTATGCACTATGCTTTT--ACAGAATGCATATATGCATA...   \n",
       "2  81374661    205  AAAGAGT---AAGACATTTCT--ATTTAT--CTGTTAATATGTTTG...   \n",
       "3  81375849   1152  AAAACAATGAACCTTGTCAATTTTTGAACAGTAAACCAAT-TGTCA...   \n",
       "4  81376851    539  CTTCAGGAGGAAATGCATTCTGTAAGATGTATTAGCATTGGCCTGC...   \n",
       "\n",
       "  total_align_len hg_dash_n mm_dash_n          hg_percent          mm_percent  \n",
       "0              43         0         0                 1.0                 1.0  \n",
       "1             875        12        66  0.9862857142857143  0.9245714285714286  \n",
       "2             219        29        14  0.8675799086757991  0.9360730593607306  \n",
       "3            1290        76       138  0.9410852713178295  0.8930232558139535  \n",
       "4             574        23        35  0.9599303135888502  0.9390243902439024  "
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "variant_1mb_synt.columns = list(variants_1mb.columns) + list(synteny_condensed.columns)\n",
    "variant_1mb_synt = variant_1mb_synt[variant_1mb_synt['hg_chr'] != '.']\n",
    "print(len(variant_1mb_synt))\n",
    "variant_1mb_synt.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "80feb0fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_index(hg_seq, idx):\n",
    "    count_letters = 0  # To count the letters encountered in the combined string\n",
    "    idx_new = 0  # Corresponding index in combined string\n",
    "    \n",
    "    for char in hg_seq:\n",
    "        if count_letters == (idx+1):\n",
    "            break\n",
    "        if char.isalpha():\n",
    "            count_letters += 1\n",
    "        idx_new += 1\n",
    "    \n",
    "    return idx_new\n",
    "\n",
    "#given two aligned strings with start and end idx\n",
    "#count up amount of same characters within the alignment subset \n",
    "def count_str(hg_seq, mm_seq, start_idx, end_idx):\n",
    "    #print('start_idx', start_idx)\n",
    "    #print('end_idx', end_idx)\n",
    "    hg_subset = hg_seq[start_idx:end_idx]\n",
    "    mm_subset = mm_seq[start_idx:end_idx]\n",
    "\n",
    "    count_same = sum(1 for a, b in zip(hg_subset, mm_subset) if a == b)\n",
    "    \n",
    "    if end_idx>start_idx:\n",
    "        prop=count_same/(end_idx-start_idx)\n",
    "    else:\n",
    "        prop=0\n",
    "    \n",
    "    \n",
    "    return(prop)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "66a007f2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "index  0\n",
      "index  10\n",
      "index  20\n",
      "index  30\n",
      "index  40\n",
      "index  50\n",
      "index  60\n",
      "index  70\n",
      "index  80\n",
      "index  90\n",
      "index  100\n",
      "index  110\n",
      "index  120\n",
      "index  130\n",
      "index  140\n",
      "index  150\n",
      "index  160\n",
      "index  170\n",
      "index  180\n",
      "index  190\n",
      "index  200\n",
      "index  210\n",
      "index  220\n",
      "index  230\n",
      "index  250\n",
      "index  260\n",
      "index  270\n",
      "index  280\n",
      "index  290\n",
      "index  300\n",
      "index  310\n",
      "index  320\n",
      "index  330\n",
      "index  340\n",
      "index  350\n",
      "index  360\n",
      "index  370\n",
      "index  380\n",
      "index  390\n",
      "index  400\n",
      "index  410\n",
      "index  420\n",
      "index  430\n",
      "index  440\n",
      "index  450\n",
      "index  460\n",
      "index  470\n",
      "index  480\n",
      "index  490\n",
      "index  500\n",
      "index  510\n",
      "index  520\n",
      "index  530\n",
      "index  540\n",
      "index  550\n",
      "index  560\n",
      "index  570\n",
      "index  580\n",
      "index  590\n",
      "index  600\n",
      "index  610\n",
      "index  620\n",
      "index  630\n",
      "index  640\n",
      "index  650\n",
      "index  660\n",
      "index  670\n",
      "index  680\n",
      "index  690\n",
      "index  700\n",
      "error in index 706\n"
     ]
    }
   ],
   "source": [
    "start=time.time()\n",
    "for i in simple.index:\n",
    "   \n",
    "    if i % 10 == 0:\n",
    "        print('index ', i)\n",
    "    \n",
    "    \n",
    "    #THIS ONE FOR 1MB SCORE \n",
    "    try:\n",
    "        #print(i)\n",
    "        variant=simple.loc[i]\n",
    "        ID=variant.ID\n",
    "        int_df=variant_1mb_synt[variant_1mb_synt['ID'] == ID]\n",
    "\n",
    "        #print(len(int_df))\n",
    "    \n",
    "    \n",
    "        if len(int_df)>0:\n",
    "\n",
    "\n",
    "            int_df['start_idx']=(np.maximum(int_df['POS_1MB'], int_df['hg_start'])) - int_df['hg_start'] \n",
    "            int_df['end_idx']=(np.minimum(int_df['END_1MB'], int_df['hg_end'])) - int_df['hg_start'] \n",
    "            int_df['n_overlap']=int_df['end_idx']-int_df['start_idx']\n",
    "\n",
    "            \n",
    "            #get alignment indices\n",
    "            int_df['start_align_idx'] = int_df.apply(lambda row: find_index(row['hg_seq'], row['start_idx']), axis=1)\n",
    "            int_df['end_align_idx'] = int_df.apply(lambda row: find_index(row['hg_seq'], row['end_idx']), axis=1)\n",
    "            int_df['n_align_overlap']=int_df['end_align_idx']-int_df['start_align_idx']\n",
    "        \n",
    "        \n",
    "            int_df['count_prop'] = int_df.apply(lambda row: count_str(row['hg_seq'], row['mm_seq'], row['start_align_idx'], row['end_align_idx']), axis=1)\n",
    "\n",
    "\n",
    "            #proportion of SV that covers a syntenic region  \n",
    "            simple.loc[i, 'synt_prop_overlap_1MB'] = int(np.sum(int_df['n_overlap'])) /abs(simple.loc[i, 'postoend_1mb'])\n",
    "            #weighted proportion of SV in syntenic region, and how similar those regions are \n",
    "            simple.loc[i, 'align_synt_prop_1MB'] =  (np.sum(int_df['count_prop'] *int_df['n_overlap'])) /abs(simple.loc[i, 'postoend_1mb'])\n",
    "\n",
    "\n",
    "\n",
    "        else:\n",
    "            simple.loc[i, 'synt_prop_overlap_1MB'] = 0\n",
    "            simple.loc[i, 'align_synt_prop_1MB'] = 0\n",
    "            \n",
    "    except:\n",
    "        print('error in index', i)\n",
    "        \n",
    "        \n",
    "    #THIS ONE FOR NORMAL SV LENGTH \n",
    "    try:\n",
    "        #print(i)\n",
    "        variant=simple.loc[i]\n",
    "        ID=variant.ID\n",
    "        int_df=variant_300kb_synt[variant_300kb_synt['ID'] == ID]\n",
    "        #print(len(int_df))\n",
    "    \n",
    "    \n",
    "        if len(int_df)>0:\n",
    "\n",
    "\n",
    "            int_df['start_idx']=(np.maximum(int_df['POS_300KB'], int_df['hg_start'])) - int_df['hg_start'] \n",
    "            int_df['end_idx']=(np.minimum(int_df['END_300KB'], int_df['hg_end'])) - int_df['hg_start'] \n",
    "            int_df['n_overlap']=int_df['end_idx']-int_df['start_idx']\n",
    "\n",
    "            \n",
    "            #get alignment indices\n",
    "            int_df['start_align_idx'] = int_df.apply(lambda row: find_index(row['hg_seq'], row['start_idx']), axis=1)\n",
    "            int_df['end_align_idx'] = int_df.apply(lambda row: find_index(row['hg_seq'], row['end_idx']), axis=1)\n",
    "            int_df['n_align_overlap']=int_df['end_align_idx']-int_df['start_align_idx']\n",
    "        \n",
    "        \n",
    "            int_df['count_prop'] = int_df.apply(lambda row: count_str(row['hg_seq'], row['mm_seq'], row['start_align_idx'], row['end_align_idx']), axis=1)\n",
    "\n",
    "\n",
    "            #proportion of SV that covers a syntenic region  \n",
    "            simple.loc[i, 'synt_prop_overlap_300KB'] = int(np.sum(int_df['n_overlap'])) /abs(simple.loc[i, 'postoend_300kb'])\n",
    "            #weighted proportion of SV in syntenic region, and how similar those regions are \n",
    "            simple.loc[i, 'align_synt_prop_300KB'] =  (np.sum(int_df['count_prop'] *int_df['n_overlap'])) /abs(simple.loc[i, 'postoend_300kb'])\n",
    "\n",
    "\n",
    "\n",
    "        else:\n",
    "            simple.loc[i, 'synt_prop_overlap_300KB'] = 0\n",
    "            simple.loc[i, 'align_synt_prop_300KB'] = 0\n",
    "            \n",
    "    except:\n",
    "        print('error 1mb in index', i)\n",
    "        \n",
    "        \n",
    "end=time.time()\n",
    "print(f\"Time taken: {end-start} seconds\")\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9d59a081",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "86f9c6df",
   "metadata": {},
   "outputs": [],
   "source": [
    "simple.to_csv('/pollard/home/shzhang/20231023_simple_top500_synteny_1mb.txt', sep='\\t', index=False)  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d739d134",
   "metadata": {},
   "outputs": [],
   "source": [
    "simple"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "efdedd65",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bd34dcd2",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0f88397b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7c43db67",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
